From b0f9731753c5fd006538b1fc4fdeed34392aecc0 Mon Sep 17 00:00:00 2001
From: Andrew Nguyen <arethusa26@gmail.com>
Date: Thu, 16 Oct 2025 15:04:20 +1100
Subject: [PATCH] iccvid: Support message ICM_DECOMPRESSEX_BEGIN

---
 dlls/iccvid/iccvid.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/iccvid/iccvid.c b/dlls/iccvid/iccvid.c
index 56a882af659..40cd4a25b27 100644
--- a/dlls/iccvid/iccvid.c
+++ b/dlls/iccvid/iccvid.c
@@ -1022,6 +1022,15 @@ LRESULT WINAPI ICCVID_DriverProc( DWORD_PTR dwDriverId, HDRVR hdrvr, UINT msg,
     case ICM_DECOMPRESS:
         return ICCVID_Decompress( info, (ICDECOMPRESS*) lParam1,
                                   (DWORD) lParam2 );
+    case ICM_DECOMPRESSEX_BEGIN:
+    {
+        ICDECOMPRESSEX *icdex = (ICDECOMPRESSEX *)lParam1;
+
+        FIXME("extended parameters not handled\n");
+
+        return ICCVID_DecompressBegin( info, (LPBITMAPINFO)icdex->lpbiSrc,
+                                       (LPBITMAPINFO)icdex->lpbiDst );
+    }
     case ICM_DECOMPRESSEX:
         return ICCVID_DecompressEx( info, (ICDECOMPRESSEX*) lParam1, 
                                   (DWORD) lParam2 );
-- 
2.51.0

