From e2a538dcaee4c59055b877b63d143a93a6e6abdb Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Tue, 15 Dec 2020 00:20:43 +0300
Subject: [PATCH] server: Fix process access rights in get_mapping_file
 handler.

---
 server/mapping.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/mapping.c b/server/mapping.c
index d5f889771e50..3df9fb97a1c5 100644
--- a/server/mapping.c
+++ b/server/mapping.c
@@ -1146,7 +1146,7 @@ DECL_HANDLER(get_mapping_file)
     struct process *process;
     struct file *file;
 
-    if (!(process = get_process_from_handle( req->process, 0 ))) return;
+    if (!(process = get_process_from_handle( req->process, PROCESS_QUERY_INFORMATION ))) return;
 
     LIST_FOR_EACH_ENTRY( view, &process->views, struct memory_view, entry )
         if (req->addr >= view->base && req->addr < view->base + view->size) break;
-- 
2.29.2

