From d62c72b607a0afed782df3b1c0bdba6eec5a538e Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 16 Oct 2020 16:53:29 -0500
Subject: [PATCH] mfmediaengine: Provide the partial topology to the media
 session.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/mfmediaengine/main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
index d93e2b2f38c..d1261e3cd73 100644
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -418,7 +418,10 @@ static HRESULT media_engine_create_topology(struct media_engine *engine, IMFMedi
             {
                 sd_video = sd;
                 IMFStreamDescriptor_AddRef(sd_video);
+                /* TODO: reintroduce this once we set up video stream nodes */
+#if 0
                 IMFPresentationDescriptor_SelectStream(pd, i);
+#endif
             }
 
             IMFMediaTypeHandler_Release(type_handler);
@@ -480,6 +483,9 @@ static HRESULT media_engine_create_topology(struct media_engine *engine, IMFMedi
             if (audio_src)
                 IMFTopologyNode_Release(audio_src);
         }
+
+        if (SUCCEEDED(hr))
+            hr = IMFMediaSession_SetTopology(engine->session, MFSESSION_SETTOPOLOGY_IMMEDIATE, topology);
     }
 
     if (topology)
-- 
2.28.0

