From 0979488b998dde4074fe4f443971fc90534c3fad Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 16 Oct 2020 16:54:57 -0500
Subject: [PATCH] mfmediaengine: Issue MF_MEDIA_ENGINE_EVENT_CANPLAY upon
 MF_TOPOSTATUS_READY.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/mfmediaengine/main.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
index d1261e3cd73..bc89b2f702a 100644
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -295,6 +295,14 @@ static HRESULT WINAPI media_engine_session_events_Invoke(IMFAsyncCallback *iface
             IMFMediaEngineNotify_EventNotify(engine->callback, event_type == MEBufferingStarted ?
                     MF_MEDIA_ENGINE_EVENT_BUFFERINGSTARTED : MF_MEDIA_ENGINE_EVENT_BUFFERINGENDED, 0, 0);
             break;
+        case MESessionTopologyStatus:
+        {
+            UINT32 topo_status = 0;
+            IMFMediaEvent_GetUINT32(event, &MF_EVENT_TOPOLOGY_STATUS, &topo_status);
+            if (topo_status == MF_TOPOSTATUS_READY)
+                IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_CANPLAY, 0, 0);
+            break;
+        }
     }
 
 failed:
-- 
2.28.0

