From 485aa86f6b509c71e657c726df99c770ca2146bc Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 20 Nov 2020 12:50:16 -0600
Subject: [PATCH] mf/topoloader: Move node connection responsibility to
 connection function.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/mf/topology.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/dlls/mf/topology.c b/dlls/mf/topology.c
index 8032e438b73..a09f6ef0ef4 100644
--- a/dlls/mf/topology.c
+++ b/dlls/mf/topology.c
@@ -2022,8 +2022,6 @@ static HRESULT topology_loader_resolve_branch(struct topoloader_context *context
     if (FAILED(IMFTopology_GetNodeByID(context->output_topology, id, &node)))
         topology_loader_clone_node(context, downstream_node, &node, context->marker + 1);
 
-    IMFTopologyNode_ConnectOutput(upstream_node, output_index, node, input_index);
-
     IMFTopologyNode_GetNodeType(upstream_node, &u_type);
     IMFTopologyNode_GetNodeType(downstream_node, &d_type);
 
@@ -2033,7 +2031,7 @@ static HRESULT topology_loader_resolve_branch(struct topoloader_context *context
         return E_FAIL;
     }
 
-    return connectors[u_type][d_type](context, upstream_node, output_index, downstream_node, input_index);
+    return connectors[u_type][d_type](context, upstream_node, output_index, node, input_index);
 }
 
 static HRESULT topology_loader_resolve_nodes(struct topoloader_context *context, unsigned int *layer_size)
-- 
2.29.2

