From d60f2f4d4e5e4eb790e5d065afb554a008506d4f Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 20 Nov 2020 12:53:20 -0600
Subject: [PATCH] mf/topoloader: Unstub IMFTopologyLoader::Load.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/mf/topology.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/dlls/mf/topology.c b/dlls/mf/topology.c
index 4756a30895b..99f3c12f190 100644
--- a/dlls/mf/topology.c
+++ b/dlls/mf/topology.c
@@ -2430,11 +2430,9 @@ static HRESULT WINAPI topology_loader_Load(IMFTopoLoader *iface, IMFTopology *in
             break;
     }
 
-    /* For now return original topology. */
-
     *ret_topology = output_topology;
 
-    return IMFTopology_CloneFrom(output_topology, input_topology);
+    return hr;
 }
 
 static const IMFTopoLoaderVtbl topologyloadervtbl =
-- 
2.29.2

