From 364f264ab7d77047fe9082cdeab885fe0461c68f Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Wed, 14 Oct 2020 11:07:05 -0500
Subject: [PATCH] mf/session: Unconditionally deliver NULL (EOS) samples.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/mf/session.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 1a7439a13c3..07e29cd013f 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -2858,11 +2858,12 @@ static void session_deliver_sample_to_node(struct media_session *session, IMFTop
                 LIST_FOR_EACH_ENTRY_SAFE(sample_entry, sample_entry2, &topo_node->u.transform.outputs[i].samples,
                         struct sample, entry)
                 {
-                    if (!topo_node->u.transform.outputs[i].requests)
+                    if (!topo_node->u.transform.outputs[i].requests && sample_entry->sample)
                         break;
 
                     session_deliver_sample_to_node(session, downstream_node, downstream_input, sample_entry->sample);
-                    topo_node->u.transform.outputs[i].requests--;
+                    if (sample_entry->sample)
+                        topo_node->u.transform.outputs[i].requests--;
 
                     transform_release_sample(sample_entry);
                 }
-- 
2.29.2

