From fea795145aa7b64bbdc99a79c162226d1caa1c77 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Mon, 11 May 2020 16:03:09 -0500
Subject: [PATCH] winegstreamer: Translate MPEG-4 Section-2 attributes to caps.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/winegstreamer/mfplat.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/winegstreamer/mfplat.c b/dlls/winegstreamer/mfplat.c
index 9b75a709168..bd2a0ed1cb2 100644
--- a/dlls/winegstreamer/mfplat.c
+++ b/dlls/winegstreamer/mfplat.c
@@ -1026,6 +1026,14 @@ GstCaps *caps_from_mf_media_type(IMFMediaType *type)
 
             user_data_to_codec_data(type, output);
         }
+        else if (IsEqualGUID(&subtype, &MFVideoFormat_M4S2))
+        {
+            output = gst_caps_new_empty_simple("video/mpeg");
+            gst_caps_set_simple(output, "mpegversion", G_TYPE_INT, 4, NULL);
+            gst_caps_set_simple(output, "systemstream", G_TYPE_BOOLEAN, FALSE, NULL);
+
+            user_data_to_codec_data(type, output);
+        }
         else
         {
             GstVideoFormat format = GST_VIDEO_FORMAT_UNKNOWN;
-- 
2.28.0

