From 0f61590d7908f62de6f1771e49743245d7ff34bd Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 15 Oct 2020 12:18:10 -0500
Subject: [PATCH] HACK: Flush decoder when changing times.

---
 dlls/mf/session.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 6fce432da72..856daebf6b4 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -2094,7 +2094,10 @@ static void session_set_presentation_clock(struct media_session *session)
     LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
     {
         if (node->type == MF_TOPOLOGY_TRANSFORM_NODE)
+        {
+            IMFTransform_ProcessMessage(node->object.transform, MFT_MESSAGE_COMMAND_FLUSH, 0);
             IMFTransform_ProcessMessage(node->object.transform, MFT_MESSAGE_NOTIFY_START_OF_STREAM, 0);
+        }
     }
 
     if (!(session->presentation.flags & SESSION_FLAG_PRESENTATION_CLOCK_SET))
-- 
2.28.0

