From 7afa96cafc4f0c15360439b161df866eb2bec29b Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@codeweavers.com>
Date: Tue, 25 Nov 2014 20:31:58 +0100
Subject: [PATCH] user32: Try harder to find a target for mouse messages

---
 dlls/win32u/message.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/message.c b/dlls/win32u/message.c
index 0cd20684b4e..08e2d047045 100644
--- a/dlls/win32u/message.c
+++ b/dlls/win32u/message.c
@@ -2463,7 +2463,7 @@ static BOOL process_mouse_message( MSG *msg, UINT hw_id, ULONG_PTR extra_info, H
     {
         HWND orig = msg->hwnd;
 
-        msg->hwnd = window_from_point( msg->hwnd, msg->pt, &hittest );
+        msg->hwnd = window_from_point( 0, msg->pt, &hittest );
         if (!msg->hwnd) /* As a heuristic, try the next window if it's the owner of orig */
         {
             HWND next = get_window_relative( orig, GW_HWNDNEXT );
-- 
2.43.0

