From 1af9d5003cb6d65be02fb68779e208488c242b82 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Sat, 3 Sep 2022 08:57:05 -0600
Subject: [PATCH] kernelbase: Add test for reparse point copy behavior.

Signed-off-by: Erich E. Hoover <erich.e.hoover@gmail.com>
---
 dlls/ntdll/tests/file.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index 6ccdb3c02dc..b1290f9ef7b 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -5762,6 +5762,12 @@ static void test_reparse_points(void)
     bret = MoveFileW(new_path, reparse_path);
     ok(bret, "Failed to move and rename reparse point.\n");
 
+    /* Check copying a reparse point to another location */
+    lstrcpyW(new_path, path);
+    lstrcatW(new_path, new_reparseW);
+    bret = CopyFileW(reparse_path, new_path, TRUE);
+    ok(!bret, "Reparse points cannot be copied.\n");
+
 cleanup:
     /* Cleanup */
     pRtlFreeUnicodeString(&nameW);
-- 
2.35.1

