From 08e37ba50a5dc930c0d46199836d72100008ee31 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Sun, 22 Nov 2020 22:35:50 -0700
Subject: kernelbase: Convert FILE_FLAG_OPEN_REPARSE_POINT for passing to
 ntdll.

Signed-off-by: Erich E. Hoover <erich.e.hoover@gmail.com>
---
 dlls/kernelbase/file.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/kernelbase/file.c b/dlls/kernelbase/file.c
index 5c7b649d805..a9e05be62da 100644
--- a/dlls/kernelbase/file.c
+++ b/dlls/kernelbase/file.c
@@ -727,6 +727,8 @@ static UINT get_nt_file_options( DWORD attributes )
         options |= FILE_SEQUENTIAL_ONLY;
     if (attributes & FILE_FLAG_WRITE_THROUGH)
         options |= FILE_WRITE_THROUGH;
+    if (attributes & FILE_FLAG_OPEN_REPARSE_POINT)
+        options |= FILE_OPEN_REPARSE_POINT;
     return options;
 }
 
-- 
2.17.1

