From d41ad5ae59d2d74b2252f4860dd076717bf27e3b Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Sat, 12 Dec 2020 17:40:27 -0700
Subject: server: Properly handle deleting dangling symlinks.

---
 server/fd.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/server/fd.c b/server/fd.c
index 0d8eece7bcb..78b1ae97f03 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -1983,6 +1983,7 @@ struct fd *open_fd( struct fd *root, const char *name, int flags, mode_t *mode,
     {
         fd->unlink_name = path;
         fd->unix_name = realpath( path, NULL );
+        if (!fd->unix_name) fd->unix_name = dup_fd_name( root, name ); /* dangling symlink */
     }
 
     closed_fd->unix_fd = fd->unix_fd;
@@ -2495,7 +2496,7 @@ static void set_fd_disposition( struct fd *fd, int unlink )
             file_set_error();
             return;
         }
-        if (S_ISREG( st.st_mode ))  /* can't unlink files we don't have permission to write */
+        if (S_ISREG( st.st_mode ) || S_ISLNK( st.st_mode ))  /* can't unlink files we don't have permission to write */
         {
             if (!(st.st_mode & (S_IWUSR | S_IWGRP | S_IWOTH)))
             {
-- 
2.17.1

