From 3abdca9dc957d524aa00ccd518aaa2a41a348582 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Sat, 6 Feb 2021 16:15:03 -0700
Subject: [PATCH] ntdll: Rewrite reparse point prefix symlink if the prefix
 location changes.

Signed-off-by: Erich E. Hoover <erich.e.hoover@gmail.com>
---
 dlls/ntdll/unix/file.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 12df28a516e..75153ae7620 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -3903,6 +3903,7 @@ cleanup:
 NTSTATUS get_reparse_point_unix(const char *unix_src, REPARSE_DATA_BUFFER *buffer, ULONG *size)
 {
     char link_dir[PATH_MAX], original_dir[PATH_MAX], *d;
+    const char prefix_string[] = "${WINEPREFIX}";
     int link_path_len, buffer_len, encoded_len;
     char link_path[PATH_MAX];
     REPARSE_DATA_BUFFER header;
@@ -4008,6 +4009,18 @@ NTSTATUS get_reparse_point_unix(const char *unix_src, REPARSE_DATA_BUFFER *buffe
         chdir( link_dir );
     }
 
+    /* if the prefix location has moved then update the Unix prefix passthrough link */
+    strcpy( link_dir, link_path );
+    link_dir[strlen(link_dir)-1] = 0;
+    chdir( link_dir );
+    link_path_len = readlink( prefix_string, link_path, sizeof(link_path) );
+    if (link_path_len > 0) link_path[link_path_len] = 0;
+    if (link_path_len > 0 && strcmp( config_dir, link_path) != 0)
+    {
+        unlink( prefix_string );
+        symlink( config_dir, prefix_string );
+    }
+
     /* revert to the original directory */
     chdir( original_dir );
 
-- 
2.35.1

