From 0e5ab82fbcc56e61c6290fed5451ab9d5d3ef997 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Sat, 6 Feb 2021 16:15:46 -0700
Subject: ntdll: Strip the wine prefix from reparse point paths external to the
 prefix.

Signed-off-by: Erich E. Hoover <erich.e.hoover@gmail.com>
---
 dlls/ntdll/unix/file.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 0c2117ca801..19a9584d4b3 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -6023,6 +6023,33 @@ static void ignore_server_ioctl_struct_holes( ULONG code, const void *in_buffer,
 }
 
 
+void strip_external_path( char *path, SIZE_T *len )
+{
+    static char *unix_root = NULL;
+    static int unix_root_len = 0;
+
+    if (unix_root == NULL)
+    {
+        OBJECT_ATTRIBUTES attr;
+        UNICODE_STRING nameW;
+        WCHAR *nt_name;
+
+        if (unix_to_nt_file_name( "/", &nt_name ) != STATUS_SUCCESS) return;
+        nameW.Buffer = nt_name;
+        nameW.Length = wcslen(nt_name) * sizeof(WCHAR);
+        InitializeObjectAttributes( &attr, &nameW, OBJ_CASE_INSENSITIVE, 0, NULL );
+        nt_to_unix_file_name( &attr, &unix_root, FILE_OPEN );
+        free( nt_name );
+        if (unix_root == NULL) return;
+        unix_root_len = strlen(unix_root);
+    }
+
+    if (strncmp( unix_root, path, unix_root_len ) != 0) return;
+    *len -= unix_root_len;
+    memmove( path, &path[unix_root_len - 1], *len + 1 );
+}
+
+
 /*
  * Retrieve the unix name corresponding to a file handle, remove that directory, and then symlink
  * the requested directory to the location of the old directory.
@@ -6167,6 +6194,8 @@ have_dest:
             goto cleanup;
         }
     }
+    else
+        strip_external_path( unix_dest, &unix_dest_len );
 
     TRACE( "Linking %s to %s\n", unix_src, &unix_dest[relative_offset] );
 
-- 
2.17.1

