From 5952a53690460f2bafac92f10dfeee15aaa709c1 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Sat, 6 Feb 2021 16:16:17 -0700
Subject: [PATCH] ntdll: Add a marker to reparse point paths to indicate the
 prefix location.

---
 dlls/ntdll/unix/file.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 346e1e8c84c..325ac6ed8a1 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -6068,6 +6068,26 @@ void strip_external_path( char *path, SIZE_T *len )
 }
 
 
+char *mark_prefix_end( char *path, SIZE_T *len )
+{
+    static char marker[] = "////.//.//"; /* "P" (0x50) encoded as a path (0=/ 1=./) */
+    int new_path_len = *len + sizeof(marker) - 1;
+    static int config_dir_len = 0;
+    char *new_path;
+
+    if (!config_dir_len) config_dir_len = strlen(config_dir);
+    if (path[config_dir_len] != '/') return path;
+    if (strncmp( config_dir, path, config_dir_len ) != 0) return path;
+    if (!(new_path = malloc( new_path_len ))) return path;
+    *len = new_path_len;
+    strcpy( new_path, config_dir );
+    strcat( new_path, marker );
+    strcat( new_path, &path[config_dir_len] );
+    free( path );
+    return new_path;
+}
+
+
 /*
  * Retrieve the unix name corresponding to a file handle, remove that directory, and then symlink
  * the requested directory to the location of the old directory.
@@ -6213,7 +6233,10 @@ have_dest:
         }
     }
     else
+    {
         strip_external_path( unix_dest, &unix_dest_len );
+        unix_dest = mark_prefix_end( unix_dest, &unix_dest_len );
+    }
 
     TRACE( "Linking %s to %s\n", unix_src, &unix_dest[relative_offset] );
 
-- 
2.30.2

