From f678f5184cb4eef1ab47a4f8b4949c83f1af9791 Mon Sep 17 00:00:00 2001
From: Daniel Wendt <daniel.wendt@linux.com>
Date: Tue, 10 Dec 2013 14:55:32 +0100
Subject: [PATCH] gdi32: fix for rotated ellipse

Bug: http://bugs.winehq.org/show_bug.cgi?id=35331
---
 dlls/gdi32/dibdrv/graphics.c | 60 +++++++++++++++++++++++++-----------
 1 file changed, 42 insertions(+), 18 deletions(-)

diff --git a/dlls/gdi32/dibdrv/graphics.c b/dlls/gdi32/dibdrv/graphics.c
index ea03cac886f..412d19249f0 100644
--- a/dlls/gdi32/dibdrv/graphics.c
+++ b/dlls/gdi32/dibdrv/graphics.c
@@ -1541,6 +1541,23 @@ BOOL CDECL dibdrv_RoundRect( PHYSDEV dev, INT left, INT top, INT right, INT bott
     BOOL ret = TRUE;
     HRGN outline = 0, interior = 0;
 
+    BOOL exclude_rotation_translation = FALSE;
+    XFORM old;
+    XFORM rotation_and_translation;
+
+    if (dc->attr->graphics_mode == GM_ADVANCED)
+    {
+        XFORM xf;
+        NtGdiGetTransform( pdev->dev.hdc, 0x203, &old );
+        xf = old;
+        if (xform_has_rotate_and_uniform_scale_and_shear( &xf ) &&
+            xform_decompose_rotation_and_translation( &xf, &rotation_and_translation ))
+        {
+            NtGdiModifyWorldTransform( pdev->dev.hdc, &xf, MWT_SET );
+            exclude_rotation_translation = TRUE;
+        }
+    }
+
     if (!get_pen_device_rect( dc, pdev, &rect, left, top, right, bottom )) return TRUE;
 
     pt[0].x = pt[0].y = 0;
@@ -1561,23 +1578,6 @@ BOOL CDECL dibdrv_RoundRect( PHYSDEV dev, INT left, INT top, INT right, INT bott
         return FALSE;
     }
 
-    if (pdev->brush.style != BS_NULL &&
-        !(interior = NtGdiCreateRoundRectRgn( rect.left, rect.top, rect.right + 1, rect.bottom + 1,
-                                              ellipse_width, ellipse_height )))
-    {
-        free( points );
-        if (outline) NtGdiDeleteObjectApp( outline );
-        return FALSE;
-    }
-
-    /* if not using a region, paint the interior first so the outline can overlap it */
-    if (interior && !outline)
-    {
-        ret = brush_region( pdev, interior );
-        NtGdiDeleteObjectApp( interior );
-        interior = 0;
-    }
-
     count = ellipse_first_quadrant( ellipse_width, ellipse_height, points );
 
     if (dc->attr->arc_direction == AD_CLOCKWISE)
@@ -1621,13 +1621,37 @@ BOOL CDECL dibdrv_RoundRect( PHYSDEV dev, INT left, INT top, INT right, INT bott
     }
     count = end + 1;
 
+    if (exclude_rotation_translation == TRUE)
+    {
+        NtGdiModifyWorldTransform( pdev->dev.hdc, &rotation_and_translation, MWT_SET );
+        /* apply rotation and translation to calculated points */
+        NtGdiTransformPoints( dev->hdc, points, points, count, NtGdiLPtoDP );
+        /* restore origin matrix */
+        NtGdiModifyWorldTransform( pdev->dev.hdc, &old, MWT_SET );
+    }
+
+    if (pdev->brush.style != BS_NULL &&
+        !(interior = NtGdiPolyPolyDraw( ULongToHandle(ALTERNATE), points, (const UINT *)&count, 1, NtGdiPolyPolygonRgn )))
+    {
+        free( points );
+        if (outline) NtGdiDeleteObjectApp( outline );
+            return FALSE;
+    }
+
+    /* if not using a region, paint the interior first so the outline can overlap it */
+    if (interior && !outline)
+    {
+        ret = brush_region( pdev, interior );
+        NtGdiDeleteObjectApp( interior );
+        interior = 0;
+    }
+
     reset_dash_origin( pdev );
     pdev->pen_lines( pdev, count, points, TRUE, outline );
     add_pen_lines_bounds( pdev, count, points, outline );
 
     if (interior)
     {
-        NtGdiCombineRgn( interior, interior, outline, RGN_DIFF );
         ret = brush_region( pdev, interior );
         NtGdiDeleteObjectApp( interior );
     }
-- 
2.33.0

