From 76d256b461af14a89c54da1cf6c2002118372d1f Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 12 Dec 2014 05:06:31 +0100
Subject: [PATCH] win32u: Fix return value of ScrollWindowEx for invisible
 windows.

---
 dlls/win32u/dce.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/dlls/win32u/dce.c b/dlls/win32u/dce.c
index cfa3510c60b..2315a16d1b4 100644
--- a/dlls/win32u/dce.c
+++ b/dlls/win32u/dce.c
@@ -1631,10 +1631,13 @@ INT WINAPI NtUserScrollWindowEx( HWND hwnd, INT dx, INT dy, const RECT *rect,
     rdw_flags = (flags & SW_ERASE) && (flags & SW_INVALIDATE) ?
         RDW_INVALIDATE | RDW_ERASE  : RDW_INVALIDATE;
 
-    if (!is_window_drawable( hwnd, TRUE )) return ERROR;
     hwnd = get_full_window_handle( hwnd );
 
-    get_client_rect( hwnd, &rc, get_thread_dpi() );
+    if (!is_window_drawable( hwnd, TRUE ))
+        SetRectEmpty( &rc );
+    else
+        get_client_rect( hwnd, &rc, get_thread_dpi() );
+    
     if (clip_rect) intersect_rect( &cliprc, &rc, clip_rect );
     else cliprc = rc;
 
-- 
2.43.0

