From 7720661daf3fe45cfaa42f2b69e55d274d89a2bc Mon Sep 17 00:00:00 2001
From: Jefferson Carpenter <jeffersoncarpenter2@gmail.com>
Date: Thu, 4 Nov 2021 22:30:05 +0000
Subject: [PATCH] msxml3: Write to DOMDocument mxwriter destination in
 endDocument.

---
 dlls/msxml3/mxwriter.c | 43 +++++++++++++++++++++++++++++++++++++++---
 1 file changed, 40 insertions(+), 3 deletions(-)

diff --git a/dlls/msxml3/mxwriter.c b/dlls/msxml3/mxwriter.c
index f766aa85955..17f1e41205e 100644
--- a/dlls/msxml3/mxwriter.c
+++ b/dlls/msxml3/mxwriter.c
@@ -168,6 +168,7 @@ typedef struct
     BSTR element;
 
     IStream *dest;
+    IXMLDOMDocument *dest_doc;
 
     output_buffer buffer;
 } mxwriter;
@@ -845,6 +846,7 @@ static ULONG WINAPI mxwriter_Release(IMXWriter *iface)
         free_output_buffer(&This->buffer);
 
         if (This->dest) IStream_Release(This->dest);
+        if (This->dest_doc) IXMLDOMDocument_Release(This->dest_doc);
         SysFreeString(This->version);
         SysFreeString(This->encoding);
 
@@ -909,6 +911,8 @@ static HRESULT WINAPI mxwriter_put_output(IMXWriter *iface, VARIANT dest)
     {
         if (This->dest) IStream_Release(This->dest);
         This->dest = NULL;
+        if (This->dest_doc) IXMLDOMDocument_Release(This->dest_doc);
+        This->dest_doc = NULL;
         close_output_buffer(This);
         break;
     }
@@ -924,12 +928,33 @@ static HRESULT WINAPI mxwriter_put_output(IMXWriter *iface, VARIANT dest)
 
             if (This->dest) IStream_Release(This->dest);
             This->dest = stream;
+            if (This->dest_doc) IXMLDOMDocument_Release(This->dest_doc);
+            This->dest_doc = NULL;
             break;
         }
 
         FIXME("unhandled interface type for VT_UNKNOWN destination\n");
         return E_NOTIMPL;
     }
+    case VT_DISPATCH:
+    {
+        IXMLDOMDocument *doc;
+
+        hr = IDispatch_QueryInterface(V_DISPATCH(&dest), &IID_IXMLDOMDocument, (void**)&doc);
+        if (hr == S_OK)
+        {
+            close_output_buffer(This);
+
+            if (This->dest) IStream_Release(This->dest);
+            This->dest = NULL;
+            if (This->dest_doc) IXMLDOMDocument_Release(This->dest_doc);
+            This->dest_doc = doc;
+            break;
+        }
+
+        FIXME("unhandled interface type for VT_DISPATCH destination\n");
+        return E_NOTIMPL;
+    }
     default:
         FIXME("unhandled destination type %s\n", debugstr_variant(&dest));
         return E_NOTIMPL;
@@ -1192,8 +1217,8 @@ static ULONG WINAPI SAXContentHandler_Release(ISAXContentHandler *iface)
 
 static HRESULT WINAPI SAXContentHandler_putDocumentLocator(ISAXContentHandler *iface, ISAXLocator *locator)
 {
-    TRACE("%p, %p.\n", iface, locator);
-
+    mxwriter *This = impl_from_ISAXContentHandler( iface );
+    TRACE("(%p)->(%p)\n", This, locator);
     return S_OK;
 }
 
@@ -1232,10 +1257,21 @@ static HRESULT WINAPI SAXContentHandler_startDocument(ISAXContentHandler *iface)
 
 static HRESULT WINAPI SAXContentHandler_endDocument(ISAXContentHandler *iface)
 {
+    HRESULT hr;
+    VARIANT dest;
+    VARIANT_BOOL success;
     mxwriter *This = impl_from_ISAXContentHandler( iface );
     TRACE("(%p)\n", This);
     This->prop_changed = FALSE;
-    return flush_output_buffer(This);
+
+    hr = flush_output_buffer(This);
+    if (FAILED(hr)) return hr;
+
+    if (This->dest_doc) {
+        mxwriter_get_output(&This->IMXWriter_iface, &dest);
+        return IXMLDOMDocument_loadXML(This->dest_doc, V_BSTR(&dest), &success);
+    }
+    return S_OK;
 }
 
 static HRESULT WINAPI SAXContentHandler_startPrefixMapping(
@@ -2610,6 +2646,7 @@ HRESULT MXWriter_create(MSXML_VERSION version, void **ppObj)
     This->newline = FALSE;
 
     This->dest = NULL;
+    This->dest_doc = NULL;
 
     hr = init_output_buffer(This->xml_enc, &This->buffer);
     if (hr != S_OK) {
-- 
2.51.0

