From 00ab0d49080a1c03b95f3ce6c15f0948f4242ba5 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 09/65] dcomp/tests: Add IDCompositionDevice::CreateVisual()
 tests.

---
 dlls/dcomp/tests/dcomp.c | 71 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 71 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index c9e1c25fe81..c2c9ec9c166 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -292,6 +292,76 @@ done:
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_device_CreateVisual(void)
+{
+    IDCompositionDesktopDevice *desktop_device;
+    IDCompositionDevice *dcomp_device;
+    IDCompositionVisual2 *visual2;
+    IDCompositionVisual *visual;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    HRESULT hr;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    /* Test CreateVisual() with IDCompositionDevice */
+    hr = pDCompositionCreateDevice(dxgi_device, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Parameter checks */
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, NULL);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (SUCCEEDED(hr))
+    {
+        /* IDCompositionVisual objects created from a device from DCompositionCreateDevice() doesn't
+         * support IDCompositionVisual2 */
+        hr = IDCompositionVisual_QueryInterface(visual, &IID_IDCompositionVisual2, (void *)&visual2);
+        ok(hr == E_NOINTERFACE, "Got unexpected hr %#lx.\n", hr);
+        IDCompositionVisual_Release(visual);
+    }
+
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+
+    /* Test CreateVisual() with IDCompositionDesktopDevice */
+    if (!pDCompositionCreateDevice2)
+    {
+        win_skip("DCompositionCreateDevice2() is unavailable.\n");
+        goto done;
+    }
+
+    hr = pDCompositionCreateDevice2((IUnknown *)dxgi_device, &IID_IDCompositionDesktopDevice,
+            (void **)&desktop_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Parameter checks */
+    hr = IDCompositionDesktopDevice_CreateVisual(desktop_device, NULL);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionDesktopDevice_CreateVisual(desktop_device, &visual2);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (SUCCEEDED(hr))
+        IDCompositionVisual2_Release(visual2);
+
+    refcount = IDCompositionDesktopDevice_Release(desktop_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+
+done:
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -316,6 +386,7 @@ START_TEST(dcomp)
     test_DCompositionCreateDevice();
     test_DCompositionCreateDevice2();
     test_device_CreateTargetForHwnd();
+    test_device_CreateVisual();
 
     FreeLibrary(module);
 }
-- 
2.51.0

