From e144d713ee8ff05144596fb8e6cc98d99b09f654 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 12/65] dcomp/tests: Add IDCompositionVisual::SetContent()
 tests.

---
 dlls/dcomp/tests/dcomp.c | 101 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 101 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 2eea1d3cbfc..cfcd1d878a6 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -41,6 +41,15 @@ static void check_interface_(unsigned int line, void *iface_ptr, REFIID iid, BOO
         IUnknown_Release(unk);
 }
 
+static HWND create_window(void)
+{
+    RECT r = {0, 0, 640, 480};
+
+    AdjustWindowRect(&r, WS_OVERLAPPEDWINDOW | WS_VISIBLE, FALSE);
+    return CreateWindowW(L"static", L"dcomp_test", WS_OVERLAPPEDWINDOW | WS_VISIBLE, 0, 0,
+            r.right - r.left, r.bottom - r.top, NULL, NULL, NULL, NULL);
+}
+
 static IDXGIDevice *create_device(unsigned int flags)
 {
     IDXGIDevice *dxgi_device;
@@ -67,6 +76,51 @@ success:
     return dxgi_device;
 }
 
+#define get_factory(a, b) get_factory_(__LINE__, a, b)
+static void get_factory_(unsigned int line, IDXGIDevice *device, IDXGIFactory **factory)
+{
+    IDXGIAdapter *adapter;
+    HRESULT hr;
+
+    hr = IDXGIDevice_GetAdapter(device, &adapter);
+    ok_(__FILE__, line)(hr == S_OK, "Failed to get adapter, hr %#lx.\n", hr);
+    hr = IDXGIAdapter_GetParent(adapter, &IID_IDXGIFactory, (void **)factory);
+    ok_(__FILE__, line)(hr == S_OK, "Failed to get parent, hr %#lx.\n", hr);
+    IDXGIAdapter_Release(adapter);
+}
+
+#define create_swapchain(a, b) create_swapchain_(__LINE__, a, b)
+static IDXGISwapChain *create_swapchain_(unsigned int line, IDXGIDevice *device, HWND window)
+{
+    DXGI_SWAP_CHAIN_DESC desc;
+    IDXGISwapChain *swapchain;
+    IDXGIFactory *factory;
+    HRESULT hr;
+
+    desc.BufferDesc.Width = 640;
+    desc.BufferDesc.Height = 480;
+    desc.BufferDesc.RefreshRate.Numerator = 60;
+    desc.BufferDesc.RefreshRate.Denominator = 1;
+    desc.BufferDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
+    desc.BufferDesc.ScanlineOrdering = DXGI_MODE_SCANLINE_ORDER_UNSPECIFIED;
+    desc.BufferDesc.Scaling = DXGI_MODE_SCALING_UNSPECIFIED;
+    desc.SampleDesc.Count = 1;
+    desc.SampleDesc.Quality = 0;
+    desc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+    desc.BufferCount = 1;
+    desc.OutputWindow = window;
+    desc.Windowed = TRUE;
+    desc.SwapEffect =  DXGI_SWAP_EFFECT_SEQUENTIAL;
+    desc.Flags = 0;
+
+    get_factory(device, &factory);
+    hr = IDXGIFactory_CreateSwapChain(factory, (IUnknown *)device, &desc, &swapchain);
+    ok_(__FILE__, line)(hr == S_OK, "Failed to create swapchain, hr %#lx.\n", hr);
+    IDXGIFactory_Release(factory);
+
+    return swapchain;
+}
+
 static void test_DCompositionCreateDevice(void)
 {
     IDCompositionDevice *dcomp_device;
@@ -354,6 +408,52 @@ done:
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_visual_SetContent(void)
+{
+    IDCompositionDevice *dcomp_device;
+    IDCompositionVisual *visual;
+    IDXGISwapChain *swapchain;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    HRESULT hr;
+    HWND hwnd;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hwnd = create_window();
+    swapchain = create_swapchain(dxgi_device, hwnd);
+    hr = pDCompositionCreateDevice(dxgi_device, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Obviously IDCompositionDevice doesn't have IDXGISwapChain1 */
+    hr = IDCompositionVisual_SetContent(visual, (IUnknown *)dcomp_device);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionVisual_SetContent(visual, (IUnknown *)swapchain);
+    todo_wine
+    ok(hr == S_OK || broken(hr == DXGI_ERROR_UNSUPPORTED) /* win8 and win10 v1507 TestBot */,
+            "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionVisual_SetContent(visual, NULL);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    IDCompositionVisual_Release(visual);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    IDXGISwapChain_Release(swapchain);
+    DestroyWindow(hwnd);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -379,6 +479,7 @@ START_TEST(dcomp)
     test_DCompositionCreateDevice2();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
+    test_visual_SetContent();
 
     FreeLibrary(module);
 }
-- 
2.51.0

