From d81fdf6a8cf44e9efba3a6544a034433c76c4b52 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 14/65] dcomp/tests: Add IDCompositionTarget::SetRoot() tests.

---
 dlls/dcomp/tests/dcomp.c | 64 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 64 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index dd5c12a6993..2f959106d8c 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -451,6 +451,69 @@ static void test_visual_SetContent(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_target_SetRoot(void)
+{
+    IDCompositionTarget *target, *target2;
+    IDCompositionDevice *dcomp_device;
+    IDCompositionVisual *visual;
+    IDXGISwapChain *swapchain;
+    IDXGIDevice *dxgi_device;
+    HWND hwnd, hwnd2, hwnd3;
+    ULONG refcount;
+    HRESULT hr;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hwnd = create_window();
+    swapchain = create_swapchain(dxgi_device, hwnd);
+    hr = pDCompositionCreateDevice(dxgi_device, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionVisual_SetContent(visual, (IUnknown *)swapchain);
+    ok(hr == S_OK || broken(hr == DXGI_ERROR_UNSUPPORTED) /* win8 and win10 v1507 TestBot */,
+            "Got unexpected hr %#lx.\n", hr);
+    hwnd2 = create_window();
+    hr = IDCompositionDevice_CreateTargetForHwnd(dcomp_device, hwnd2, TRUE, &target);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hwnd3 = create_window();
+    hr = IDCompositionDevice_CreateTargetForHwnd(dcomp_device, hwnd3, TRUE, &target2);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionTarget_SetRoot(target, visual);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* SetRoot with a visual already bound to a target */
+    hr = IDCompositionTarget_SetRoot(target2, visual);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionTarget_SetRoot(target, NULL);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionTarget_SetRoot(target2, visual);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    DestroyWindow(hwnd3);
+    DestroyWindow(hwnd2);
+    IDCompositionTarget_Release(target2);
+    IDCompositionTarget_Release(target);
+    IDCompositionVisual_Release(visual);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    IDXGISwapChain_Release(swapchain);
+    DestroyWindow(hwnd);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -476,6 +539,7 @@ START_TEST(dcomp)
     test_DCompositionCreateDevice2();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
+    test_target_SetRoot();
     test_visual_SetContent();
 
     FreeLibrary(module);
-- 
2.51.0

