From 219325696b01649938c8a588fdb35a4bcb73702b Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 19/65] dcomp/tests: Add tests for
 DCompositionCreateSharedVisualHandle().

---
 dlls/dcomp/tests/dcomp.c | 71 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 71 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index fe1aefb9e19..81139aec8ef 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -18,7 +18,12 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#include <stdarg.h>
+
+#include "ntstatus.h"
+#define WIN32_NO_STATUS
 #define COBJMACROS
+#include <winternl.h>
 #include "initguid.h"
 #include <d2d1_1.h>
 #include <d3d10_1.h>
@@ -28,6 +33,28 @@
 static HRESULT (WINAPI *pDCompositionCreateDevice)(IDXGIDevice *dxgi_device, REFIID iid, void **device);
 static HRESULT (WINAPI *pDCompositionCreateDevice2)(IUnknown *rendering_device, REFIID iid, void **device);
 
+static void *get_stack_pointer(void)
+{
+    void *stack_pointer = NULL;
+
+#ifdef __i386__
+    __asm__ __volatile__("movl %%esp, %0"
+                         : "=r"(stack_pointer) /* output */
+                         :                     /* no input */
+                         :                     /* no clobbered registers */
+    );
+#elif __x86_64__
+    __asm__ __volatile__("movq %%rsp, %0"
+                         : "=r"(stack_pointer) /* output */
+                         :                     /* no input */
+                         :                     /* no clobbered registers */
+    );
+#else
+#error "Unsupported architecture"
+#endif
+    return stack_pointer;
+}
+
 #define check_interface(a, b, c) check_interface_(__LINE__, a, b, c)
 static void check_interface_(unsigned int line, void *iface_ptr, REFIID iid, BOOL supported)
 {
@@ -285,6 +312,49 @@ static void test_DCompositionCreateDevice2(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_DCompositionCreateSharedVisualHandle(void)
+{
+    HRESULT (WINAPI *pDCompositionCreateSharedVisualHandle)(HANDLE *ret_handle);
+    char buffer[1024] = {0};
+    OBJECT_TYPE_INFORMATION *type = (OBJECT_TYPE_INFORMATION *)buffer;
+    void *stack_pointer, *old_stack_pointer;
+    NTSTATUS status;
+    HMODULE module;
+    HANDLE handle;
+    ULONG len = 0;
+    HRESULT hr;
+    BOOL ret;
+
+    module = GetModuleHandleW(L"dcomp.dll");
+    ok(!!module, "GetModuleHandleW failed.\n");
+
+    pDCompositionCreateSharedVisualHandle = (void *)GetProcAddress(module, (LPCSTR)1040);
+    ok(!!pDCompositionCreateSharedVisualHandle, "Failed to load function at ordinal 1040.\n");
+
+    hr = pDCompositionCreateSharedVisualHandle(NULL);
+    ok(hr == STATUS_INVALID_PARAMETER, "Got unexpected hr %#lx.\n", hr);
+
+    old_stack_pointer = get_stack_pointer();
+
+    handle = NULL;
+    hr = pDCompositionCreateSharedVisualHandle(&handle);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer.\n");
+
+    status = NtQueryObject(handle, ObjectTypeInformation, buffer, sizeof(buffer), &len);
+    todo_wine
+    ok(!status, "Got %#lx.\n", status);
+    if (!status)
+    {
+        ok(!wcscmp(type->TypeName.Buffer, L"Composition"), "Got %s.\n", debugstr_w(type->TypeName.Buffer));
+
+        ret = CloseHandle(handle);
+        ok(ret, "CloseHandle failed.\n");
+    }
+}
+
 static void test_device_CreateTargetForHwnd(void)
 {
     IDCompositionTarget *target, *target2, *target3;
@@ -745,6 +815,7 @@ START_TEST(dcomp)
 
     test_DCompositionCreateDevice();
     test_DCompositionCreateDevice2();
+    test_DCompositionCreateSharedVisualHandle();
     test_device_Commit();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
-- 
2.51.0

