From 292afa67f0735397824cbbe276006ec81a4be35f Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 20/65] dcomp/tests: Add tests for
 DCompositionWaitForCompositorClock().

---
 dlls/dcomp/tests/dcomp.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 81139aec8ef..ece7dfbd291 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -355,6 +355,18 @@ static void test_DCompositionCreateSharedVisualHandle(void)
     }
 }
 
+static void test_DCompositionWaitForCompositorClock(void)
+{
+    DWORD (WINAPI *pDCompositionWaitForCompositorClock)(UINT count, const HANDLE *handles, DWORD timeout);
+    HMODULE module;
+
+    module = GetModuleHandleW(L"dcomp.dll");
+    ok(!!module, "GetModuleHandleW failed.\n");
+
+    pDCompositionWaitForCompositorClock = (void *)GetProcAddress(module, "DCompositionWaitForCompositorClock");
+    ok(!pDCompositionWaitForCompositorClock, "GetProcAddress() succeeded.\n");
+}
+
 static void test_device_CreateTargetForHwnd(void)
 {
     IDCompositionTarget *target, *target2, *target3;
@@ -816,6 +828,7 @@ START_TEST(dcomp)
     test_DCompositionCreateDevice();
     test_DCompositionCreateDevice2();
     test_DCompositionCreateSharedVisualHandle();
+    test_DCompositionWaitForCompositorClock();
     test_device_Commit();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
-- 
2.51.0

