From f4452e86eaafe7be55d077b96b80b28bc3ae3345 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 26/65] dcomp/tests: Add tests for
 IDCompositionDesktopDevicePartner_Unknown1().

---
 dlls/dcomp/tests/dcomp.c | 75 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 75 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index f96b4a6e6d6..020a4939c11 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -890,6 +890,80 @@ static void test_DCompositionCreateDevice3(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_IDCompositionDesktopDevicePartner(void)
+{
+    HRESULT (WINAPI *pDCompositionCreateSharedVisualHandle)(HANDLE *ret_handle);
+    IDCompositionDesktopDevicePartner *partner;
+    void *stack_pointer, *old_stack_pointer;
+    IDCompositionDevice *dcomp_device;
+    HANDLE shared_visual_handle;
+    IDCompositionTarget *target;
+    IDCompositionVisual *visual;
+    IDXGIDevice *dxgi_device;
+    HMODULE module;
+    ULONG refcount;
+    HRESULT hr;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hr = pDCompositionCreateDevice3((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, (void **)&partner);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Test IDCompositionDesktopDevicePartner_Unknown1 */
+    module = GetModuleHandleW(L"dcomp.dll");
+    ok(!!module, "GetModuleHandleW failed.\n");
+
+    pDCompositionCreateSharedVisualHandle = (void *)GetProcAddress(module, (LPCSTR)1040);
+    ok(!!pDCompositionCreateSharedVisualHandle, "Failed to load function at ordinal 1040.\n");
+
+    /* Creating a IDCompositionVisual from a shared visual handle */
+    hr = pDCompositionCreateSharedVisualHandle(&shared_visual_handle);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    old_stack_pointer = get_stack_pointer();
+
+    hr = IDCompositionDesktopDevicePartner_Unknown1(partner, shared_visual_handle, &IID_IDCompositionVisual, (void **)&visual);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer.\n");
+
+    check_interface(visual, &IID_IDCompositionVisual, TRUE);
+    check_interface(visual, &IID_IDCompositionVisual2, TRUE);
+    check_interface(visual, &IID_IDCompositionTarget, FALSE);
+
+    IDCompositionVisual_Release(visual);
+    CloseHandle(shared_visual_handle);
+
+    /* Creating a IDCompositionVisual from a shared visual handle */
+    hr = pDCompositionCreateSharedVisualHandle(&shared_visual_handle);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionDesktopDevicePartner_Unknown1(partner, shared_visual_handle, &IID_IDCompositionTarget, (void **)&target);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    check_interface(target, &IID_IDCompositionVisual, FALSE);
+    check_interface(target, &IID_IDCompositionVisual2, FALSE);
+    check_interface(target, &IID_IDCompositionTarget, TRUE);
+
+    IDCompositionTarget_Release(target);
+    CloseHandle(shared_visual_handle);
+
+    IDCompositionDesktopDevicePartner_Release(partner);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -917,6 +991,7 @@ START_TEST(dcomp)
     test_DCompositionCreateDevice3();
     test_DCompositionCreateSharedVisualHandle();
     test_DCompositionWaitForCompositorClock();
+    test_IDCompositionDesktopDevicePartner();
     test_device_Commit();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
-- 
2.51.0

