From bc81f9c0d04b50b8a9bca5ef213e2e4d3737e02c Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 28/65] dcomp/tests: Add tests for IDCompositionVisualUnknown.

---
 dlls/dcomp/tests/dcomp.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 020a4939c11..596a94e9298 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -506,10 +506,14 @@ static void test_device_CreateVisual(void)
 
     hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    check_interface(visual, &IID_IUnknown, TRUE);
+    check_interface(visual, &IID_IDCompositionVisual, TRUE);
     /* IDCompositionVisual objects created from a device from DCompositionCreateDevice() doesn't
      * support IDCompositionVisual2 */
-    hr = IDCompositionVisual_QueryInterface(visual, &IID_IDCompositionVisual2, (void *)&visual2);
-    ok(hr == E_NOINTERFACE, "Got unexpected hr %#lx.\n", hr);
+    check_interface(visual, &IID_IDCompositionVisual2, FALSE);
+    check_interface(visual, &IID_IDCompositionVisualUnknown, TRUE);
+
     IDCompositionVisual_Release(visual);
 
     refcount = IDCompositionDevice_Release(dcomp_device);
@@ -532,6 +536,13 @@ static void test_device_CreateVisual(void)
 
     hr = IDCompositionDesktopDevice_CreateVisual(desktop_device, &visual2);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    check_interface(visual2, &IID_IUnknown, TRUE);
+    check_interface(visual2, &IID_IDCompositionVisual, TRUE);
+    check_interface(visual2, &IID_IDCompositionVisual2, TRUE);
+    check_interface(visual2, &IID_IDCompositionVisualUnknown, TRUE);
+    check_inherited_interface(visual2, &IID_IDCompositionVisual2, &IID_IDCompositionVisualUnknown);
+
     IDCompositionVisual2_Release(visual2);
 
     refcount = IDCompositionDesktopDevice_Release(desktop_device);
-- 
2.51.0

