From 1292b6626532bb0bbc62a323f4c003036c7ca07f Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 34/65] dcomp/tests: Add tests for IDCompositionDeviceUnknown.

---
 dlls/dcomp/tests/dcomp.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index b5c048ef8dd..8f69760bcc3 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -233,6 +233,8 @@ static void test_DCompositionCreateDevice(void)
     check_interface(dcomp_device, &IID_IDCompositionDevice2, FALSE);
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevice, FALSE);
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, TRUE);
+    check_interface(dcomp_device, &IID_IDCompositionDeviceUnknown, TRUE);
+    check_inherited_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, &IID_IDCompositionDeviceUnknown);
 
     refcount = IDCompositionDevice_Release(dcomp_device);
     ok(!refcount, "Device has %lu references left.\n", refcount);
@@ -302,6 +304,8 @@ static void test_DCompositionCreateDevice2(void)
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevice, TRUE);
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, TRUE);
     check_inherited_interface(dcomp_device, &IID_IDCompositionDesktopDevice, &IID_IDCompositionDesktopDevicePartner);
+    check_interface(dcomp_device, &IID_IDCompositionDeviceUnknown, TRUE);
+    check_inherited_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, &IID_IDCompositionDeviceUnknown);
 
     refcount = IDCompositionDevice_Release(dcomp_device);
     ok(!refcount, "Device has %lu references left.\n", refcount);
@@ -880,6 +884,8 @@ static void test_DCompositionCreateDevice3(void)
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevice, TRUE);
     check_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, TRUE);
     check_inherited_interface(dcomp_device, &IID_IDCompositionDesktopDevice, &IID_IDCompositionDesktopDevicePartner);
+    check_interface(dcomp_device, &IID_IDCompositionDeviceUnknown, TRUE);
+    check_inherited_interface(dcomp_device, &IID_IDCompositionDesktopDevicePartner, &IID_IDCompositionDeviceUnknown);
 
     refcount = IDCompositionDevice_Release(dcomp_device);
     ok(!refcount, "Device has %lu references left.\n", refcount);
-- 
2.51.0

