From a317e9d36c4e504d881753bd584a2e69670f0138 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 36/65] dcomp/tests: Add tests for desktop_device_Unknown17().

---
 dlls/dcomp/tests/dcomp.c | 45 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 45 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 8f69760bcc3..1f386dfcd7d 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -313,6 +313,7 @@ static void test_DCompositionCreateDevice2(void)
     /* Parameter checks */
     hr = pDCompositionCreateDevice2(NULL, &IID_IDCompositionDevice, (void **)&dcomp_device);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
     refcount = IDCompositionDevice_Release(dcomp_device);
     ok(!refcount, "Device has %lu references left.\n", refcount);
 
@@ -993,6 +994,49 @@ static void test_IDCompositionDesktopDevicePartner(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_IDCompositionDeviceUnknown(void)
+{
+    IDCompositionDeviceUnknown *device_unknown;
+    void *stack_pointer, *old_stack_pointer;
+    IDCompositionDevice *dcomp_device;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    HRESULT hr;
+    void *object;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hr = pDCompositionCreateDevice3((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDeviceUnknown, (void **)&device_unknown);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    old_stack_pointer = get_stack_pointer();
+
+    hr = IDCompositionDeviceUnknown_Unknown17(device_unknown, NULL);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer.\n");
+
+    object = (void *)0xdeadbeef;
+    hr = IDCompositionDeviceUnknown_Unknown17(device_unknown, &object);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    ok(object == NULL, "Got object %p.\n", object);
+
+    IDCompositionDeviceUnknown_Release(device_unknown);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 static void test_IDCompositionVisualUnknown(void)
 {
     IDCompositionDesktopDevice *desktop_device;
@@ -1083,6 +1127,7 @@ START_TEST(dcomp)
     test_DCompositionCreateSharedVisualHandle();
     test_DCompositionWaitForCompositorClock();
     test_IDCompositionDesktopDevicePartner();
+    test_IDCompositionDeviceUnknown();
     test_IDCompositionVisualUnknown();
     test_device_Commit();
     test_device_CreateTargetForHwnd();
-- 
2.51.0

