From 4b9804fe2541b29ca48b256ee0f8f3d119f36462 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 38/65] dcomp/tests: Add tests for desktop_device_Unknown16().

---
 dlls/dcomp/tests/dcomp.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 1f386dfcd7d..a1314e32e99 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1016,7 +1016,21 @@ static void test_IDCompositionDeviceUnknown(void)
 
     hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDeviceUnknown, (void **)&device_unknown);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    /* Test IDCompositionDeviceUnknown_Unknown16() */
+    old_stack_pointer = get_stack_pointer();
 
+    hr = IDCompositionDeviceUnknown_Unknown16(device_unknown, NULL);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer.\n");
+
+    object = (void *)0xdeadbeef;
+    hr = IDCompositionDeviceUnknown_Unknown16(device_unknown, &object);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    ok(object == NULL, "Got object %p.\n", object);
+
+    /* Test IDCompositionDeviceUnknown_Unknown17() */
     old_stack_pointer = get_stack_pointer();
 
     hr = IDCompositionDeviceUnknown_Unknown17(device_unknown, NULL);
-- 
2.51.0

