From 938627886e0321b736759bc8af1517f69987a7b6 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 41/65] dcomp/tests: Add tests for
 IDCompositionVisualUnknown_Unknown14().

---
 dlls/dcomp/tests/dcomp.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index fedac470817..499bd21ed3b 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1096,6 +1096,14 @@ static void test_IDCompositionVisualUnknown(void)
 
     old_stack_pointer = get_stack_pointer();
 
+    hr = IDCompositionVisualUnknown_Unknown14(visual_unknown, 0);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer.\n");
+
+    old_stack_pointer = get_stack_pointer();
+
     hr = IDCompositionVisualUnknown_Unknown19(visual_unknown, 0);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
-- 
2.51.0

