From 1b0608437d1b75a61b62468e14dc9825a05292da Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 42/65] dcomp/tests: Add visual_SetOffsetX() tests.

---
 dlls/dcomp/tests/dcomp.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 499bd21ed3b..78ca4264070 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1286,6 +1286,42 @@ static void test_visual_SetBackFaceVisibility(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_visual_SetOffsetX(void)
+{
+    IDCompositionDevice *dcomp_device;
+    IDCompositionVisual *visual;
+    IDXGISwapChain *swapchain;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    HRESULT hr;
+    HWND hwnd;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hwnd = create_window();
+    swapchain = create_swapchain(dxgi_device, hwnd);
+    hr = pDCompositionCreateDevice(dxgi_device, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionVisual_SetOffsetX(visual, 0);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    IDCompositionVisual_Release(visual);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    IDXGISwapChain_Release(swapchain);
+    DestroyWindow(hwnd);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 static void test_visual_AddVisual(void)
 {
     IDCompositionVisual *parent_visual, *parent_visual2, *child_visual, *ref_visual;
@@ -1435,6 +1471,7 @@ START_TEST(dcomp)
     test_visual_SetBitmapInterpolationMode();
     test_visual_SetBorderMode();
     test_visual_SetBackFaceVisibility();
+    test_visual_SetOffsetX();
     test_visual_AddVisual();
 
     FreeLibrary(module);
-- 
2.51.0

