From 0b11913e7d9bbdbedca0b1c2bbb9b2ae8cb00cad Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 44/65] dcomp/tests: Add visual_SetOffsetXAnimation() tests.

---
 dlls/dcomp/tests/dcomp.c | 60 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 60 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 78ca4264070..301ce549876 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1322,6 +1322,65 @@ static void test_visual_SetOffsetX(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_visual_SetOffsetXAnimation(void)
+{
+    IDCompositionAnimation *animation;
+    IDCompositionDevice *dcomp_device;
+    IDCompositionVisual *visual;
+    IDXGISwapChain *swapchain;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    HRESULT hr;
+    HWND hwnd;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hwnd = create_window();
+    swapchain = create_swapchain(dxgi_device, hwnd);
+    hr = pDCompositionCreateDevice(dxgi_device, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_CreateAnimation(dcomp_device, &animation);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (FAILED(hr))
+        goto done;
+
+    /* NULL animation pointer */
+    hr = IDCompositionVisual_SetOffsetXAnimation(visual, NULL);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    /* An animation which didn't end */
+    hr = IDCompositionVisual_SetOffsetXAnimation(visual, animation);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    /* Normal animation */
+    hr = IDCompositionAnimation_End(animation, 1, 1);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionVisual_SetOffsetXAnimation(visual, animation);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    IDCompositionAnimation_Release(animation);
+done:
+    IDCompositionVisual_Release(visual);
+    refcount = IDCompositionDevice_Release(dcomp_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+    IDXGISwapChain_Release(swapchain);
+    DestroyWindow(hwnd);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 static void test_visual_AddVisual(void)
 {
     IDCompositionVisual *parent_visual, *parent_visual2, *child_visual, *ref_visual;
@@ -1472,6 +1531,7 @@ START_TEST(dcomp)
     test_visual_SetBorderMode();
     test_visual_SetBackFaceVisibility();
     test_visual_SetOffsetX();
+    test_visual_SetOffsetXAnimation();
     test_visual_AddVisual();
 
     FreeLibrary(module);
-- 
2.51.0

