From ad4aa22dc7aa0f8424f3b2903912a4e491a095fe Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 45/65] dcomp/tests: Remove succeeding todo_wine.

---
 dlls/dcomp/tests/dcomp.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 301ce549876..e35d011879f 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1169,7 +1169,6 @@ static void test_visual_SetBitmapInterpolationMode(void)
     for (i = 0; i < ARRAY_SIZE(tests); i++)
     {
         hr = IDCompositionVisual_SetBitmapInterpolationMode(visual, tests[i].mode);
-        todo_wine
         ok(hr == tests[i].hr, "Got unexpected hr %#lx.\n", hr);
     }
 
@@ -1221,7 +1220,6 @@ static void test_visual_SetBorderMode(void)
     for (i = 0; i < ARRAY_SIZE(tests); i++)
     {
         hr = IDCompositionVisual_SetBorderMode(visual, tests[i].mode);
-        todo_wine
         ok(hr == tests[i].hr, "Got unexpected hr %#lx.\n", hr);
     }
 
@@ -1273,7 +1271,6 @@ static void test_visual_SetBackFaceVisibility(void)
     for (i = 0; i < ARRAY_SIZE(tests); i++)
     {
         hr = IDCompositionVisual2_SetBackFaceVisibility(visual, tests[i].visibility);
-        todo_wine
         ok(hr == tests[i].hr, "Got unexpected hr %#lx.\n", hr);
     }
 
@@ -1310,7 +1307,6 @@ static void test_visual_SetOffsetX(void)
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
     hr = IDCompositionVisual_SetOffsetX(visual, 0);
-    todo_wine
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
     IDCompositionVisual_Release(visual);
-- 
2.51.0

