From f8137cc24ce45045f9f9f430c3109c89f418988b Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:35 +0800
Subject: [PATCH 46/65] dcomp/tests: Add tests for
 IDCompositionDevice2::CreateSurfaceFactory().

---
 dlls/dcomp/tests/dcomp.c | 56 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 56 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index e35d011879f..c9abf7d26d6 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1488,6 +1488,61 @@ static void test_visual_AddVisual(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_device_CreateSurfaceFactory(void)
+{
+    IDCompositionSurfaceFactory *surface_factory;
+    IDCompositionDevice2 *dcomp_device2;
+    IDCompositionDevice *dcomp_device;
+    IDXGIDevice *dxgi_device;
+    ID2D1Device *d2d_device;
+    ULONG refcount;
+    HRESULT hr;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hr = pDCompositionCreateDevice2((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDevice2, (void *)&dcomp_device2);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* NULL rendering device pointer */
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, NULL, &surface_factory);
+    todo_wine
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    /* Rendering device pointer is not an IDXGIDevice or ID2D1Device */
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)dcomp_device, &surface_factory);
+    todo_wine
+    ok(hr == E_NOINTERFACE, "Got unexpected hr %#lx.\n", hr);
+
+    /* Rendering device pointer is an IDXGIDevice */
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)dxgi_device, &surface_factory);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (hr == S_OK)
+        IDCompositionSurfaceFactory_Release(surface_factory);
+
+     /* Rendering device pointer is an ID2D1Device */
+    hr = D2D1CreateDevice(dxgi_device, NULL, &d2d_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)d2d_device, &surface_factory);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (hr == S_OK)
+        IDCompositionSurfaceFactory_Release(surface_factory);
+    ID2D1Device_Release(d2d_device);
+
+    IDCompositionDevice2_Release(dcomp_device2);
+    IDCompositionDevice_Release(dcomp_device);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -1519,6 +1574,7 @@ START_TEST(dcomp)
     test_IDCompositionDeviceUnknown();
     test_IDCompositionVisualUnknown();
     test_device_Commit();
+    test_device_CreateSurfaceFactory();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
     test_target_SetRoot();
-- 
2.51.0

