From fa726b75a4fa93deb385de7a99d23bb4af84f691 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:36 +0800
Subject: [PATCH 48/65] dcomp/tests: Add tests for
 IDCompositionSurfaceFactory::CreateSurface().

---
 dlls/dcomp/tests/dcomp.c | 91 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 91 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index f650de806c5..abaf76299c3 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1537,6 +1537,96 @@ static void test_device_CreateSurfaceFactory(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_surface_factory_CreateSurface(void)
+{
+    IDCompositionSurfaceFactory *surface_factories[2];
+    IDCompositionDevice2 *dcomp_device2;
+    IDCompositionDevice *dcomp_device;
+    IDCompositionSurface *surface;
+    IDXGIDevice *dxgi_device;
+    ID2D1Device *d2d_device;
+    ULONG refcount;
+    unsigned int i, j;
+    HRESULT hr;
+
+    static const struct
+    {
+        UINT width;
+        UINT height;
+        DXGI_FORMAT pixel_format;
+        DXGI_ALPHA_MODE alpha_mode;
+        HRESULT hr;
+    }
+    tests[] =
+    {
+        /* Invalid width */
+        {0, 1, DXGI_FORMAT_R8G8B8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, E_INVALIDARG},
+        /* Invalid height */
+        {1, 0, DXGI_FORMAT_R8G8B8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, E_INVALIDARG},
+        /* Invalid pixel format */
+        {1, 1, DXGI_FORMAT_UNKNOWN, DXGI_ALPHA_MODE_IGNORE, E_INVALIDARG},
+        /* Invalid alpha mode format */
+        {1, 1, DXGI_FORMAT_UNKNOWN, DXGI_ALPHA_MODE_STRAIGHT, E_INVALIDARG},
+        /* Valid pixel formats */
+        {1, 1, DXGI_FORMAT_B8G8R8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, S_OK},
+        {1, 1, DXGI_FORMAT_R8G8B8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, S_OK},
+        {1, 1, DXGI_FORMAT_R16G16B16A16_FLOAT, DXGI_ALPHA_MODE_IGNORE, S_OK},
+        /* Valid alpha mode */
+        {1, 1, DXGI_FORMAT_B8G8R8A8_UNORM, DXGI_ALPHA_MODE_UNSPECIFIED, S_OK},
+        {1, 1, DXGI_FORMAT_B8G8R8A8_UNORM, DXGI_ALPHA_MODE_PREMULTIPLIED, S_OK},
+        {1, 1, DXGI_FORMAT_B8G8R8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, S_OK},
+    };
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hr = pDCompositionCreateDevice2((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDevice2, (void *)&dcomp_device2);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Rendering device pointer is an IDXGIDevice */
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)dxgi_device, &surface_factories[0]);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+     /* Rendering device pointer is an ID2D1Device */
+    hr = D2D1CreateDevice(dxgi_device, NULL, &d2d_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)d2d_device, &surface_factories[1]);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    ID2D1Device_Release(d2d_device);
+
+    for (i = 0; i < ARRAY_SIZE(surface_factories); i++)
+    {
+        winetest_push_context("%d", i);
+
+        for (j = 0; j < ARRAY_SIZE(tests); j++)
+        {
+            winetest_push_context("%d", i);
+
+            hr = IDCompositionSurfaceFactory_CreateSurface(surface_factories[i], tests[j].width,
+                    tests[j].height, tests[j].pixel_format, tests[j].alpha_mode, &surface);
+            todo_wine
+            ok(hr == tests[j].hr, "Got unexpected hr %#lx.\n", hr);
+            if (hr == S_OK)
+                IDCompositionSurface_Release(surface);
+            winetest_pop_context();
+        }
+
+        IDCompositionSurfaceFactory_Release(surface_factories[i]);
+        winetest_pop_context();
+    }
+
+    IDCompositionDevice2_Release(dcomp_device2);
+    IDCompositionDevice_Release(dcomp_device);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 START_TEST(dcomp)
 {
     HMODULE module;
@@ -1571,6 +1661,7 @@ START_TEST(dcomp)
     test_device_CreateSurfaceFactory();
     test_device_CreateTargetForHwnd();
     test_device_CreateVisual();
+    test_surface_factory_CreateSurface();
     test_target_SetRoot();
     test_visual_SetContent();
     test_visual_SetBitmapInterpolationMode();
-- 
2.51.0

