From bf9f2dbe9015e9db93e8116dc7296f6fc15f139d Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:36 +0800
Subject: [PATCH 53/65] dcomp/tests: Add tests for committing visual content of
 IDCompositionSurface.

---
 dlls/dcomp/tests/dcomp.c | 99 +++++++++++++++++++++++++++++++++++++++-
 1 file changed, 98 insertions(+), 1 deletion(-)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index d33e5977015..04f51a005d4 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -774,11 +774,15 @@ static void _expect_rendered_color(int line, HWND hwnd, COLORREF expected_color)
 
 static void test_device_Commit(void)
 {
+    IDCompositionSurfaceFactory *surface_factory;
     DXGI_SWAP_CHAIN_DESC1 swapchain_desc;
+    IDCompositionDevice2 *dcomp_device2;
     IDCompositionDevice *dcomp_device;
+    IDCompositionSurface *surface;
     ID3D10Device1 *d3d10_device;
     IDCompositionTarget *target;
     IDCompositionVisual *visual;
+    IDXGISurface *dxgi_surface;
     IDXGISwapChain1 *swapchain;
     IDXGIDevice *dxgi_device;
     IDXGIFactory2 *factory2;
@@ -786,6 +790,7 @@ static void test_device_Commit(void)
     D2D1_COLOR_F color;
     ULONG refcount;
     HBRUSH brush;
+    POINT offset;
     HRESULT hr;
     HWND hwnd;
     RECT rect;
@@ -827,12 +832,17 @@ static void test_device_Commit(void)
             &swapchain_desc, NULL, &swapchain);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
-    hr = pDCompositionCreateDevice(NULL, &IID_IDCompositionDevice, (void **)&dcomp_device);
+    hr = pDCompositionCreateDevice2((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDevice2,
+            (void *)&dcomp_device2);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
     hr = IDCompositionDevice_CreateVisual(dcomp_device, &visual);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
+    /* Test Commit() when visual content is a IDXGISwapChain1 */
     hr = IDCompositionVisual_SetContent(visual, (IUnknown *)swapchain);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
@@ -872,9 +882,96 @@ static void test_device_Commit(void)
     render_color_swapchain(swapchain, &color);
     expect_rendered_color(hwnd, RGB(0, 0, 0xff));
 
+    /* Test Commit() when visual content is a IDCompositionSurface */
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)dxgi_device, &surface_factory);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionSurfaceFactory_CreateSurface(surface_factory, 640, 480, DXGI_FORMAT_B8G8R8A8_UNORM,
+            DXGI_ALPHA_MODE_IGNORE, &surface);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionVisual_SetContent(visual, (IUnknown *)surface);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Render surface green */
+    hr = IDCompositionSurface_BeginDraw(surface, NULL, &IID_IDXGISurface, (void **)&dxgi_surface, &offset);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (hr == S_OK)
+    {
+        set_color(&color, 0.0f, 1.0f, 0.0f, 1.0f);
+        render_color_surface(dxgi_surface, &color);
+        IDXGISurface_Release(dxgi_surface);
+        hr = IDCompositionSurface_EndDraw(surface);
+        todo_wine
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    }
+
+    hr = IDCompositionDevice_Commit(dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    todo_wine
+    expect_rendered_color(hwnd, RGB(0, 0xff, 0x00));
+
+    /* Render surface yellow */
+    hr = IDCompositionSurface_BeginDraw(surface, NULL, &IID_IDXGISurface, (void **)&dxgi_surface, &offset);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (hr == S_OK)
+    {
+        set_color(&color, 1.0f, 1.0f, 0.0f, 1.0f);
+        render_color_surface(dxgi_surface, &color);
+        IDXGISurface_Release(dxgi_surface);
+        hr = IDCompositionSurface_EndDraw(surface);
+        todo_wine
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    }
+
+    /* Remains green before calling Commit() */
+    todo_wine
+    expect_rendered_color(hwnd, RGB(0, 0xff, 0x00));
+
+    hr = IDCompositionDevice_Commit(dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    todo_wine
+    expect_rendered_color(hwnd, RGB(0xff, 0xff, 0x00));
+
+    /* Test Commit() after two draws. Render surface white and then magenta. */
+    hr = IDCompositionSurface_BeginDraw(surface, NULL, &IID_IDXGISurface, (void **)&dxgi_surface, &offset);
+    todo_wine
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    if (hr == S_OK)
+    {
+        set_color(&color, 1.0f, 1.0f, 1.0f, 1.0f);
+        render_color_surface(dxgi_surface, &color);
+        IDXGISurface_Release(dxgi_surface);
+        hr = IDCompositionSurface_EndDraw(surface);
+        todo_wine
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+        hr = IDCompositionSurface_BeginDraw(surface, NULL, &IID_IDXGISurface, (void **)&dxgi_surface, &offset);
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        set_color(&color, 1.0f, 0.0f, 1.0f, 1.0f);
+        render_color_surface(dxgi_surface, &color);
+        IDXGISurface_Release(dxgi_surface);
+        hr = IDCompositionSurface_EndDraw(surface);
+        todo_wine
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    }
+
+    hr = IDCompositionDevice_Commit(dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    todo_wine
+    expect_rendered_color(hwnd, RGB(0xff, 0x00, 0xff));
+
+    hr = IDCompositionVisual_SetContent(visual, NULL);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    IDCompositionSurface_Release(surface);
+    IDCompositionSurfaceFactory_Release(surface_factory);
+
     DestroyWindow(hwnd);
     IDCompositionTarget_Release(target);
     IDCompositionVisual_Release(visual);
+    IDCompositionDevice2_Release(dcomp_device2);
     refcount = IDCompositionDevice_Release(dcomp_device);
     ok(!refcount, "Device has %lu references left.\n", refcount);
     IDXGISwapChain1_Release(swapchain);
-- 
2.51.0

