From ba96a12627076fe0684410b42b35bf7fd8af2fff Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 13 Mar 2026 10:13:36 +0800
Subject: [PATCH 64/65] dcomp/tests: Add tests for
 IDCompositionSurfaceUnknown::Unknown3().

---
 dlls/dcomp/tests/dcomp.c | 96 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 96 insertions(+)

diff --git a/dlls/dcomp/tests/dcomp.c b/dlls/dcomp/tests/dcomp.c
index 07fd1746457..c0cfb2b0c9e 100644
--- a/dlls/dcomp/tests/dcomp.c
+++ b/dlls/dcomp/tests/dcomp.c
@@ -1683,6 +1683,101 @@ static void test_device_CreateSurfaceFactory(void)
     ok(!refcount, "Device has %lu references left.\n", refcount);
 }
 
+static void test_surface_interface(void)
+{
+    static const unsigned int width = 640, height = 480;
+    unsigned char *stack_pointer, *old_stack_pointer;
+    IDCompositionSurfaceFactory *surface_factory;
+    IDCompositionSurfaceUnknown *surface_unknown;
+    IDCompositionDevice2 *dcomp_device2;
+    IDCompositionDevice *dcomp_device;
+    IDCompositionSurface *surface;
+    IDXGISurface *dxgi_surface;
+    IDXGIDevice *dxgi_device;
+    ULONG refcount;
+    POINT offset;
+    HRESULT hr;
+    RECT rect;
+
+    if (!(dxgi_device = create_device(D3D10_CREATE_DEVICE_BGRA_SUPPORT)))
+    {
+        skip("Failed to create device.\n");
+        return;
+    }
+
+    hr = pDCompositionCreateDevice2((IUnknown *)dxgi_device, &IID_IDCompositionDevice,
+            (void **)&dcomp_device);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice_QueryInterface(dcomp_device, &IID_IDCompositionDevice2, (void *)&dcomp_device2);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IDCompositionDevice2_CreateSurfaceFactory(dcomp_device2, (IUnknown *)dxgi_device, &surface_factory);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionSurfaceFactory_CreateSurface(surface_factory, width, height,
+            DXGI_FORMAT_B8G8R8A8_UNORM, DXGI_ALPHA_MODE_IGNORE, &surface);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    check_interface(surface, &IID_IDCompositionSurface, TRUE);
+    check_interface(surface, &IID_IDCompositionSurfaceUnknown, TRUE);
+    check_inherited_interface(surface, &IID_IDCompositionSurfaceUnknown, &IID_IDCompositionSurface);
+
+    /* Test IDCompositionSurfaceUnknown */
+    hr = IDCompositionSurface_QueryInterface(surface, &IID_IDCompositionSurfaceUnknown, (void **)&surface_unknown);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Test IDCompositionSurfaceUnknown::Unknown3 */
+    old_stack_pointer = get_stack_pointer();
+
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 0, 0);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    stack_pointer = get_stack_pointer();
+    ok(stack_pointer == old_stack_pointer, "Got unexpected stack pointer, offset %d.\n",
+            (int)(stack_pointer - old_stack_pointer));
+
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 0, 1);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 1, 0);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 1, 1);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* IDCompositionSurfaceUnknown::Unknown3 after BeginDraw() */
+    SetRect(&rect, 0, 0, 1, 1);
+    hr = IDCompositionSurface_BeginDraw(surface, &rect, &IID_IDXGISurface, (void **)&dxgi_surface,
+            &offset);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 2, 2);
+    ok(hr == DCOMPOSITION_ERROR_SURFACE_BEING_RENDERED, "Got unexpected hr %#lx.\n", hr);
+
+    IDXGISurface_Release(dxgi_surface);
+    hr = IDCompositionSurface_EndDraw(surface);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    /* Check BeginDraw after IDCompositionSurfaceUnknown::Unknown3(). This shows that
+     * IDCompositionSurfaceUnknown::Unknown3() resizes the surface because it should succeed if the
+     * size is kept at 640x480 according to other tests */
+    hr = IDCompositionSurfaceUnknown_Unknown3(surface_unknown, 100, 100);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    SetRect(&rect, 0, 0, 100, 101);
+    hr = IDCompositionSurface_BeginDraw(surface, &rect, &IID_IDXGISurface, (void **)&dxgi_surface,
+            &offset);
+    ok(hr == E_INVALIDARG, "Got unexpected hr %#lx.\n", hr);
+
+    IDCompositionSurfaceUnknown_Release(surface_unknown);
+
+    IDCompositionSurface_Release(surface);
+    IDCompositionSurfaceFactory_Release(surface_factory);
+    IDCompositionDevice2_Release(dcomp_device2);
+    IDCompositionDevice_Release(dcomp_device);
+    refcount = IDXGIDevice_Release(dxgi_device);
+    ok(!refcount, "Device has %lu references left.\n", refcount);
+}
+
 static void test_surface_factory_CreateSurface(void)
 {
     IDCompositionSurfaceFactory *surface_factories[2];
@@ -2145,6 +2240,7 @@ START_TEST(dcomp)
     test_device_CreateVisual();
     test_surface_factory_CreateSurface();
     test_surface_begin_end_Draw();
+    test_surface_interface();
     test_target_SetRoot();
     test_visual_SetContent();
     test_visual_SetBitmapInterpolationMode();
-- 
2.51.0

