From b5ae43cfd5cd0bce708d3c60eb7543f3ffab6d78 Mon Sep 17 00:00:00 2001
From: Stian Low <wineryyyyy@gmail.com>
Date: Mon, 8 Jun 2026 11:41:27 +1000
Subject: [PATCH] dcomp: Allow IDCompositionDevice3 interface

---
 dlls/dcomp/device.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/dcomp/device.c b/dlls/dcomp/device.c
index 8984ed275c5..6e047ff286f 100644
--- a/dlls/dcomp/device.c
+++ b/dlls/dcomp/device.c
@@ -75,6 +75,7 @@ static HRESULT STDMETHODCALLTYPE device_QueryInterface(IDCompositionDevice *ifac
     }
     else if ((device->version >= 2
               && (IsEqualGUID(iid, &IID_IDCompositionDevice2)
+                  || IsEqualGUID(iid, &IID_IDCompositionDevice3)
                   || IsEqualGUID(iid, &IID_IDCompositionDesktopDevice)))
               || IsEqualGUID(iid, &IID_IDCompositionDesktopDevicePartner)
               || IsEqualGUID(iid, &IID_IDCompositionDeviceUnknown))
-- 
2.53.0

