From 99b908c02eccbaa226e0180d430efb6be1df5475 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sun, 30 Jan 2022 10:52:20 +1100
Subject: [PATCH] xactengine3_7: Fix for older compilers

---
 dlls/xactengine3_7/xact_dll.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/dlls/xactengine3_7/xact_dll.c b/dlls/xactengine3_7/xact_dll.c
index 62084f9234f..a0a7d9d6d32 100644
--- a/dlls/xactengine3_7/xact_dll.c
+++ b/dlls/xactengine3_7/xact_dll.c
@@ -859,16 +859,16 @@ static void FACTCALL fact_notification_cb(const FACTNotification *notification)
 
     switch (notification->type)
     {
-        case XACTNOTIFICATIONTYPE_SOUNDBANKDESTROYED:
+        case (int)XACTNOTIFICATIONTYPE_SOUNDBANKDESTROYED:
             note.soundBank.pSoundBank = FACTSoundBank_GetPrivateContext(notification->wave.pSoundBank);
             break;
 #if XACT3_VER >= 0x0205
-        case XACTNOTIFICATIONTYPE_WAVEDESTROYED:
-        case XACTNOTIFICATIONTYPE_WAVELOOPED:
-        case XACTNOTIFICATIONTYPE_WAVEPLAY:
-        case XACTNOTIFICATIONTYPE_WAVEPREPARED:
+        case (int)XACTNOTIFICATIONTYPE_WAVEDESTROYED:
+        case (int)XACTNOTIFICATIONTYPE_WAVELOOPED:
+        case (int)XACTNOTIFICATIONTYPE_WAVEPLAY:
+        case (int)XACTNOTIFICATIONTYPE_WAVEPREPARED:
 #endif
-        case XACTNOTIFICATIONTYPE_WAVESTOP:
+        case (int)XACTNOTIFICATIONTYPE_WAVESTOP:
             note.wave.cueIndex = notification->wave.cueIndex;
             note.wave.pCue = FACTCue_GetPrivateContext(notification->wave.pCue);
             note.wave.pSoundBank = FACTSoundBank_GetPrivateContext(notification->wave.pSoundBank);
@@ -878,12 +878,12 @@ static void FACTCALL fact_notification_cb(const FACTNotification *notification)
             note.wave.pWaveBank = FACTWaveBank_GetPrivateContext(notification->wave.pWaveBank);
             break;
 
-        case XACTNOTIFICATIONTYPE_CUEPLAY:
-        case XACTNOTIFICATIONTYPE_CUEPREPARED:
-        case XACTNOTIFICATIONTYPE_CUESTOP:
+        case (int)XACTNOTIFICATIONTYPE_CUEPLAY:
+        case (int)XACTNOTIFICATIONTYPE_CUEPREPARED:
+        case (int)XACTNOTIFICATIONTYPE_CUESTOP:
             note.cue.pCue = FACTCue_GetPrivateContext(notification->cue.pCue);
             /* Fall through */
-        case XACTNOTIFICATIONTYPE_CUEDESTROYED:
+        case (int)XACTNOTIFICATIONTYPE_CUEDESTROYED:
             note.cue.cueIndex = notification->cue.cueIndex;
             note.cue.pSoundBank = FACTSoundBank_GetPrivateContext(notification->cue.pSoundBank);
             break;
-- 
2.34.1

