From 0c8d2be596270def5721fe0ae2d11732e63f6b3b Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 4 Jan 2024 12:27:13 +1100
Subject: [PATCH] include: Add more D3D_FEATURE_LEVEL_ defines

---
 include/d3dcommon.idl | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/d3dcommon.idl b/include/d3dcommon.idl
index 7a53a84079d..5ffb4cd5b5b 100644
--- a/include/d3dcommon.idl
+++ b/include/d3dcommon.idl
@@ -100,6 +100,7 @@ typedef enum D3D_DRIVER_TYPE
 
 typedef enum D3D_FEATURE_LEVEL
 {
+    D3D_FEATURE_LEVEL_1_0_CORE = 0x1000,
     D3D_FEATURE_LEVEL_9_1 = 0x9100,
     D3D_FEATURE_LEVEL_9_2 = 0x9200,
     D3D_FEATURE_LEVEL_9_3 = 0x9300,
@@ -109,6 +110,7 @@ typedef enum D3D_FEATURE_LEVEL
     D3D_FEATURE_LEVEL_11_1 = 0xb100,
     D3D_FEATURE_LEVEL_12_0 = 0xc000,
     D3D_FEATURE_LEVEL_12_1 = 0xc100,
+    D3D_FEATURE_LEVEL_12_2 = 0xc200,
 } D3D_FEATURE_LEVEL;
 
 cpp_quote("#define D3D_FL9_1_REQ_TEXTURE1D_U_DIMENSION          2048")
-- 
2.43.0

