From 61f65bb6024f63c6ab278ae6c9844d2d38339be8 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 16 Feb 2024 11:01:45 +1100
Subject: [PATCH] combase: Avoid implicit cast of interface pointer.

---
 dlls/combase/string.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/combase/string.c b/dlls/combase/string.c
index 1345870cf2f..e60ea251011 100644
--- a/dlls/combase/string.c
+++ b/dlls/combase/string.c
@@ -55,12 +55,12 @@ static inline struct hstring_private *impl_from_HSTRING(HSTRING string)
 
 static inline struct hstring_private *impl_from_HSTRING_HEADER(HSTRING_HEADER *header)
 {
-    return CONTAINING_RECORD(header, struct hstring_private, header);
+    return CONTAINING_RECORD((struct hstring_header*)header, struct hstring_private, header);
 }
 
 static inline struct hstring_private *impl_from_HSTRING_BUFFER(HSTRING_BUFFER buffer)
 {
-    return CONTAINING_RECORD(buffer, struct hstring_private, buffer);
+    return CONTAINING_RECORD( (void*)buffer, struct hstring_private, buffer);
 }
 
 static BOOL alloc_string(UINT32 len, HSTRING *out)
-- 
2.43.0

