From 31ad180c1c96e6e9a2e666ebbe90c3d8b78cefa4 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 16 Feb 2024 11:01:45 +1100
Subject: [PATCH] combase: Avoid implicit cast of interface pointer.

---
 dlls/combase/string.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/combase/string.c b/dlls/combase/string.c
index 1345870cf2f..dda6ab78141 100644
--- a/dlls/combase/string.c
+++ b/dlls/combase/string.c
@@ -55,7 +55,7 @@ static inline struct hstring_private *impl_from_HSTRING(HSTRING string)
 
 static inline struct hstring_private *impl_from_HSTRING_HEADER(HSTRING_HEADER *header)
 {
-    return CONTAINING_RECORD(header, struct hstring_private, header);
+    return CONTAINING_RECORD((struct hstring_header*)header, struct hstring_private, header);
 }
 
 static inline struct hstring_private *impl_from_HSTRING_BUFFER(HSTRING_BUFFER buffer)
-- 
2.43.0

