From bab45b9b71bac540d36487f0b8493885b6564a81 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 22 Mar 2016 21:54:48 +0100
Subject: [PATCH] d3d11: Avoid implicit cast of interface pointer.

---
 dlls/d3d11/view.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/d3d11/view.c b/dlls/d3d11/view.c
index 228bed0ed66..23df10c0ab7 100644
--- a/dlls/d3d11/view.c
+++ b/dlls/d3d11/view.c
@@ -2282,7 +2282,7 @@ struct d3d_shader_resource_view *unsafe_impl_from_ID3D10ShaderResourceView(ID3D1
     if (!iface)
         return NULL;
     assert(iface->lpVtbl == (ID3D10ShaderResourceViewVtbl *)&d3d10_shader_resource_view_vtbl);
-    return CONTAINING_RECORD(iface, struct d3d_shader_resource_view, ID3D10ShaderResourceView1_iface);
+    return CONTAINING_RECORD((ID3D10ShaderResourceView1 *)iface, struct d3d_shader_resource_view, ID3D10ShaderResourceView1_iface);
 }
 
 /* ID3D11UnorderedAccessView methods */
-- 
2.43.0

