From 8f905b6f5651f2235d172ddb7ce204c7ab580ab2 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 8 Jun 2021 09:46:25 +1000
Subject: [PATCH] dxgi: Avoid implicit cast of interface pointer.

---
 dlls/dxgi/output.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dxgi/output.c b/dlls/dxgi/output.c
index e828ed0d9d8..2c212a24821 100644
--- a/dlls/dxgi/output.c
+++ b/dlls/dxgi/output.c
@@ -708,7 +708,7 @@ struct dxgi_output *unsafe_impl_from_IDXGIOutput(IDXGIOutput *iface)
     if (!iface)
         return NULL;
     assert(iface->lpVtbl == (IDXGIOutputVtbl *)&dxgi_output_vtbl);
-    return CONTAINING_RECORD(iface, struct dxgi_output, IDXGIOutput6_iface);
+    return CONTAINING_RECORD((IDXGIOutput6*)iface, struct dxgi_output, IDXGIOutput6_iface);
 }
 
 static void dxgi_output_init(struct dxgi_output *output, unsigned int output_idx,
-- 
2.43.0

