From 947fb9823f81a34e14e70a16b4688a89918d4791 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 8 Jun 2021 11:38:46 +1000
Subject: [PATCH] wbemdisp: Avoid implicit cast of interface pointer.

---
 dlls/wbemdisp/locator.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/wbemdisp/locator.c b/dlls/wbemdisp/locator.c
index c3e0c86f4ae..70f94a0d02b 100644
--- a/dlls/wbemdisp/locator.c
+++ b/dlls/wbemdisp/locator.c
@@ -1657,7 +1657,7 @@ static struct object *unsafe_object_impl_from_IDispatch(IDispatch *iface)
         FIXME( "External implementations are not supported.\n" );
         return NULL;
     }
-    return CONTAINING_RECORD(iface, struct object, ISWbemObject_iface);
+    return CONTAINING_RECORD((ISWbemObject *)iface, struct object, ISWbemObject_iface);
 }
 
 static HRESULT SWbemObject_create( struct services *services, IWbemClassObject *wbem_object,
@@ -3492,7 +3492,7 @@ static struct namedvalueset *unsafe_valueset_impl_from_IDispatch(IDispatch *ifac
         FIXME( "External implementations are not supported.\n" );
         return NULL;
     }
-    return CONTAINING_RECORD(iface, struct namedvalueset, ISWbemNamedValueSet_iface);
+    return CONTAINING_RECORD((ISWbemNamedValueSet*)iface, struct namedvalueset, ISWbemNamedValueSet_iface);
 }
 
 HRESULT SWbemNamedValueSet_create( void **obj )
-- 
2.43.0

