From 272753daf4f20b7fe3c646eae4fd78f1b254c9a6 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 8 Jun 2021 09:57:01 +1000
Subject: [PATCH] msctf: Avoid implicit cast of interface pointer.

---
 dlls/msctf/range.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/msctf/range.c b/dlls/msctf/range.c
index 9690a0a15c4..e15123eeb89 100644
--- a/dlls/msctf/range.c
+++ b/dlls/msctf/range.c
@@ -54,7 +54,7 @@ static inline Range *impl_from_ITfRangeACP(ITfRangeACP *iface)
 
 static Range *unsafe_impl_from_ITfRange(ITfRange *iface)
 {
-    return CONTAINING_RECORD(iface, Range, ITfRangeACP_iface);
+    return CONTAINING_RECORD((ITfRangeACP*)iface, Range, ITfRangeACP_iface);
 }
 
 static void Range_Destructor(Range *This)
-- 
2.30.2

