From 4801b547b70654507cb8a25dcfd1efc862dcf467 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 8 Jun 2021 11:38:46 +1000
Subject: [PATCH] wbemdisp: Avoid implicit cast of interface pointer.

---
 dlls/wbemdisp/locator.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/wbemdisp/locator.c b/dlls/wbemdisp/locator.c
index e0a1872d38f..5bb2a187f2c 100644
--- a/dlls/wbemdisp/locator.c
+++ b/dlls/wbemdisp/locator.c
@@ -1751,7 +1751,7 @@ static struct object *unsafe_object_impl_from_IDispatch(IDispatch *iface)
         FIXME( "External implementations are not supported.\n" );
         return NULL;
     }
-    return CONTAINING_RECORD(iface, struct object, ISWbemObject_iface);
+    return CONTAINING_RECORD((ISWbemObject *)iface, struct object, ISWbemObject_iface);
 }
 
 static HRESULT SWbemObject_create( struct services *services, IWbemClassObject *wbem_object,
@@ -3587,7 +3587,7 @@ static struct namedvalueset *unsafe_valueset_impl_from_IDispatch(IDispatch *ifac
         FIXME( "External implementations are not supported.\n" );
         return NULL;
     }
-    return CONTAINING_RECORD(iface, struct namedvalueset, ISWbemNamedValueSet_iface);
+    return CONTAINING_RECORD((ISWbemNamedValueSet*)iface, struct namedvalueset, ISWbemNamedValueSet_iface);
 }
 
 HRESULT SWbemNamedValueSet_create( void **obj )
-- 
2.30.2

