From 0f24bf7b15e0e22fda2106e8573b5422b80eaa04 Mon Sep 17 00:00:00 2001
From: Arkadiusz Hiler <ahiler@codeweavers.com>
Date: Wed, 10 Feb 2021 13:23:33 +0100
Subject: [PATCH] hidclass.sys: Assign rawinput handles through device
 properties.

---
 dlls/hidclass.sys/pnp.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/dlls/hidclass.sys/pnp.c b/dlls/hidclass.sys/pnp.c
index 1d82215bbfe..7eaa28a411d 100644
--- a/dlls/hidclass.sys/pnp.c
+++ b/dlls/hidclass.sys/pnp.c
@@ -25,6 +25,8 @@
 #include "ntddmou.h"
 #include "ddk/hidtypes.h"
 #include "ddk/wdm.h"
+#include "initguid.h"
+#include "devpkey.h"
 #include "regstr.h"
 #include "winuser.h"
 #include "wine/debug.h"
@@ -87,6 +89,8 @@ static NTSTATUS get_device_id(DEVICE_OBJECT *device, BUS_QUERY_ID_TYPE type, WCH
     return irp_status.u.Status;
 }
 
+DEFINE_DEVPROPKEY(DEVPROPKEY_HID_HANDLE, 0xbc62e415, 0xf4fe, 0x405c, 0x8e, 0xda, 0x63, 0x6f, 0xb5, 0x9f, 0x08, 0x98, 2);
+
 /* user32 reserves 1 & 2 for winemouse and winekeyboard,
  * keep this in sync with user_private.h */
 #define WINE_MOUSE_HANDLE 1
@@ -241,6 +245,15 @@ static NTSTATUS WINAPI driver_add_device(DRIVER_OBJECT *driver, DEVICE_OBJECT *b
     else
         pdo_ext->u.pdo.rawinput_handle = alloc_rawinput_handle();
 
+    status = IoSetDevicePropertyData(child_pdo, &DEVPROPKEY_HID_HANDLE, LOCALE_NEUTRAL,
+                                     PLUGPLAY_PROPERTY_PERSISTENT, DEVPROP_TYPE_UINT32,
+                                     sizeof(pdo_ext->u.pdo.rawinput_handle), &pdo_ext->u.pdo.rawinput_handle);
+    if (status != STATUS_SUCCESS)
+    {
+        FIXME("failed to set device property %x\n", status);
+        return status;
+    }
+
     pdo_ext->u.pdo.poll_interval = DEFAULT_POLL_INTERVAL;
 
     pdo_ext->u.pdo.ring_buffer = RingBuffer_Create(
-- 
2.30.2

