From 10cd3e92839fcc77d77fd1acb1fdbfe669e8ee85 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:44:43 +1000
Subject: [PATCH] odbc32: Load function pointer for SQLBindParam

---
 dlls/odbc32/proxyodbc.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index d31a08af14f..1add621ca31 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -50,6 +50,7 @@ struct win32_funcs
     SQLRETURN (WINAPI *SQLAllocHandleStd)(SQLSMALLINT,SQLHANDLE,SQLHANDLE*);
     SQLRETURN (WINAPI *SQLAllocStmt)(SQLHDBC,SQLHSTMT*);
     SQLRETURN (WINAPI *SQLBindCol)(SQLHSTMT,SQLUSMALLINT,SQLSMALLINT,SQLPOINTER,SQLLEN,SQLLEN*);
+    SQLRETURN (WINAPI *SQLBindParam)(SQLHSTMT,SQLUSMALLINT,SQLSMALLINT,SQLSMALLINT,SQLULEN,SQLSMALLINT,SQLPOINTER,SQLLEN*);
     SQLRETURN (WINAPI *SQLBindParameter)(SQLHSTMT,SQLUSMALLINT,SQLSMALLINT,SQLSMALLINT,SQLSMALLINT,SQLULEN,
                                          SQLSMALLINT,SQLPOINTER,SQLLEN,SQLLEN*);
     SQLRETURN (WINAPI *SQLBrowseConnect)(SQLHDBC,SQLCHAR*,SQLSMALLINT,SQLCHAR*,SQLSMALLINT,SQLSMALLINT*);
@@ -206,6 +207,7 @@ static BOOL load_function_table( HMODULE module, struct win32_driver *driver )
     LOAD_FUNCPTR( SQLAllocHandleStd )
     LOAD_FUNCPTR( SQLAllocStmt )
     LOAD_FUNCPTR( SQLBindCol )
+    LOAD_FUNCPTR( SQLBindParam )
     LOAD_FUNCPTR( SQLBindParameter )
     LOAD_FUNCPTR( SQLBrowseConnect )
     LOAD_FUNCPTR( SQLBrowseConnectW )
-- 
2.50.1

