From 49adfd79f730714b11c4d5f983ced05b26ffa52a Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 13:42:26 +1000
Subject: [PATCH] odbc32: Support freeing SQL handles for ODBC v2 drivers.

---
 dlls/odbc32/proxyodbc.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 4799ca0a11e..678d9260fbb 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1867,6 +1867,25 @@ static SQLRETURN free_handle_win32( SQLSMALLINT type, struct handle *handle )
 {
     if (handle->win32_funcs->SQLFreeHandle)
         return handle->win32_funcs->SQLFreeHandle( type, handle->win32_handle );
+    else
+    {
+        /* ODBC v2 */
+        if (type == SQL_HANDLE_ENV)
+        {
+            if (handle->win32_funcs->SQLFreeEnv)
+                return handle->win32_funcs->SQLFreeEnv( handle->win32_handle );
+        }
+        else if (type == SQL_HANDLE_DBC)
+        {
+            if (handle->win32_funcs->SQLFreeConnect)
+                return handle->win32_funcs->SQLFreeConnect( handle->win32_handle );
+        }
+        else if (type == SQL_HANDLE_STMT)
+        {
+            if (handle->win32_funcs->SQLFreeStmt)
+                return handle->win32_funcs->SQLFreeStmt( handle->win32_handle, SQL_CLOSE );
+        }
+    }
     return SQL_ERROR;
 }
 
-- 
2.43.0

