From 8256bfe5ed31fd521b0baa9085819557653be5e3 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 18 Jul 2024 14:17:44 +1000
Subject: [PATCH] odbc32: SQLSetEnvAttr isnt supported in ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 2a62fd88d64..4f23c761f5a 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1162,7 +1162,7 @@ static int has_suffix( const WCHAR *str, const WCHAR *suffix )
 
 static SQLRETURN set_env_attr( struct environment *env, SQLINTEGER attr, SQLPOINTER value, SQLINTEGER len )
 {
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     if (env->hdr.unix_handle)
     {
@@ -1171,7 +1171,8 @@ static SQLRETURN set_env_attr( struct environment *env, SQLINTEGER attr, SQLPOIN
     }
     else if (env->hdr.win32_handle)
     {
-        ret = env->hdr.win32_funcs->SQLSetEnvAttr( env->hdr.win32_handle, attr, value, len );
+        if (env->hdr.win32_funcs->SQLSetEnvAttr)
+            ret = env->hdr.win32_funcs->SQLSetEnvAttr( env->hdr.win32_handle, attr, value, len );
     }
 
     return ret;
-- 
2.43.0

