From 0408d4baa23ab59ba26cc3869fc494c0a2de15a0 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:40:32 +1000
Subject: [PATCH] odbc32: SQLBindParameter handle fallback function

---
 dlls/odbc32/proxyodbc.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 35519b0aa81..cd1159e8ee6 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -5981,9 +5981,16 @@ static SQLRETURN bind_parameter_win32( struct statement *stmt, SQLUSMALLINT para
 {
     value_type = map_type( stmt, TRUE, value_type );
     param_type = map_type( stmt, TRUE, param_type );
+
     if (stmt->hdr.win32_funcs->SQLBindParameter)
         return stmt->hdr.win32_funcs->SQLBindParameter( stmt->hdr.win32_handle, param, io_type, value_type,
                                                         param_type, size, digits, value, buflen, len );
+    else if(stmt->hdr.win32_funcs->SQLBindParam)
+    {
+        return stmt->hdr.win32_funcs->SQLBindParam( stmt->hdr.win32_handle, param, value_type, param_type,
+                                                 size, digits, value, len);
+    }
+
     return SQL_ERROR;
 }
 
-- 
2.53.0

