From 9a04cd1f7199385e7521fd77f1118682e50192cf Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:13:33 +1000
Subject: [PATCH] odbc32: Support Driver in connection string

---
 dlls/odbc32/proxyodbc.c | 97 +++++++++++++++++++++++++++++++----------
 1 file changed, 74 insertions(+), 23 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 9a4f08440ea..c19d401168b 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -753,11 +753,20 @@ static WCHAR *strdupAW( const char *src )
     return dst;
 }
 
-static HKEY open_odbcini_key( HKEY root )
+static HKEY open_odbcini_key( HKEY root, BOOL use_dsn )
 {
     static const WCHAR sourcesW[] = L"Software\\ODBC\\ODBC.INI";
+    static const WCHAR driversW[] = L"Software\\ODBC\\ODBCINST.INI";
     HKEY key;
-    if (!RegCreateKeyExW( root, sourcesW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+
+    if (use_dsn)
+    {
+        if (!RegCreateKeyExW( root, sourcesW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+    }
+    else
+    {
+        if (!RegCreateKeyExW( root, driversW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+    }
     return NULL;
 }
 
@@ -771,12 +780,12 @@ static WCHAR *get_reg_value( HKEY key, const WCHAR *name )
     return NULL;
 }
 
-static WCHAR *get_driver_filename( const SQLWCHAR *source )
+static WCHAR *get_driver_filename( const SQLWCHAR *source, BOOL use_dsn )
 {
     HKEY key_root, key_source;
     WCHAR *ret = NULL;
 
-    if (!(key_root = open_odbcini_key( HKEY_CURRENT_USER ))) return NULL;
+    if (!(key_root = open_odbcini_key( HKEY_CURRENT_USER, use_dsn ))) return NULL;
     if (!RegOpenKeyExW( key_root, source, 0, KEY_READ, &key_source ))
     {
         ret = get_reg_value( key_source, L"Driver" );
@@ -785,7 +794,7 @@ static WCHAR *get_driver_filename( const SQLWCHAR *source )
     RegCloseKey( key_root );
     if (ret) return ret;
 
-    if (!(key_root = open_odbcini_key( HKEY_LOCAL_MACHINE ))) return NULL;
+    if (!(key_root = open_odbcini_key( HKEY_LOCAL_MACHINE, use_dsn ))) return NULL;
     if (!RegOpenKeyExW( key_root, source, 0, KEY_READ, &key_source ))
     {
         ret = get_reg_value( key_source, L"Driver" );
@@ -919,7 +928,7 @@ SQLRETURN WINAPI SQLConnect(SQLHDBC ConnectionHandle, SQLCHAR *ServerName, SQLSM
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!servername || !(filename = get_driver_filename( servername )))
+    if (!servername || !(filename = get_driver_filename( servername, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -2650,6 +2659,32 @@ static WCHAR *get_datasource( const WCHAR *connection_string )
     return ret;
 }
 
+static WCHAR *get_driver( const WCHAR *connection_string )
+{
+    const WCHAR *p = connection_string, *q;
+    WCHAR *ret = NULL;
+    unsigned int len;
+
+    if (!p) return NULL;
+    while (*p)
+    {
+        if (!wcsnicmp( p, L"DRIVER=", 7 ))
+        {
+            p += 7;
+            q = wcschr( p, ';' );
+            len = q ? (q - p) : wcslen( p );
+            if ((ret = malloc( (len + 1) * sizeof(WCHAR) )))
+            {
+                memcpy( ret, p, len * sizeof(WCHAR) );
+                ret[len] = 0;
+                break;
+            }
+        }
+        p++;
+    }
+    return ret;
+}
+
 /*************************************************************************
  *				SQLBrowseConnect           [ODBC32.055]
  */
@@ -2659,6 +2694,7 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength, %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_an((const char *)InConnectionString, StringLength1),
@@ -2666,13 +2702,16 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -3319,6 +3358,7 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString, %p,"
           " BufferLength, %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -3327,13 +3367,16 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -3490,7 +3533,7 @@ SQLRETURN WINAPI SQLConnectW(SQLHDBC ConnectionHandle, WCHAR *ServerName, SQLSMA
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!(filename = get_driver_filename( ServerName )))
+    if (!(filename = get_driver_filename( ServerName, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -4056,6 +4099,7 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString %p,"
           " BufferLength %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -4064,13 +4108,16 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -4346,6 +4393,7 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_wn(InConnectionString, StringLength1), StringLength1,
@@ -4353,13 +4401,16 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
-- 
2.43.0

