From a9f272e877306607e502a268cb232609f56eed85 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:19:22 +1000
Subject: [PATCH] odbc32: SQLColAttributeW support fallback function

---
 dlls/odbc32/proxyodbc.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index c19d401168b..3ccb331a6be 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -3758,6 +3758,21 @@ SQLRETURN WINAPI SQLSetCursorNameW(SQLHSTMT StatementHandle, WCHAR *CursorName,
     return ret;
 }
 
+static SQLINTEGER map_odbc3_to_2(SQLINTEGER fieldid)
+{
+    switch( fieldid )
+    {
+        case SQL_DESC_COUNT:
+            return SQL_COLUMN_COUNT;
+        case SQL_DESC_NULLABLE:
+            return SQL_COLUMN_NULLABLE;
+        case SQL_DESC_NAME:
+            return SQL_COLUMN_NAME;
+        default:
+            return fieldid;
+    }
+}
+
 /*************************************************************************
  *				SQLColAttributeW          [ODBC32.127]
  */
@@ -3792,9 +3807,18 @@ SQLRETURN WINAPI SQLColAttributeW(SQLHSTMT StatementHandle, SQLUSMALLINT ColumnN
     }
     else if (handle->win32_handle)
     {
-        ret = handle->win32_funcs->SQLColAttributeW( handle->win32_handle, ColumnNumber, FieldIdentifier,
+        if (handle->win32_funcs->SQLColAttributeW)
+            ret = handle->win32_funcs->SQLColAttributeW( handle->win32_handle, ColumnNumber, FieldIdentifier,
                                                      CharacterAttribute, BufferLength, StringLength,
                                                      NumericAttribute );
+        else if(handle->win32_funcs->SQLColAttributesW)
+        {
+            /* ODBC v2 */
+            FieldIdentifier = map_odbc3_to_2(FieldIdentifier);
+            ret = handle->win32_funcs->SQLColAttributesW( handle->win32_handle, ColumnNumber, FieldIdentifier,
+                                                     CharacterAttribute, BufferLength, StringLength,
+                                                     NumericAttribute );
+        }
     }
 
     TRACE("Returning %d\n", ret);
-- 
2.43.0

