From c6edaf28df9e495ff173910c6e0124d244784f41 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 18 Jul 2024 14:34:49 +1000
Subject: [PATCH] odbc32: Support Allocating SQL handles for ODBC v2 drivers.

---
 dlls/odbc32/proxyodbc.c | 31 +++++++++++++++++++++++++------
 1 file changed, 25 insertions(+), 6 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 6fe38e7a863..6168786084d 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -439,13 +439,32 @@ static SQLRETURN alloc_handle_unix( SQLSMALLINT type, struct handle *input, stru
 
 static SQLRETURN alloc_handle_win32( SQLSMALLINT type, struct handle *input, struct handle *output )
 {
+    SQLRETURN ret = SQL_ERROR;
     if (input->win32_funcs->SQLAllocHandle)
     {
-        SQLRETURN ret = input->win32_funcs->SQLAllocHandle( type, input->win32_handle, &output->win32_handle );
-        if (SUCCESS( ret )) output->win32_funcs = input->win32_funcs;
-        return ret;
+        ret = input->win32_funcs->SQLAllocHandle( type, input->win32_handle, &output->win32_handle );
     }
-    return SQL_ERROR;
+    else
+    {
+        /* ODBC v2 */
+        if (type == SQL_HANDLE_ENV)
+        {
+            if (input->win32_funcs->SQLAllocEnv)
+                ret = input->win32_funcs->SQLAllocEnv( &output->win32_handle );
+        }
+        else if (type == SQL_HANDLE_DBC)
+        {
+            if (input->win32_funcs->SQLAllocConnect)
+                ret = input->win32_funcs->SQLAllocConnect( input, &output->win32_handle );
+        }
+        else if (type == SQL_HANDLE_STMT)
+        {
+            if (input->win32_funcs->SQLAllocStmt)
+                ret = input->win32_funcs->SQLAllocStmt( input, &output->win32_handle );
+        }
+    }
+    if (SUCCESS( ret )) output->win32_funcs = input->win32_funcs;
+    return ret;
 }
 
 /*************************************************************************
@@ -1045,7 +1064,7 @@ static SQLRETURN create_env( struct handle *handle, BOOL is_unix )
     }
     else
     {
-        if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_ENV, NULL, &handle->win32_handle ))) return ret;
+        if ((ret = alloc_handle_win32(SQL_HANDLE_ENV, handle, handle ))) return ret;
     }
 
     return prepare_env( handle );
@@ -1101,7 +1120,7 @@ static SQLRETURN create_con( struct handle *handle )
     }
     else
     {
-        if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_DBC, parent->win32_handle, &handle->win32_handle )))
+        if ((ret = alloc_handle_win32(SQL_HANDLE_DBC, parent, handle )))
             return ret;
     }
 
-- 
2.43.0

