From fda9d9bf29a6d1a3091dc648be8702b8b2b5cd99 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:23:01 +1000
Subject: [PATCH] odbc32: SQLError/W handle NULL handles

---
 dlls/odbc32/proxyodbc.c | 32 ++++++++++++++++++++++++++------
 1 file changed, 26 insertions(+), 6 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 3ccb331a6be..31a789b26a4 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1191,8 +1191,19 @@ SQLRETURN WINAPI SQLError(SQLHENV EnvironmentHandle, SQLHDBC ConnectionHandle, S
     }
     else if ((env && env->win32_handle) || (con && con->win32_handle) || (stmt && stmt->win32_handle))
     {
-        ret = env->win32_funcs->SQLError( env->win32_handle, con->win32_handle, stmt->win32_handle, SqlState,
-                                          NativeError, MessageText, BufferLength, TextLength );
+        const struct win32_funcs *win32_funcs = NULL;
+
+        if (env) win32_funcs = env->win32_funcs;
+        else if (con) win32_funcs = con->win32_funcs;
+        else if (stmt) win32_funcs = con->win32_funcs;
+
+        if(win32_funcs)
+            ret = win32_funcs->SQLError( env ? env->win32_handle : NULL,
+                                          con ? con->win32_handle : NULL,
+                                          stmt ? stmt->win32_handle : NULL,
+                                          SqlState, NativeError, MessageText, BufferLength, TextLength );
+        else
+            ERR("No function map found\n");
     }
 
     if (SUCCESS( ret ))
@@ -3633,10 +3644,19 @@ SQLRETURN WINAPI SQLErrorW(SQLHENV EnvironmentHandle, SQLHDBC ConnectionHandle,
     }
     else if ((env && env->win32_handle) || (con && con->win32_handle) || (stmt && stmt->win32_handle))
     {
-        ret = env->win32_funcs->SQLErrorW( env ? env->win32_handle : NULL,
-                                           con ? con->win32_handle : NULL,
-                                           stmt ? stmt->win32_handle : NULL,
-                                           SqlState, NativeError, MessageText, BufferLength, TextLength );
+        const struct win32_funcs *win32_funcs = NULL;
+
+        if (env) win32_funcs = env->win32_funcs;
+        else if (con) win32_funcs = con->win32_funcs;
+        else if (stmt) win32_funcs = con->win32_funcs;
+
+        if(win32_funcs)
+            ret = win32_funcs->SQLErrorW( env ? env->win32_handle : NULL,
+                                          con ? con->win32_handle : NULL,
+                                          stmt ? stmt->win32_handle : NULL,
+                                          SqlState, NativeError, MessageText, BufferLength, TextLength );
+        else
+            ERR("No function map found\n");
     }
 
     if (SUCCESS(ret ))
-- 
2.43.0

