From 652327cf649f9d10794f18d6c3a2c3452f42f947 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:13:33 +1000
Subject: [PATCH] odbc32: Support Driver in connection string

---
 dlls/odbc32/proxyodbc.c | 100 ++++++++++++++++++++++++++++++----------
 1 file changed, 75 insertions(+), 25 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 94504a8cc26..df8fa06dc57 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -989,23 +989,28 @@ static HKEY open_odbcinst_key( void )
     return NULL;
 }
 
-static WCHAR *get_driver_filename( const SQLWCHAR *source )
+static WCHAR *get_driver_filename( const SQLWCHAR *source, BOOL is_dsn )
 {
     HKEY key_sources, key_odbcinst, key_driver;
     WCHAR *driver_name, *ret;
 
-    if (!(key_sources = open_sources_key( HKEY_CURRENT_USER ))) return NULL;
-    if (!(driver_name = get_reg_value( key_sources, source )))
+    if (is_dsn)
     {
-        RegCloseKey( key_sources );
-        if (!(key_sources = open_sources_key( HKEY_LOCAL_MACHINE ))) return NULL;
+        if (!(key_sources = open_sources_key( HKEY_CURRENT_USER ))) return NULL;
         if (!(driver_name = get_reg_value( key_sources, source )))
         {
             RegCloseKey( key_sources );
-            return NULL;
+            if (!(key_sources = open_sources_key( HKEY_LOCAL_MACHINE ))) return NULL;
+            if (!(driver_name = get_reg_value( key_sources, source )))
+            {
+                RegCloseKey( key_sources );
+                return NULL;
+            }
         }
+        RegCloseKey( key_sources );
     }
-    RegCloseKey( key_sources );
+    else
+        driver_name = wcsdup(source);
 
     if (!(key_odbcinst = open_odbcinst_key()) || RegOpenKeyExW( key_odbcinst, driver_name, 0, KEY_READ, &key_driver ))
     {
@@ -1175,7 +1180,7 @@ SQLRETURN WINAPI SQLConnect(SQLHDBC ConnectionHandle, SQLCHAR *ServerName, SQLSM
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!servername || !(filename = get_driver_filename( servername )))
+    if (!servername || !(filename = get_driver_filename( servername, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -3983,6 +3988,35 @@ static SQLRETURN browse_connect_unix_a( struct handle *handle, SQLCHAR *in_conn_
     return ODBC_CALL( SQLBrowseConnect, &params );
 }
 
+static WCHAR *get_driver( const WCHAR *connection_string )
+{
+    const WCHAR *p = connection_string, *q;
+    WCHAR *ret = NULL;
+    unsigned int len;
+
+    if (!p) return NULL;
+    while (*p)
+    {
+        if (!wcsnicmp( p, L"DRIVER=", 7 ))
+        {
+            p += 7;
+            if (*p == '{')
+                q = wcschr( ++p, '}' );
+            else
+                q = wcschr( p, ';' );
+            len = q ? (q - p) : wcslen( p );
+            if ((ret = malloc( (len + 1) * sizeof(WCHAR) )))
+            {
+                memcpy( ret, p, len * sizeof(WCHAR) );
+                ret[len] = 0;
+                break;
+            }
+        }
+        p++;
+    }
+    return ret;
+}
+
 /*************************************************************************
  *				SQLBrowseConnect           [ODBC32.055]
  */
@@ -3992,6 +4026,7 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL is_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength, %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_sqlstr(InConnectionString, StringLength1),
@@ -3999,13 +4034,16 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        is_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, is_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -5083,6 +5121,7 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL is_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString %p,"
           " BufferLength %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -5091,13 +5130,16 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        is_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, is_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -5310,7 +5352,7 @@ SQLRETURN WINAPI SQLConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *ServerName, SQL
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!(filename = get_driver_filename( ServerName )))
+    if (!(filename = get_driver_filename( ServerName, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -6137,6 +6179,7 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL is_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString %p,"
           " BufferLength %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -6145,13 +6188,16 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        is_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, is_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -6553,6 +6599,7 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL is_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_sqlwstr(InConnectionString, StringLength1), StringLength1,
@@ -6560,13 +6607,16 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        is_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, is_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
-- 
2.43.0

