From 02839f3df7ab1c8417a3ecd507a2b7c24dc6963f Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:13:33 +1000
Subject: [PATCH 04/15] odbc32: Support Driver in connection string

---
 dlls/odbc32/proxyodbc.c | 100 +++++++++++++++++++++++++++++++---------
 1 file changed, 77 insertions(+), 23 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index e8e08fb7d00..511fd8c9de3 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -821,11 +821,20 @@ static WCHAR *strdupAW( const char *src )
     return dst;
 }
 
-static HKEY open_odbcini_key( HKEY root )
+static HKEY open_odbcini_key( HKEY root, BOOL use_dsn )
 {
     static const WCHAR sourcesW[] = L"Software\\ODBC\\ODBC.INI";
+    static const WCHAR driversW[] = L"Software\\ODBC\\ODBCINST.INI";
     HKEY key;
-    if (!RegCreateKeyExW( root, sourcesW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+
+    if (use_dsn)
+    {
+        if (!RegCreateKeyExW( root, sourcesW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+    }
+    else
+    {
+        if (!RegCreateKeyExW( root, driversW, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &key, NULL )) return key;
+    }
     return NULL;
 }
 
@@ -839,12 +848,12 @@ static WCHAR *get_reg_value( HKEY key, const WCHAR *name )
     return NULL;
 }
 
-static WCHAR *get_driver_filename( const SQLWCHAR *source )
+static WCHAR *get_driver_filename( const SQLWCHAR *source, BOOL use_dsn )
 {
     HKEY key_root, key_source;
     WCHAR *ret = NULL;
 
-    if (!(key_root = open_odbcini_key( HKEY_CURRENT_USER ))) return NULL;
+    if (!(key_root = open_odbcini_key( HKEY_CURRENT_USER, use_dsn ))) return NULL;
     if (!RegOpenKeyExW( key_root, source, 0, KEY_READ, &key_source ))
     {
         ret = get_reg_value( key_source, L"Driver" );
@@ -853,7 +862,7 @@ static WCHAR *get_driver_filename( const SQLWCHAR *source )
     RegCloseKey( key_root );
     if (ret) return ret;
 
-    if (!(key_root = open_odbcini_key( HKEY_LOCAL_MACHINE ))) return NULL;
+    if (!(key_root = open_odbcini_key( HKEY_LOCAL_MACHINE, use_dsn ))) return NULL;
     if (!RegOpenKeyExW( key_root, source, 0, KEY_READ, &key_source ))
     {
         ret = get_reg_value( key_source, L"Driver" );
@@ -987,7 +996,7 @@ SQLRETURN WINAPI SQLConnect(SQLHDBC ConnectionHandle, SQLCHAR *ServerName, SQLSM
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!servername || !(filename = get_driver_filename( servername )))
+    if (!servername || !(filename = get_driver_filename( servername, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -2828,6 +2837,35 @@ static SQLRETURN browse_connect_unix_a( struct handle *handle, SQLCHAR *in_conn_
     return ODBC_CALL( SQLBrowseConnect, &params );
 }
 
+static WCHAR *get_driver( const WCHAR *connection_string )
+{
+    const WCHAR *p = connection_string, *q;
+    WCHAR *ret = NULL;
+    unsigned int len;
+
+    if (!p) return NULL;
+    while (*p)
+    {
+        if (!wcsnicmp( p, L"DRIVER=", 7 ))
+        {
+            p += 7;
+            if (*p == '{')
+                q = wcschr( ++p, '}' );
+            else
+                q = wcschr( p, ';' );
+            len = q ? (q - p) : wcslen( p );
+            if ((ret = malloc( (len + 1) * sizeof(WCHAR) )))
+            {
+                memcpy( ret, p, len * sizeof(WCHAR) );
+                ret[len] = 0;
+                break;
+            }
+        }
+        p++;
+    }
+    return ret;
+}
+
 /*************************************************************************
  *				SQLBrowseConnect           [ODBC32.055]
  */
@@ -2837,6 +2875,7 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength, %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_sqlstr(InConnectionString, StringLength1),
@@ -2844,13 +2883,16 @@ SQLRETURN WINAPI SQLBrowseConnect(SQLHDBC ConnectionHandle, SQLCHAR *InConnectio
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -3567,6 +3609,7 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource = NULL, *filename = NULL, *connection_string = strdupAW( (const char *)InConnectionString );
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString %p,"
           " BufferLength %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -3575,13 +3618,16 @@ SQLRETURN WINAPI SQLDriverConnect(SQLHDBC ConnectionHandle, SQLHWND WindowHandle
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!connection_string || !(datasource = get_datasource( connection_string )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( connection_string )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -3764,7 +3810,7 @@ SQLRETURN WINAPI SQLConnectW(SQLHDBC ConnectionHandle, WCHAR *ServerName, SQLSMA
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    if (!(filename = get_driver_filename( ServerName )))
+    if (!(filename = get_driver_filename( ServerName, TRUE )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -4395,6 +4441,7 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, WindowHandle %p, InConnectionString %s, Length %d, OutConnectionString %p,"
           " BufferLength %d, Length2 %p, DriverCompletion %d)\n", ConnectionHandle, WindowHandle,
@@ -4403,13 +4450,16 @@ SQLRETURN WINAPI SQLDriverConnectW(SQLHDBC ConnectionHandle, SQLHWND WindowHandl
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
@@ -4701,6 +4751,7 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
     struct handle *handle = ConnectionHandle;
     WCHAR *datasource, *filename = NULL;
     SQLRETURN ret = SQL_ERROR;
+    BOOL use_dsn = TRUE;
 
     TRACE("(ConnectionHandle %p, InConnectionString %s, StringLength1 %d, OutConnectionString %p, BufferLength %d, "
           "StringLength2 %p)\n", ConnectionHandle, debugstr_sqlwstr(InConnectionString, StringLength1), StringLength1,
@@ -4708,13 +4759,16 @@ SQLRETURN WINAPI SQLBrowseConnectW(SQLHDBC ConnectionHandle, SQLWCHAR *InConnect
 
     if (!handle) return SQL_INVALID_HANDLE;
 
-    /* FIXME: try DRIVER attribute if DSN is absent */
     if (!(datasource = get_datasource( InConnectionString )))
     {
-        WARN( "can't find data source\n" );
-        goto done;
+        use_dsn = FALSE;
+        if (!(datasource = get_driver( InConnectionString )))
+        {
+            WARN( "can't find data source\n" );
+            goto done;
+        }
     }
-    if (!(filename = get_driver_filename( datasource )))
+    if (!(filename = get_driver_filename( datasource, use_dsn )))
     {
         WARN( "can't find driver filename\n" );
         goto done;
-- 
2.43.0

