From 7d25d7fa0ee65c209e8c58f6ce276416931d8052 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 18 Jul 2024 07:13:48 +1000
Subject: [PATCH] odbc32: Record loaded driver SQL_ATTR_ODBC_VERSION

---
 dlls/odbc32/proxyodbc.c | 4 ++++
 dlls/odbc32/unixlib.h   | 1 +
 2 files changed, 5 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 4fdd242f3b5..c0ab9db9879 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -464,6 +464,7 @@ static struct environment *create_environment( void )
     if (!(ret = calloc( 1, sizeof(*ret) ))) return NULL;
     init_object( &ret->hdr, SQL_HANDLE_ENV, NULL );
     ret->attr_version = SQL_OV_ODBC2;
+    ret->driver_ver = SQL_OV_ODBC2;
     return ret;
 }
 
@@ -1265,6 +1266,9 @@ static SQLRETURN set_env_attr( struct environment *env, SQLINTEGER attr, SQLPOIN
     }
     else if (env->hdr.win32_handle)
     {
+        if (env->hdr.win32_funcs->SQLGetEnvAttr)
+           ret = env->hdr.win32_funcs->SQLGetEnvAttr( env->hdr.win32_handle, SQL_ATTR_ODBC_VERSION, &env->driver_ver, 0, NULL );
+
         if (env->hdr.win32_funcs->SQLSetEnvAttr)
             ret = env->hdr.win32_funcs->SQLSetEnvAttr( env->hdr.win32_handle, attr, value, len );
     }
diff --git a/dlls/odbc32/unixlib.h b/dlls/odbc32/unixlib.h
index c865a83711d..4ff13a68add 100644
--- a/dlls/odbc32/unixlib.h
+++ b/dlls/odbc32/unixlib.h
@@ -205,6 +205,7 @@ struct environment
     UINT32 sources_idx;
     void  *sources_key;
     BOOL   sources_system;
+    UINT32 driver_ver;
 };
 
 struct connection
-- 
2.47.2

