From c036cbda4e49b15e0a8dea55bb346929db5ce6c2 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:19:22 +1000
Subject: [PATCH 05/15] odbc32: SQLColAttributeW support fallback function

---
 dlls/odbc32/proxyodbc.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 511fd8c9de3..1f51ac2a27e 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -4060,6 +4060,21 @@ static SQLRETURN col_attribute_unix_w( struct handle *handle, SQLUSMALLINT col,
     return ret;
 }
 
+static SQLINTEGER map_odbc3_to_2(SQLINTEGER fieldid)
+{
+    switch( fieldid )
+    {
+        case SQL_DESC_COUNT:
+            return SQL_COLUMN_COUNT;
+        case SQL_DESC_NULLABLE:
+            return SQL_COLUMN_NULLABLE;
+        case SQL_DESC_NAME:
+            return SQL_COLUMN_NAME;
+        default:
+            return fieldid;
+    }
+}
+
 static SQLRETURN col_attribute_win32_w( struct handle *handle, SQLUSMALLINT col, SQLUSMALLINT field_id,
                                         SQLPOINTER char_attr, SQLSMALLINT buflen, SQLSMALLINT *retlen,
                                         SQLLEN *num_attr )
@@ -4067,6 +4082,14 @@ static SQLRETURN col_attribute_win32_w( struct handle *handle, SQLUSMALLINT col,
     if (handle->win32_funcs->SQLColAttributeW)
         return handle->win32_funcs->SQLColAttributeW( handle->win32_handle, col, field_id, char_attr, buflen,
                                                       retlen, num_attr );
+    else if(handle->win32_funcs->SQLColAttributesW)
+    {
+        /* ODBC v2 */
+        field_id = map_odbc3_to_2(field_id);
+        return handle->win32_funcs->SQLColAttributesW( handle->win32_handle, col, field_id,
+                                                 char_attr, buflen, retlen,
+                                                 num_attr );
+    }
     if (handle->win32_funcs->SQLColAttribute) FIXME( "Unicode to ANSI conversion not handled\n" );
     return SQL_ERROR;
 }
-- 
2.43.0

