From 2a020e9d35ae25858e0642b06c025dc686594fec Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 22:04:39 +1000
Subject: [PATCH] odbc32: SQLSetStmtAttrW support fallback for ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 742823cf08d..d0f3a837745 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -7948,6 +7948,11 @@ static SQLRETURN set_stmt_attr_win32_w( struct statement *stmt, SQLINTEGER attr,
     if (stmt->hdr.win32_funcs->SQLSetStmtAttrW)
         return stmt->hdr.win32_funcs->SQLSetStmtAttrW( stmt->hdr.win32_handle, attr, value, len );
     if (stmt->hdr.win32_funcs->SQLSetStmtAttr) FIXME( "Unicode to ANSI conversion not handled\n" );
+    else if (stmt->hdr.win32_funcs->SQLSetStmtOption)
+    {
+        /* ODBC v2.0 */
+        return stmt->hdr.win32_funcs->SQLSetStmtOption( stmt->hdr.win32_handle, attr, (SQLULEN)value );
+    }
     return SQL_ERROR;
 }
 
-- 
2.47.2

