From 7e2e9b4af4d8e56a9076510fc34bdfd44c9b3416 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 28 Aug 2024 07:57:17 +1000
Subject: [PATCH] odbc32: SQLColAttributes/W fall throught for unknown field
 ids

---
 dlls/odbc32/proxyodbc.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 3e5d840ef63..0893537aef9 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1049,7 +1049,6 @@ static SQLRETURN col_attribute_win32_a( struct statement *stmt, SQLUSMALLINT col
 
         default:
             FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
         }
 
         return stmt->hdr.win32_funcs->SQLColAttributes( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
@@ -6391,7 +6390,6 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
 
         default:
             FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
         }
 
         ret = stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
-- 
2.50.1

