From e78c554ba0dcbd24048d9ee716805b00b815f079 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 28 Aug 2024 07:57:17 +1000
Subject: [PATCH] odbc32: SQLColAttributes/W fall throught for unknown field
 ids

---
 dlls/odbc32/proxyodbc.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 4f31e869511..6669d0dbbfb 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1085,7 +1085,6 @@ static SQLRETURN col_attribute_win32_a( struct statement *stmt, SQLUSMALLINT col
 
         default:
             FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
         }
 
         ret = stmt->hdr.win32_funcs->SQLColAttributes( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
@@ -6778,7 +6777,6 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
 
         default:
             FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
         }
 
         ret = stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
-- 
2.53.0

