From 20374d93364244b20dc9c87b53c40c375d4f1235 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 12 Jul 2024 14:55:47 +1000
Subject: [PATCH] odbc32: SQLSetConnectAttr/W handle fallback function

---
 dlls/odbc32/proxyodbc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 7105f3534ce..dd94e070796 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6123,6 +6123,11 @@ static SQLRETURN set_connect_attr_win32_w( struct handle *handle, SQLINTEGER att
     if (handle->win32_funcs->SQLSetConnectAttrW)
         return handle->win32_funcs->SQLSetConnectAttrW( handle->win32_handle, attr, value, len );
     if (handle->win32_funcs->SQLSetConnectAttr) FIXME( "Unicode to ANSI conversion not handled\n" );
+    else if(handle->win32_funcs->SQLSetConnectOptionW)
+    {
+        /* ODBC v2 */
+        return handle->win32_funcs->SQLSetConnectOptionW( handle->win32_handle, attr, (SQLULEN)value );
+    }
     return SQL_ERROR;
 }
 
-- 
2.43.0

