From 53c9ccb1d5008cf348268e6a053ff6414bc4773e Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 10 Jul 2024 15:17:00 +1000
Subject: [PATCH] odbc32: Store handles when requesting information of Columns.

---
 dlls/odbc32/proxyodbc.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 632738b9533..ca598e404b1 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6315,7 +6315,27 @@ SQLRETURN WINAPI SQLGetStmtAttrW(SQLHSTMT StatementHandle, SQLINTEGER Attribute,
     }
     else if (stmt->hdr.win32_handle)
     {
-        ret = get_stmt_attr_win32_w( stmt, Attribute, Value, BufferLength, StringLength );
+        switch(Attribute)
+        {
+            case SQL_ATTR_APP_ROW_DESC:
+            case SQL_ATTR_IMP_ROW_DESC:
+            case SQL_ATTR_APP_PARAM_DESC:
+            case SQL_ATTR_IMP_PARAM_DESC:
+            {
+                struct descriptor *desc;
+                if ((desc = create_descriptor( stmt )))
+                {
+                    ret = stmt->hdr.win32_funcs->SQLGetStmtAttrW( stmt->hdr.win32_handle, Attribute,
+                                                             &desc->hdr.win32_handle,
+                                                             BufferLength, StringLength);
+                    desc->hdr.win32_funcs = stmt->hdr.win32_funcs;
+                    *((SQLHDESC*)Value) = desc;
+                }
+                break;
+            }
+            default:
+                ret = get_stmt_attr_win32_w( stmt, Attribute, Value, BufferLength, StringLength );
+        }
     }
 
     TRACE("Returning %d\n", ret);
-- 
2.43.0

