From 8c26d09014cdf26756bd808e2c076d9d10f6295f Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 13 Jul 2024 15:23:10 +1000
Subject: [PATCH 11/15] odbc32: SQLSetDescFieldW pass correct handle onto
 driver

---
 dlls/odbc32/proxyodbc.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 3a9c2802d45..c70b0d80012 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -4387,7 +4387,7 @@ SQLRETURN WINAPI SQLGetStmtAttrW(SQLHSTMT StatementHandle, SQLINTEGER Attribute,
                                  SQLINTEGER BufferLength, SQLINTEGER *StringLength)
 {
     struct handle *handle = StatementHandle;
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     TRACE("(StatementHandle %p, Attribute %d, Value %p, BufferLength %d, StringLength %p)\n", StatementHandle,
           Attribute, Value, BufferLength, StringLength);
@@ -5299,7 +5299,8 @@ done:
 SQLRETURN WINAPI SQLSetDescFieldW(SQLHDESC DescriptorHandle, SQLSMALLINT RecNumber, SQLSMALLINT FieldIdentifier,
                                   SQLPOINTER Value, SQLINTEGER BufferLength)
 {
-    struct handle *handle = DescriptorHandle;
+    struct SQLHDESC_data *hdesc = DescriptorHandle;
+    struct handle *handle = hdesc->parent;
     SQLRETURN ret = SQL_ERROR;
 
     TRACE("(DescriptorHandle %p, RecNumber %d, FieldIdentifier %d, Value %p, BufferLength %d)\n", DescriptorHandle,
@@ -5315,7 +5316,11 @@ SQLRETURN WINAPI SQLSetDescFieldW(SQLHDESC DescriptorHandle, SQLSMALLINT RecNumb
     }
     else if (handle->win32_handle)
     {
-        ret = handle->win32_funcs->SQLSetDescFieldW( handle->win32_handle, RecNumber, FieldIdentifier, Value,
+        if (handle->imp_param_desc.driver_hdesc)
+            ret = handle->win32_funcs->SQLSetDescFieldW( handle->imp_param_desc.driver_hdesc, RecNumber, FieldIdentifier, Value,
+                                                     BufferLength );
+        else
+            ret = handle->win32_funcs->SQLSetDescFieldW( handle->app_param_desc.driver_hdesc, RecNumber, FieldIdentifier, Value,
                                                      BufferLength );
     }
 
-- 
2.43.0

