From cee76bad632261533d15e6d642367fbb66176564 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 13 Jul 2024 15:23:10 +1000
Subject: [PATCH] odbc32: SQLSetDescFieldW pass correct handle onto driver

---
 dlls/odbc32/proxyodbc.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 7120f24b76b..ddf6c0ac9f5 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6086,7 +6086,7 @@ SQLRETURN WINAPI SQLGetStmtAttrW(SQLHSTMT StatementHandle, SQLINTEGER Attribute,
                                  SQLINTEGER BufferLength, SQLINTEGER *StringLength)
 {
     struct handle *handle = StatementHandle;
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     TRACE("(StatementHandle %p, Attribute %d, Value %p, BufferLength %d, StringLength %p)\n", StatementHandle,
           Attribute, Value, BufferLength, StringLength);
@@ -7268,7 +7268,8 @@ static SQLRETURN set_desc_field_win32_w( struct handle *handle, SQLSMALLINT reco
 SQLRETURN WINAPI SQLSetDescFieldW(SQLHDESC DescriptorHandle, SQLSMALLINT RecNumber, SQLSMALLINT FieldIdentifier,
                                   SQLPOINTER Value, SQLINTEGER BufferLength)
 {
-    struct handle *handle = DescriptorHandle;
+    struct SQLHDESC_data *hdesc = DescriptorHandle;
+    struct handle *handle = hdesc->parent;
     SQLRETURN ret = SQL_ERROR;
 
     TRACE("(DescriptorHandle %p, RecNumber %d, FieldIdentifier %d, Value %p, BufferLength %d)\n", DescriptorHandle,
@@ -7282,7 +7283,11 @@ SQLRETURN WINAPI SQLSetDescFieldW(SQLHDESC DescriptorHandle, SQLSMALLINT RecNumb
     }
     else if (handle->win32_handle)
     {
-        ret = set_desc_field_win32_w( handle, RecNumber, FieldIdentifier, Value, BufferLength );
+        if (handle->imp_param_desc.driver_hdesc)
+            ret = handle->win32_funcs->SQLSetDescFieldW( handle->imp_param_desc.driver_hdesc, RecNumber, FieldIdentifier, Value,
+                                                     BufferLength );
+        else
+            ret = set_desc_field_win32_w( handle, RecNumber, FieldIdentifier, Value, BufferLength );
     }
 
     TRACE("Returning %d\n", ret);
-- 
2.43.0

