From 285cfb3fbfb048b2a08a9c9b9c211a2972fa7f53 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 26 Apr 2025 17:32:49 +1000
Subject: [PATCH] odbc32: SQLDescribeColW add ANSI fallback

---
 dlls/odbc32/proxyodbc.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 6669d0dbbfb..bd4cd2bcbb5 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6372,11 +6372,18 @@ static SQLRETURN describe_col_win32_w( struct statement *stmt, SQLUSMALLINT col_
         ret = stmt->hdr.win32_funcs->SQLDescribeColW( stmt->hdr.win32_handle, col_number, col_name, buf_len,
                                                        name_len, data_type, col_size, decimal_digits, nullable );
     }
-    else if (stmt->hdr.win32_funcs->SQLDescribeCol) FIXME( "Unicode to ANSI conversion not handled\n" );
+    else if (stmt->hdr.win32_funcs->SQLDescribeCol)
+    {
+        SQLCHAR *name = (SQLCHAR*)strdupWA( (WCHAR*)col_name );
 
+        ret = stmt->hdr.win32_funcs->SQLDescribeCol( stmt->hdr.win32_handle, col_number, name, buf_len, name_len,
+                                                        data_type, col_size, decimal_digits, nullable);
+        free(name);
+    }
 
     if (SUCCESS( ret ) && data_type)
         *data_type = map_type( stmt, FALSE, *data_type );
+
     return ret;
 }
 
-- 
2.53.0

