From 37a034464df09450fe1d4226bf43ca7588c7894d Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 21:55:20 +1000
Subject: [PATCH] odbc32: SQLGetData support ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index be20e967aea..1836164796f 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -2223,7 +2223,20 @@ static SQLRETURN get_data_win32( struct handle *handle, SQLUSMALLINT column, SQL
                                  SQLLEN buflen, SQLLEN *retlen )
 {
     if (handle->win32_funcs->SQLGetData)
+    {
+        if (handle->env_attr_version == SQL_OV_ODBC2)
+        {
+            if (type == SQL_C_TYPE_TIME)
+                type = SQL_C_TIME;
+            else if (type == SQL_C_TYPE_DATE)
+                type = SQL_C_DATE;
+            else if (type == SQL_C_TYPE_TIMESTAMP)
+                type = SQL_C_TIMESTAMP;
+        }
+
         return handle->win32_funcs->SQLGetData( handle->win32_handle, column, type, value, buflen, retlen );
+    }
+
     return SQL_ERROR;
 }
 
-- 
2.43.0

