From 3a85cd3851edb650d8cac1d4b2275150c8314bdd Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sun, 14 Jul 2024 20:05:13 +1000
Subject: [PATCH] odbc32: Support Creating Environment handle with ODBC 2
 driver

---
 dlls/odbc32/proxyodbc.c | 31 +++++++++++++++++++++++++------
 1 file changed, 25 insertions(+), 6 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 30f3d5f34e9..f8b68bf44a5 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -814,7 +814,7 @@ static int has_suffix( const WCHAR *str, const WCHAR *suffix )
 
 static SQLRETURN set_env_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTER value, SQLINTEGER len )
 {
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     if (handle->unix_handle)
     {
@@ -823,7 +823,8 @@ static SQLRETURN set_env_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTE
     }
     else if (handle->win32_handle)
     {
-        ret = handle->win32_funcs->SQLSetEnvAttr( handle->win32_handle, attr, value, len );
+        if (handle->win32_funcs->SQLSetEnvAttr)
+           ret = handle->win32_funcs->SQLSetEnvAttr( handle->win32_handle, attr, value, len );
     }
     return ret;
 }
@@ -848,7 +849,15 @@ static SQLRETURN create_env( struct handle *handle, BOOL is_unix )
     }
     else
     {
-        if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_ENV, NULL, &handle->win32_handle ))) return ret;
+        if (handle->win32_funcs->SQLAllocHandle)
+        {
+            if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_ENV, NULL, &handle->win32_handle ))) return ret;
+        }
+        else if(handle->win32_funcs->SQLAllocEnv)
+        {
+            /* ODBC v2.0 */
+            if ((ret = handle->win32_funcs->SQLAllocEnv(&handle->win32_handle ))) return ret;
+        }
     }
 
     return prepare_env( handle );
@@ -856,7 +865,7 @@ static SQLRETURN create_env( struct handle *handle, BOOL is_unix )
 
 static SQLRETURN set_con_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTER value, SQLINTEGER len )
 {
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     if (handle->unix_handle)
     {
@@ -904,8 +913,18 @@ static SQLRETURN create_con( struct handle *handle )
     }
     else
     {
-        if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_DBC, parent->win32_handle, &handle->win32_handle )))
-            return ret;
+        if (handle->win32_funcs->SQLAllocHandle)
+        {
+            if ((ret = handle->win32_funcs->SQLAllocHandle( SQL_HANDLE_DBC, parent->win32_handle, &handle->win32_handle )))
+                return ret;
+        }
+        else if (handle->win32_funcs->SQLAllocConnect)
+        {
+            if ((ret = handle->win32_funcs->SQLAllocConnect( parent->win32_handle, &handle->win32_handle )))
+                return ret;
+        }
+        else
+            return SQL_ERROR;
     }
 
     return prepare_con( handle );
-- 
2.43.0

