From c41312e28ac2c500aad9e5cdbbbee4670f4b4a86 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 22:03:03 +1000
Subject: [PATCH 14/15] odbc32: SQLColAttributesW support ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index a66ec99b913..a9d23c94504 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -4144,11 +4144,27 @@ static SQLRETURN col_attribute_win32_w( struct handle *handle, SQLUSMALLINT col,
                                                       retlen, num_attr );
     else if(handle->win32_funcs->SQLColAttributesW)
     {
+        SQLRETURN ret;
+
         /* ODBC v2 */
         field_id = map_odbc3_to_2(field_id);
-        return handle->win32_funcs->SQLColAttributesW( handle->win32_handle, col, field_id,
+        ret = handle->win32_funcs->SQLColAttributesW( handle->win32_handle, col, field_id,
                                                  char_attr, buflen, retlen,
                                                  num_attr );
+        /* Convert back for ODBC3 drivers */
+        if (num_attr && field_id == SQL_COLUMN_TYPE &&
+                handle->env_attr_version == SQL_OV_ODBC2 &&
+                handle->driver_ver == SQL_OV_ODBC2)
+        {
+            if (*num_attr == SQL_TIME)
+                *num_attr = SQL_TYPE_TIME;
+            else if (*num_attr == SQL_DATETIME)
+                *num_attr = SQL_TYPE_DATE;
+            else if (*num_attr == SQL_TIMESTAMP)
+                *num_attr = SQL_TYPE_TIMESTAMP;
+        }
+
+        return ret;
     }
     if (handle->win32_funcs->SQLColAttribute) FIXME( "Unicode to ANSI conversion not handled\n" );
     return SQL_ERROR;
-- 
2.43.0

