From 7fd12044418fbd610b920961db964ff0acd098f4 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 22:03:03 +1000
Subject: [PATCH] odbc32: SQLColAttributesW support ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index b2bf2a7d58c..f7a6535db37 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -5973,11 +5973,27 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
                                                        retlen, num_attr );
     else if(stmt->hdr.win32_funcs->SQLColAttributesW)
     {
+        SQLRETURN ret;
+
         /* ODBC v2 */
         field_id = map_odbc3_to_2(field_id);
-        return stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id,
-                                                 char_attr, buflen, retlen,
-                                                 num_attr );
+        ret = stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id,
+                                                 char_attr, buflen, retlen, num_attr );
+
+        /* Convert back for ODBC3 drivers */
+        if (num_attr && field_id == SQL_COLUMN_TYPE &&
+                ((struct environment*)(stmt->hdr.parent))->attr_version == SQL_OV_ODBC2 &&
+                ((struct environment*)(stmt->hdr.parent))->driver_ver == SQL_OV_ODBC2)
+        {
+            if (*num_attr == SQL_TIME)
+                *num_attr = SQL_TYPE_TIME;
+            else if (*num_attr == SQL_DATETIME)
+                *num_attr = SQL_TYPE_DATE;
+            else if (*num_attr == SQL_TIMESTAMP)
+                *num_attr = SQL_TYPE_TIMESTAMP;
+        }
+
+        return ret;
     }
     if (stmt->hdr.win32_funcs->SQLColAttribute) FIXME( "Unicode to ANSI conversion not handled\n" );
 
-- 
2.43.0

