From b86180b88e1d485bb628a195e2315be2bada2898 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 22:04:39 +1000
Subject: [PATCH] odbc32: SQLSetStmtAttrW support fallback for ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 3429e8a26ce..2464ddfebee 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -7350,6 +7350,11 @@ static SQLRETURN set_stmt_attr_win32_w( struct handle *handle, SQLINTEGER attr,
     if (handle->win32_funcs->SQLSetStmtAttrW)
         return handle->win32_funcs->SQLSetStmtAttrW( handle->win32_handle, attr, value, len );
     if (handle->win32_funcs->SQLSetStmtAttr) FIXME( "Unicode to ANSI conversion not handled\n" );
+    else if (handle->win32_funcs->SQLSetStmtOption)
+    {
+        /* ODBC v2.0 */
+        return handle->win32_funcs->SQLSetStmtOption( handle->win32_handle, attr, (SQLULEN)value );
+    }
     return SQL_ERROR;
 }
 
-- 
2.43.0

