From 0f53e844d9fba3ff36f340d989f7d16223a62dfd Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 16:58:44 +1000
Subject: [PATCH] odbc32: SQLAllocHandle use fallback for ODBC 2.0

---
 dlls/odbc32/proxyodbc.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index dd68872743e..e6ffd4d6fc7 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -440,7 +440,20 @@ SQLRETURN WINAPI SQLAllocHandle(SQLSMALLINT HandleType, SQLHANDLE InputHandle, S
     }
     else if (input->win32_handle)
     {
-        ret = input->win32_funcs->SQLAllocHandle( HandleType, input->win32_handle, &output->win32_handle );
+        if(input->win32_funcs->SQLAllocHandle)
+            ret = input->win32_funcs->SQLAllocHandle( HandleType, input->win32_handle, &output->win32_handle );
+        else
+        {
+            /* ODBC v2.0 */
+            if (HandleType == SQL_HANDLE_STMT)
+            {
+                if (input->win32_funcs->SQLAllocStmt)
+                {
+                    ret = input->win32_funcs->SQLAllocStmt( input->win32_handle, &output->win32_handle );
+                }
+            }
+        }
+
         if (SUCCESS( ret )) output->win32_funcs = input->win32_funcs;
     }
 
-- 
2.43.0

