From 3472c0211aa199df18e0d328d31d9efc290b89b4 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 18 Jul 2024 07:13:48 +1000
Subject: [PATCH] odbc32: Record loaded driver SQL_ATTR_ODBC_VERSION

---
 dlls/odbc32/proxyodbc.c | 4 ++++
 dlls/odbc32/unixlib.h   | 1 +
 2 files changed, 5 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index e6ffd4d6fc7..611362dbd7c 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -376,6 +376,7 @@ static struct handle *create_handle( struct handle *parent )
     if (!(ret = calloc( 1, sizeof(*ret) ))) return NULL;
     ret->parent = parent;
     ret->env_attr_version = SQL_OV_ODBC2;
+    ret->driver_ver = SQL_OV_ODBC2;
     ret->row_count = 1;
     return ret;
 }
@@ -872,6 +873,9 @@ static SQLRETURN set_env_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTE
     }
     else if (handle->win32_handle)
     {
+        if (handle->win32_funcs->SQLGetEnvAttr)
+           ret = handle->win32_funcs->SQLGetEnvAttr( handle->win32_handle, SQL_ATTR_ODBC_VERSION, &handle->driver_ver, 0, NULL );
+
         if (handle->win32_funcs->SQLSetEnvAttr)
            ret = handle->win32_funcs->SQLSetEnvAttr( handle->win32_handle, attr, value, len );
     }
diff --git a/dlls/odbc32/unixlib.h b/dlls/odbc32/unixlib.h
index fc69857c8b7..61b8429768d 100644
--- a/dlls/odbc32/unixlib.h
+++ b/dlls/odbc32/unixlib.h
@@ -211,6 +211,7 @@ struct handle
     UINT32 sources_idx;
     void  *sources_key;
     BOOL   sources_system;
+    UINT32 driver_ver;
     /* parameter bindings */
     struct param_binding bind_col;
     struct param_binding bind_parameter;
-- 
2.43.0

