From c32d12a345ce533dd7e785c439cede340cfef17e Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 27 Aug 2024 07:25:34 +1000
Subject: [PATCH] odbc32: Always pass through the field id to
 SQLColAttributes/W

---
 dlls/odbc32/proxyodbc.c | 14 ++------------
 1 file changed, 2 insertions(+), 12 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index f6a5ea01e22..82859650993 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1004,13 +1004,8 @@ static SQLRETURN col_attribute_win32_a( struct statement *stmt, SQLUSMALLINT col
             field_id = SQL_DESC_NULLABLE;
             break;
 
-        case SQL_COLUMN_TYPE:
-        case SQL_COLUMN_DISPLAY_SIZE:
-            break;
-
         default:
-            FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
+            /* Pass through field id. */
         }
 
         return stmt->hdr.win32_funcs->SQLColAttributes( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
@@ -6230,13 +6225,8 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
             field_id = SQL_DESC_NULLABLE;
             break;
 
-        case SQL_COLUMN_TYPE:
-        case SQL_COLUMN_DISPLAY_SIZE:
-            break;
-
         default:
-            FIXME( "field id %u not handled\n", field_id );
-            return SQL_ERROR;
+            /* Pass through field id. */
         }
 
         ret = stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id, char_attr, buflen,
-- 
2.45.2

