From bf50a23bd4bd098a51b4944f1c22a9d2a71fc454 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 20 Jul 2024 12:56:26 +1000
Subject: [PATCH] odbc32: set_con_attr return success by default

---
 dlls/odbc32/proxyodbc.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index d0661406fae..cf13669d051 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1219,7 +1219,7 @@ static SQLRETURN create_env( struct environment *env, BOOL is_unix )
 
 static SQLRETURN set_con_attr( struct connection *con, SQLINTEGER attr, SQLPOINTER value, SQLINTEGER len )
 {
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     if (con->hdr.unix_handle)
     {
@@ -1242,6 +1242,17 @@ static SQLRETURN set_con_attr( struct connection *con, SQLINTEGER attr, SQLPOINT
             ret = con->hdr.win32_funcs->SQLSetConnectAttrW( con->hdr.win32_handle, attr, value, len );
         else if (con->hdr.win32_funcs->SQLSetConnectAttr)
             ret = con->hdr.win32_funcs->SQLSetConnectAttr( con->hdr.win32_handle, attr, value, len );
+        else if (con->hdr.win32_funcs->SQLSetConnectOptionW)
+        {
+            ret = con->hdr.win32_funcs->SQLSetConnectOptionW( con->hdr.win32_handle, attr, (SQLULEN)value );
+
+            /* If an option isn't supported on other driver, then just return success for now. */
+            if(ret != SQL_SUCCESS)
+            {
+                FIXME("Driver returned %d for option %d\n", ret, attr);
+                ret = SQL_SUCCESS;
+            }
+        }
     }
 
     return ret;
-- 
2.43.0

