From a286fe2aca1bd8f7c2a95675394fe554d16ef7fa Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 20 Jul 2024 12:56:26 +1000
Subject: [PATCH] odbc32: set_con_attr return success by default

---
 dlls/odbc32/proxyodbc.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index dee73f0618e..d29069cb021 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -994,7 +994,7 @@ static SQLRETURN create_env( struct handle *handle, BOOL is_unix )
 
 static SQLRETURN set_con_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTER value, SQLINTEGER len )
 {
-    SQLRETURN ret = SQL_ERROR;
+    SQLRETURN ret = SQL_SUCCESS;
 
     if (handle->unix_handle)
     {
@@ -1017,6 +1017,17 @@ static SQLRETURN set_con_attr( struct handle *handle, SQLINTEGER attr, SQLPOINTE
             ret = handle->win32_funcs->SQLSetConnectAttrW( handle->win32_handle, attr, value, len );
         else if (handle->win32_funcs->SQLSetConnectAttr)
             ret = handle->win32_funcs->SQLSetConnectAttr( handle->win32_handle, attr, value, len );
+        else if (handle->win32_funcs->SQLSetConnectOptionW)
+        {
+            ret = handle->win32_funcs->SQLSetConnectOptionW( handle->win32_handle, attr, (SQLULEN)value );
+
+            /* If an option isn't supported on other driver, then just return success for now. */
+            if(ret != SQL_SUCCESS)
+            {
+                FIXME("Driver returned %d for option %d\n", ret, attr);
+                ret = SQL_SUCCESS;
+            }
+        }
     }
     return ret;
 }
-- 
2.43.0

