From 610fdcdfc1a63432a6476adc46050fbe30db4f81 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 21:55:20 +1000
Subject: [PATCH] odbc32: SQLGetData support ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 611362dbd7c..9ca4f35c332 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1750,6 +1750,15 @@ SQLRETURN WINAPI SQLGetData(SQLHSTMT StatementHandle, SQLUSMALLINT ColumnNumber,
     }
     else if (handle->win32_handle)
     {
+        if (handle->env_attr_version == SQL_OV_ODBC2)
+        {
+            if (TargetType == SQL_C_TYPE_TIME)
+                TargetType = SQL_C_TIME;
+            else if (TargetType == SQL_C_TYPE_DATE)
+                TargetType = SQL_C_DATE;
+            else if (TargetType == SQL_C_TYPE_TIMESTAMP)
+                TargetType = SQL_C_TIMESTAMP;
+        }
         ret = handle->win32_funcs->SQLGetData( handle->win32_handle, ColumnNumber, TargetType, TargetValue,
                                                BufferLength, StrLen_or_Ind );
     }
-- 
2.43.0

