From daff4a3d283837c83f33bcc0141af4078e6251ac Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 14 Aug 2024 07:52:10 +1000
Subject: [PATCH] odbc32: Pass statment when creating a descriptor

---
 dlls/odbc32/proxyodbc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index c12910161ab..2e55e1ee06e 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -547,7 +547,7 @@ static SQLRETURN alloc_descriptors( struct statement *stmt )
     unsigned int i;
     for (i = 0; i < ARRAY_SIZE(stmt->desc); i++)
     {
-        if (!(stmt->desc[i] = create_descriptor( NULL )))
+        if (!(stmt->desc[i] = create_descriptor( stmt )))
         {
             free_descriptors( stmt );
             return SQL_ERROR;
-- 
2.43.0

