From 14b0cc45d3d19baf6c1f2534f8f31b7b5299c54f Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 10 Aug 2024 14:43:02 +1000
Subject: [PATCH] odbc32: Call drivers SQLBindCol function

---
 dlls/odbc32/proxyodbc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 0fa0102f2bb..35744622513 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -664,7 +664,7 @@ static SQLRETURN bind_col_win32( struct handle *handle, SQLUSMALLINT column, SQL
                                  SQLLEN buflen, SQLLEN *retlen )
 {
     if (handle->win32_funcs->SQLBindCol)
-        return SQLBindCol( handle->win32_handle, column, type, value, buflen, retlen );
+        return handle->win32_funcs->SQLBindCol( handle->win32_handle, column, type, value, buflen, retlen );
     return SQL_ERROR;
 }
 
-- 
2.43.0

