From 9b5c4a85620c6f86a368d420221199da2f794c63 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 14 Aug 2024 07:51:18 +1000
Subject: [PATCH] odbc32: Propgate win32_funcs to all children

---
 dlls/odbc32/proxyodbc.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 932560015b8..c12910161ab 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -397,7 +397,11 @@ static void init_object( struct object *obj, UINT32 type, struct object *parent
     obj->parent = parent;
     list_init( &obj->entry );
     list_init( &obj->children );
-    if (parent) list_add_tail( &parent->children, &obj->entry );
+    if (parent)
+    {
+        list_add_tail( &parent->children, &obj->entry );
+        obj->win32_funcs = parent->win32_funcs;
+    }
     InitializeCriticalSectionEx( &obj->cs, 0, RTL_CRITICAL_SECTION_FLAG_FORCE_DEBUG_INFO );
     obj->cs.DebugInfo->Spare[0] = (DWORD_PTR)(__FILE__ ": object.cs");
 }
-- 
2.43.0

