From fa52f2afdfb554e43ca32a137378e9aaa052a097 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 17 Jul 2024 22:04:39 +1000
Subject: [PATCH] odbc32: SQLSetStmtAttrW support fallback for ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 91881635cab..f6e2cd61436 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -5388,7 +5388,17 @@ SQLRETURN WINAPI SQLSetStmtAttrW(SQLHSTMT StatementHandle, SQLINTEGER Attribute,
     }
     else if (handle->win32_handle)
     {
-        ret = handle->win32_funcs->SQLSetStmtAttrW( handle->win32_handle, Attribute, Value, StringLength );
+        if(handle->win32_funcs->SQLSetStmtAttrW)
+            ret = handle->win32_funcs->SQLSetStmtAttrW( handle->win32_handle, Attribute, Value, StringLength );
+        else
+        {
+            /* ODBC v2.0 */
+            if (handle->win32_funcs->SQLSetStmtOption)
+            {
+                ret = handle->win32_funcs->SQLSetStmtOption( handle->win32_handle, Attribute, (SQLULEN)Value );
+            }
+        }
+
     }
 
     TRACE("Returning %d\n", ret);
-- 
2.43.0

