From 9b36b65dd25db374f53f51cc24b36637e80de2c2 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 26 Apr 2025 16:27:58 +1000
Subject: [PATCH] odbc32: SQLGetInfoW support ascii fallback.

---
 dlls/odbc32/proxyodbc.c | 31 +++++++++++++++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 7025a141bb6..495bf2ff446 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6924,13 +6924,40 @@ static SQLRETURN get_info_unix_w( struct connection *con, SQLUSMALLINT type, SQL
     return ODBC_CALL( SQLGetInfoW, &params );
 }
 
+static BOOL typeinfo_is_string( SQLSMALLINT type )
+{
+    switch (type)
+    {
+        case SQL_DRIVER_ODBC_VER:
+        case SQL_DRIVER_NAME:
+            return TRUE;
+        default:
+        {
+            FIXME("Type %d not handled\n", type);
+            return FALSE;
+        }
+    }
+}
+
 static SQLRETURN get_info_win32_w( struct connection *con, SQLUSMALLINT type, SQLPOINTER value, SQLSMALLINT buflen,
                                    SQLSMALLINT *retlen )
 {
+    SQLRETURN ret = SQL_ERROR;
+
     if (con->hdr.win32_funcs->SQLGetInfoW)
         return con->hdr.win32_funcs->SQLGetInfoW( con->hdr.win32_handle, type, value, buflen, retlen );
-    if (con->hdr.win32_funcs->SQLGetInfo) FIXME( "Unicode to ANSI conversion not handled\n" );
-    return SQL_ERROR;
+    if (con->hdr.win32_funcs->SQLGetInfo)
+    {
+        ret = con->hdr.win32_funcs->SQLGetInfo( con->hdr.win32_handle, type, value, buflen, retlen );
+        FIXME("st %s\n", debugstr_a(value));
+        if (SQL_SUCCEEDED(ret) && typeinfo_is_string(type))
+        {
+            WCHAR *p = strnAtoW(value, buflen);
+            wcscpy(value, p);
+            free(p);
+        }
+    }
+    return ret;
 }
 
 /*************************************************************************
-- 
2.47.2

