From 8f32508f96c33d9a47e3b664c8d6c39454802e4f Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 1 May 2025 08:10:04 +1000
Subject: [PATCH] odbc32: SQLPrepareW add ansi fallback

---
 dlls/odbc32/proxyodbc.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index dcd38eb1066..c057c8ee815 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -6634,10 +6634,17 @@ static SQLRETURN prepare_unix_w( struct statement *stmt, SQLWCHAR *statement, SQ
 
 static SQLRETURN prepare_win32_w( struct statement *stmt, SQLWCHAR *statement, SQLINTEGER len )
 {
+    SQLRETURN ret = SQL_ERROR;
+
     if (stmt->hdr.win32_funcs->SQLPrepareW)
-        return stmt->hdr.win32_funcs->SQLPrepareW( stmt->hdr.win32_handle, statement, len );
-    if (stmt->hdr.win32_funcs->SQLPrepare) FIXME( "Unicode to ANSI conversion not handled\n" );
-    return SQL_ERROR;
+        ret = stmt->hdr.win32_funcs->SQLPrepareW( stmt->hdr.win32_handle, statement, len );
+    else if (stmt->hdr.win32_funcs->SQLPrepare)
+    {
+        SQLCHAR *statementA = strnWtoA( statement, len );
+        ret = stmt->hdr.win32_funcs->SQLPrepare( stmt->hdr.win32_handle, statementA, SQL_NTS );
+        free(statementA);
+    }
+    return ret;
 }
 
 /*************************************************************************
-- 
2.53.0

