From f9dc463e820b2b8840ef1b98957ba6f9b67d8e32 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 5 May 2025 09:29:17 +1000
Subject: [PATCH 1/4] odbc32: SQLColAttributesW accept field id
 SQL_COLUMN_NAME.

---
 dlls/odbc32/proxyodbc.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 76c104e3a4a..ce69e3ae134 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1028,6 +1028,7 @@ static SQLRETURN col_attribute_win32_a( struct statement *stmt, SQLUSMALLINT col
             field_id = SQL_COLUMN_NAME;
             break;
 
+        case SQL_COLUMN_NAME:
         case SQL_COLUMN_TYPE:
         case SQL_COLUMN_DISPLAY_SIZE:
         case SQL_MAX_COLUMNS_IN_TABLE:
@@ -6386,6 +6387,7 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
             field_id = SQL_COLUMN_NAME;
             break;
 
+        case SQL_COLUMN_NAME:
         case SQL_COLUMN_TYPE:
         case SQL_COLUMN_DISPLAY_SIZE:
         case SQL_MAX_COLUMNS_IN_TABLE:
-- 
2.47.2

