From 085539732d3dcd3aadfda4fc022de8f33be1e509 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 28 Apr 2025 10:54:55 +1000
Subject: [PATCH] odbc32: SQLColAttributesW support ascii fallback

---
 dlls/odbc32/proxyodbc.c | 42 ++++++++++++++++++++++++++++++++++++++---
 1 file changed, 39 insertions(+), 3 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 31083792f8f..b490da9012f 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -4636,6 +4636,10 @@ static SQLRETURN col_attributes_unix_a( struct statement *stmt, SQLUSMALLINT col
     return ret;
 }
 
+static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col, SQLUSMALLINT field_id,
+                                        SQLPOINTER char_attr, SQLSMALLINT buflen, SQLSMALLINT *retlen,
+                                        SQLLEN *num_attr );
+
 static SQLRETURN col_attributes_win32_a( struct statement *stmt, SQLUSMALLINT col, SQLUSMALLINT field_id,
                                          SQLPOINTER char_attrs, SQLSMALLINT buflen, SQLSMALLINT *retlen,
                                          SQLLEN *num_attrs )
@@ -5839,11 +5843,17 @@ static SQLRETURN col_attributes_win32_w( struct statement *stmt, SQLUSMALLINT co
                                          SQLPOINTER char_attrs, SQLSMALLINT buflen, SQLSMALLINT *retlen,
                                          SQLLEN *num_attrs )
 {
+    SQLRETURN ret = SQL_ERROR;
+
     if (stmt->hdr.win32_funcs->SQLColAttributesW)
         return stmt->hdr.win32_funcs->SQLColAttributesW( stmt->hdr.win32_handle, col, field_id, char_attrs, buflen,
                                                          retlen, num_attrs );
     if (stmt->hdr.win32_funcs->SQLColAttributes) FIXME( "Unicode to ANSI conversion not handled\n" );
-    return SQL_ERROR;
+    else
+    {
+       ret = col_attribute_win32_w( stmt, col, field_id, char_attrs, buflen, retlen, num_attrs );
+    }
+    return ret;
 }
 
 /*************************************************************************
@@ -6295,8 +6305,34 @@ static SQLRETURN col_attribute_win32_w( struct statement *stmt, SQLUSMALLINT col
 
     if (stmt->hdr.win32_funcs->SQLColAttribute)
     {
-        FIXME( "Unicode to ANSI conversion not handled\n" );
-        return SQL_ERROR;
+        SQLCHAR *strA = char_attr;
+
+        if (char_attr && buflen && SQLColAttributes_KnownStringAttribute(field_id))
+        {
+            strA = malloc( buflen );
+        }
+
+        ret = stmt->hdr.win32_funcs->SQLColAttribute( stmt->hdr.win32_handle, col, field_id, strA, buflen,
+                                                         retlen, num_attr );
+        if (ret == SQL_SUCCESS && SQLColAttributes_KnownStringAttribute(field_id))
+        {
+            if (strA)
+            {
+                WCHAR *p = strnAtoW(strA, -1);
+                wcscpy(char_attr, p);
+                free(p);
+
+                if (retlen)
+                    *retlen = wcslen( char_attr ) * sizeof(WCHAR);
+            }
+            else if (retlen)
+                *retlen = *retlen * sizeof(WCHAR);
+        }
+
+        if (strA != char_attr)
+            free(strA);
+
+        return ret;
     }
 
     if (stmt->hdr.win32_funcs->SQLColAttributesW)
-- 
2.47.2

