From 0442ed8d597eb250b7ebc41261073f71d9368f98 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 2 Apr 2016 00:22:30 +0200
Subject: [PATCH] shell32: Fix copying of files when using a context menu.

---
 dlls/shell32/shlview_cmenu.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/shell32/shlview_cmenu.c b/dlls/shell32/shlview_cmenu.c
index 67f5ac80d84..634e3b07176 100644
--- a/dlls/shell32/shlview_cmenu.c
+++ b/dlls/shell32/shlview_cmenu.c
@@ -1164,6 +1164,13 @@ static BOOL DoPaste(ContextMenu *This)
 
 	    apidl = _ILCopyCidaToaPidl(&pidl, lpcida);
 
+            /*
+             * In case source is a file we need to remove the last component
+             * to obtain a IShellFolder of the parent.
+             */
+            if (_ILIsValue(pidl))
+                ILRemoveLastID(pidl);
+
 	    for (i = 0; bSuccess && i < lpcida->cidl; i++) {
 	      ITEMIDLIST *apidl_dir = NULL;
 	      ITEMIDLIST *apidl_item;
-- 
2.30.2

