From 5d231b5fb9c07a7d43db75370ea9b8e629946987 Mon Sep 17 00:00:00 2001
From: Jactry Zeng <wine@jactry.com>
Date: Fri, 3 Jun 2022 10:22:50 +1000
Subject: [PATCH] dwmapi: DwmGetCompositionTimingInfo supply defaults

---
 dlls/dwmapi/dwmapi_main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/dwmapi/dwmapi_main.c b/dlls/dwmapi/dwmapi_main.c
index 6bb086a87d1..339f05e1a30 100644
--- a/dlls/dwmapi/dwmapi_main.c
+++ b/dlls/dwmapi/dwmapi_main.c
@@ -229,6 +229,12 @@ HRESULT WINAPI DwmGetCompositionTimingInfo(HWND hwnd, DWM_TIMING_INFO *info)
     memset(info, 0, info->cbSize);
     info->cbSize = sizeof(DWM_TIMING_INFO);
 
+    info->rateRefresh.uiDenominator = 1;
+    info->rateRefresh.uiNumerator = 64;
+    info->rateCompose.uiDenominator = 1;
+    info->rateCompose.uiNumerator = 64;
+    info->qpcRefreshPeriod = 156250;
+
     return S_OK;
 }
 
-- 
2.35.1

