From 5aa6c213ced971079f236acfe25437f5d49cebbf Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 10 Jan 2023 16:22:14 +1100
Subject: [PATCH 1/4] include: Add _XHR enum values

---
 include/msxml6.idl | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/include/msxml6.idl b/include/msxml6.idl
index 4672ae80626..aa03a8a310a 100644
--- a/include/msxml6.idl
+++ b/include/msxml6.idl
@@ -256,6 +256,20 @@ typedef enum _SCHEMATYPEVARIETY
 } SCHEMATYPEVARIETY;
 cpp_quote("#endif /* __msxml_som_enums__ */")
 
+typedef [v1_enum] enum _XHR_CRED_PROMPT
+{
+    XHR_CRED_PROMPT_ALL   = 0x0,
+    XHR_CRED_PROMPT_NONE  = 0x1,
+    XHR_CRED_PROMPT_PROXY = 0x2,
+} XHR_CRED_PROMPT;
+
+typedef [v1_enum] enum _XHR_AUTH
+{
+    XHR_AUTH_ALL   = 0x0,
+    XHR_AUTH_NONE  = 0x1,
+    XHR_AUTH_PROXY = 0x2,
+} XHR_AUTH;
+
 typedef [v1_enum] enum _XHR_PROPERTY
 {
     XHR_PROP_NO_CRED_PROMPT,
-- 
2.39.0

