From a4077e33ab9e41927433479a9a8af9ed11990c04 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 12 Jan 2023 08:21:48 +1100
Subject: [PATCH 4/4] msxml3: Correct xml_http_request_2_GetResponseHeader

---
 dlls/msxml3/httprequest.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/dlls/msxml3/httprequest.c b/dlls/msxml3/httprequest.c
index 48ed444576e..111ac39f62f 100644
--- a/dlls/msxml3/httprequest.c
+++ b/dlls/msxml3/httprequest.c
@@ -2339,8 +2339,7 @@ static HRESULT WINAPI xml_http_request_2_GetResponseHeader(IXMLHTTPRequest3 *ifa
 
     TRACE("(%p)->(%s %p)\n", This, debugstr_w(header), value);
 
-    if (FAILED(hr = httprequest_getResponseHeader(&This->req, (BSTR)header, value)))
-        return hr;
+    hr = httprequest_getResponseHeader(&This->req, (BSTR)header, value);
 
 #define E_FILE_NOT_FOUND                                   _HRESULT_TYPEDEF_(0x80070002)
 
-- 
2.39.0

