From a1d1a521b9e3d3ee418fb7449c090f9d61c6bd01 Mon Sep 17 00:00:00 2001
From: Ralf Habacker <ralf.habacker@freenet.de>
Date: Wed, 28 May 2025 01:12:26 +0200
Subject: [PATCH] ws2_32: Add note in bind() for AF_UNIX sockets

---
 dlls/ws2_32/socket.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index 7a5720583d3..818ad66e30a 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -1340,6 +1340,8 @@ int WINAPI bind( SOCKET s, const struct sockaddr *addr, int len )
 
     if (addr->sa_family == AF_UNIX && *addr->sa_data)
     {
+        /* The corresponding unix path is appended to a buffer with
+         * the structure sockaddr_un and can have a length of <= PATH_MAX */
         struct sockaddr_un sun = { 0 };
         WCHAR *sun_pathW;
         memcpy(&sun, addr, len);
-- 
2.53.0

