From bea6b99a50c4de12c03d6dd68e9f246b854f9550 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 8 Jun 2024 10:24:40 +1000
Subject: [PATCH 1/2] odbccp32: Support System wide ODBC DSN keys

---
 dlls/odbccp32/odbccp32.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/dlls/odbccp32/odbccp32.c b/dlls/odbccp32/odbccp32.c
index 95bfb90781e..7846f29584f 100644
--- a/dlls/odbccp32/odbccp32.c
+++ b/dlls/odbccp32/odbccp32.c
@@ -713,19 +713,25 @@ BOOL WINAPI SQLGetInstalledDrivers(char *buf, WORD size, WORD *sizeout)
 
 static HKEY get_privateprofile_sectionkey(LPCWSTR section, LPCWSTR filename)
 {
-    HKEY hkey, hkeyfilename, hkeysection;
+    HKEY hkeysection;
     LONG ret;
+    WCHAR *regpath;
 
-    if (RegOpenKeyW(HKEY_CURRENT_USER, odbcW, &hkey))
+    regpath = malloc ( (wcslen(L"Software\\ODBC\\") + wcslen(filename) + wcslen(L"\\")
+                            + wcslen(section) + 1) * sizeof(WCHAR));
+    if (!regpath)
         return NULL;
 
-    ret = RegOpenKeyW(hkey, filename, &hkeyfilename);
-    RegCloseKey(hkey);
-    if (ret)
-        return NULL;
+    wcscpy(regpath, L"Software\\ODBC\\");
+    wcscat(regpath, filename);
+    wcscat(regpath, L"\\");
+    wcscat(regpath, section);
 
-    ret = RegOpenKeyW(hkeyfilename, section, &hkeysection);
-    RegCloseKey(hkeyfilename);
+    if ((ret = RegOpenKeyW(HKEY_CURRENT_USER, regpath, &hkeysection)) != ERROR_SUCCESS)
+    {
+        ret = RegOpenKeyW(HKEY_LOCAL_MACHINE, regpath, &hkeysection);
+    }
+    free(regpath);
 
     return ret ? NULL : hkeysection;
 }
-- 
2.43.0

