From 1d91d62e04e11defa69177084e3b4d5e55e0eae8 Mon Sep 17 00:00:00 2001
From: Alex Henrie <alexhenrie24@gmail.com>
Date: Sun, 7 May 2023 22:34:15 -0600
Subject: [PATCH] winecfg: Mention protocol associations.

---
 programs/winecfg/winecfg.rc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/winecfg/winecfg.rc b/programs/winecfg/winecfg.rc
index 3f18020e8a8..7e58f1553ce 100644
--- a/programs/winecfg/winecfg.rc
+++ b/programs/winecfg/winecfg.rc
@@ -313,7 +313,7 @@ BEGIN
     CONTROL         "",IDC_SYSPARAM_SIZE_UD,UPDOWN_CLASSA,UDS_SETBUDDYINT | UDS_ALIGNRIGHT | WS_DISABLED, 185,75,15,13
 
     GROUPBOX        "MIME types",IDC_STATIC,8,95,244,23
-    CONTROL         "Manage file &associations",IDC_ENABLE_FILE_ASSOCIATIONS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,15,105,230,10
+    CONTROL         "Manage file and protocol &associations",IDC_ENABLE_FILE_ASSOCIATIONS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,15,105,230,10
 
     PUSHBUTTON      "&Font...",IDC_SYSPARAM_FONT,190,75,55,13,WS_DISABLED
     GROUPBOX        "Folders",IDC_STATIC,8,120,244,94
-- 
2.40.1

