From 931936f51db30da1b5d0583a9be3231b0f7678f0 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 6 Feb 2023 16:03:58 +1100
Subject: [PATCH] odbc32: Forward SQLTransact request onto driver

---
 dlls/odbc32/proxyodbc.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index f2ece198084..e78b51d7486 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -1591,11 +1591,24 @@ SQLRETURN WINAPI SQLTables(SQLHSTMT StatementHandle, SQLCHAR *CatalogName, SQLSM
  */
 SQLRETURN WINAPI SQLTransact(SQLHENV EnvironmentHandle, SQLHDBC ConnectionHandle, SQLUSMALLINT CompletionType)
 {
+    struct SQLHDBC_data *connection = ConnectionHandle;
     SQLRETURN ret = SQL_ERROR;
 
-    FIXME("(EnvironmentHandle %p, ConnectionHandle %p, CompletionType %d)\n", EnvironmentHandle, ConnectionHandle,
+    TRACE("(EnvironmentHandle %p, ConnectionHandle %p, CompletionType %d)\n", EnvironmentHandle, ConnectionHandle,
           CompletionType);
 
+    if (connection->type != SQL_HANDLE_DBC)
+    {
+        WARN("Wrong connection handle type %d\n", connection->type);
+        return SQL_ERROR;
+    }
+
+    if (connection->pSQLTransact)
+    {
+        ret = connection->pSQLTransact(connection->driver_env, connection->driver_hdbc, CompletionType);
+    }
+
+    TRACE("ret %d\n", ret);
     return ret;
 }
 
-- 
2.43.0

