From d3bfdb5fc584ae6a2bab70056f0cad11e4b741cd Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 8 Feb 2023 20:19:44 +1100
Subject: [PATCH] odbc32: Forward SQLColAttribute/W request onto driver

---
 dlls/odbc32/proxyodbc.c | 119 ++++++++++++++++++++++++++++++++++------
 1 file changed, 102 insertions(+), 17 deletions(-)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 3eca2ce9999..25e8249f8fc 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -314,6 +314,36 @@ static void connection_bind_sql_funcs(struct SQLHDBC_data *connection)
     LOAD_FUNCPTR(SQLTransact);
 }
 
+static SQLINTEGER map_odbc2_to_3(SQLINTEGER fieldid)
+{
+    switch( fieldid )
+    {
+        case SQL_COLUMN_COUNT:
+            return SQL_DESC_COUNT;
+        case SQL_COLUMN_NULLABLE:
+            return SQL_DESC_NULLABLE;
+        case SQL_COLUMN_NAME:
+            return SQL_DESC_NAME;
+        default:
+            return fieldid;
+    }
+}
+static SQLINTEGER map_odbc3_to_2(SQLINTEGER fieldid)
+{
+    switch( fieldid )
+    {
+        case SQL_DESC_COUNT:
+            return SQL_COLUMN_COUNT;
+        case SQL_DESC_NULLABLE:
+            return SQL_COLUMN_NULLABLE;
+        case SQL_DESC_NAME:
+            return SQL_COLUMN_NAME;
+        default:
+            return fieldid;
+    }
+}
+
+
 /*************************************************************************
  *				SQLAllocConnect           [ODBC32.001]
  */
@@ -557,12 +587,47 @@ SQLRETURN WINAPI SQLColAttribute(SQLHSTMT StatementHandle, SQLUSMALLINT ColumnNu
                                  SQLSMALLINT BufferLength, SQLSMALLINT *StringLength,
                                  SQLLEN *NumericAttribute)
 {
+    struct SQLHSTMT_data *statement = StatementHandle;
     SQLRETURN ret = SQL_ERROR;
 
-    FIXME("(StatementHandle %p, ColumnNumber %d, FieldIdentifier %d, CharacterAttribute %p, BufferLength %d,"
+    TRACE("(StatementHandle %p, ColumnNumber %d, FieldIdentifier %d, CharacterAttribute %p, BufferLength %d,"
           " StringLength %p, NumericAttribute %p)\n", StatementHandle, ColumnNumber, FieldIdentifier,
           CharacterAttribute, BufferLength, StringLength, NumericAttribute);
 
+    if (statement->type != SQL_HANDLE_STMT)
+    {
+        WARN("Wrong handle type %d\n", statement->type);
+        return SQL_ERROR;
+    }
+
+    /* ODBC 3.0 */
+    if (statement->connection->pSQLColAttribute)
+    {
+        ret = statement->connection->pSQLColAttribute(statement->driver_stmt, ColumnNumber, FieldIdentifier,
+                                 CharacterAttribute, BufferLength, StringLength, NumericAttribute);
+    }
+    /* ODBC 2.0 */
+    else if (statement->connection->pSQLColAttributes)
+    {
+        FieldIdentifier = map_odbc3_to_2(FieldIdentifier);
+        ret = statement->connection->pSQLColAttributes(statement->driver_stmt, ColumnNumber, FieldIdentifier,
+                                 CharacterAttribute, BufferLength, StringLength, NumericAttribute);
+
+        /* Convert back for ODBC3 drivers */
+        if (NumericAttribute && FieldIdentifier == SQL_COLUMN_TYPE &&
+                statement->connection->driver_ver == SQL_OV_ODBC2 &&
+                statement->connection->environment->version == SQL_OV_ODBC3)
+        {
+            if (*NumericAttribute == SQL_TIME)
+                *NumericAttribute = SQL_TYPE_TIME;
+            else if (*NumericAttribute == SQL_DATETIME)
+                *NumericAttribute = SQL_TYPE_DATE;
+            else if (*NumericAttribute == SQL_TIMESTAMP)
+                *NumericAttribute = SQL_TYPE_TIMESTAMP;
+        }
+    }
+
+    TRACE("ret %d\n", ret);
     return ret;
 }
 
@@ -2063,21 +2128,6 @@ SQLRETURN WINAPI SQLSetScrollOptions(SQLHSTMT statement_handle, SQLUSMALLINT f_c
     return ret;
 }
 
-static SQLINTEGER map_odbc2_to_3(SQLINTEGER fieldid)
-{
-    switch( fieldid )
-    {
-        case SQL_COLUMN_COUNT:
-            return SQL_DESC_COUNT;
-        case SQL_COLUMN_NULLABLE:
-            return SQL_DESC_NULLABLE;
-        case SQL_COLUMN_NAME:
-            return SQL_DESC_NAME;
-        default:
-            return fieldid;
-    }
-}
-
 /*************************************************************************
  *				SQLColAttributesW          [ODBC32.106]
  */
@@ -2284,12 +2334,47 @@ SQLRETURN WINAPI SQLColAttributeW(SQLHSTMT StatementHandle, SQLUSMALLINT ColumnN
                                   SQLSMALLINT BufferLength, SQLSMALLINT *StringLength,
                                   SQLLEN *NumericAttribute)
 {
+    struct SQLHSTMT_data *statement = StatementHandle;
     SQLRETURN ret = SQL_ERROR;
 
-    FIXME("StatementHandle %p ColumnNumber %d FieldIdentifier %d CharacterAttribute %p BufferLength %d"
+    TRACE("StatementHandle %p ColumnNumber %d FieldIdentifier %d CharacterAttribute %p BufferLength %d"
           " StringLength %p NumericAttribute %p\n", StatementHandle, ColumnNumber, FieldIdentifier,
           CharacterAttribute, BufferLength, StringLength, NumericAttribute);
 
+    if (statement->type != SQL_HANDLE_STMT)
+    {
+        WARN("Wrong handle type %d\n", statement->type);
+        return SQL_ERROR;
+    }
+
+    /* ODBC 3.0 */
+    if (statement->connection->pSQLColAttributeW)
+    {
+        ret = statement->connection->pSQLColAttributeW(statement->driver_stmt, ColumnNumber, FieldIdentifier,
+                                 CharacterAttribute, BufferLength, StringLength, NumericAttribute);
+    }
+    /* ODBC 2.0 */
+    else if (statement->connection->pSQLColAttributesW)
+    {
+        FieldIdentifier = map_odbc3_to_2(FieldIdentifier);
+        ret = statement->connection->pSQLColAttributesW(statement->driver_stmt, ColumnNumber, FieldIdentifier,
+                                 CharacterAttribute, BufferLength, StringLength, NumericAttribute);
+
+        /* Convert back for ODBC3 drivers */
+        if (NumericAttribute && FieldIdentifier == SQL_COLUMN_TYPE &&
+                statement->connection->driver_ver == SQL_OV_ODBC2 &&
+                statement->connection->environment->version == SQL_OV_ODBC3)
+        {
+            if (*NumericAttribute == SQL_TIME)
+                *NumericAttribute = SQL_TYPE_TIME;
+            else if (*NumericAttribute == SQL_DATETIME)
+                *NumericAttribute = SQL_TYPE_DATE;
+            else if (*NumericAttribute == SQL_TIMESTAMP)
+                *NumericAttribute = SQL_TYPE_TIMESTAMP;
+        }
+    }
+
+    TRACE("ret %d\n", ret);
     return ret;
 }
 
-- 
2.43.0

