From cbfeb2a5abed1c5c532280290f70f5299894cc8f Mon Sep 17 00:00:00 2001
From: Czcibor Bohusz-Dobosz <czbd@o2.pl>
Date: Sun, 14 Jan 2024 17:50:49 +1100
Subject: [PATCH] msi: Only process cabinet files if it exists

---
 dlls/msi/files.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/msi/files.c b/dlls/msi/files.c
index 96c6c5eaa4b..c85c66ae53e 100644
--- a/dlls/msi/files.c
+++ b/dlls/msi/files.c
@@ -819,7 +819,7 @@ UINT ACTION_PatchFiles( MSIPACKAGE *package )
             data.cb      = patchfiles_cb;
             data.user    = &cursor;
 
-            if (!msi_cabextract( package, mi, &data ))
+            if (mi->cabinet && !msi_cabextract( package, mi, &data ))
             {
                 ERR("Failed to extract cabinet: %s\n", debugstr_w(mi->cabinet));
                 rc = ERROR_INSTALL_FAILURE;
-- 
2.43.0

