From ef059a443a88414b0c551608b5f8302002d5f6f0 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 28 Sep 2021 14:41:16 +1000
Subject: [PATCH] sapi: Return dump object in ISpObjectTokenEnumBuilder Next

---
 dlls/sapi/token.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/dlls/sapi/token.c b/dlls/sapi/token.c
index 859d916fa43..12b3e0e8a6e 100644
--- a/dlls/sapi/token.c
+++ b/dlls/sapi/token.c
@@ -705,6 +705,7 @@ static HRESULT WINAPI token_enum_Next( ISpObjectTokenEnumBuilder *iface,
                                        ULONG *fetched )
 {
     struct token_enum *This = impl_from_ISpObjectTokenEnumBuilder( iface );
+    HRESULT hr;
 
     TRACE( "(%p)->(%lu %p %p)\n", This, num, tokens, fetched );
 
@@ -712,8 +713,12 @@ static HRESULT WINAPI token_enum_Next( ISpObjectTokenEnumBuilder *iface,
 
     FIXME( "semi-stub: Returning an empty enumerator\n" );
 
-    if (fetched) *fetched = 0;
-    return S_FALSE;
+    hr = token_create( NULL, &IID_ISpObjectToken, (void**)tokens );
+    if (FAILED(hr))
+        return hr;
+
+    if (fetched) *fetched = 1;
+    return hr;
 }
 
 static HRESULT WINAPI token_enum_Skip( ISpObjectTokenEnumBuilder *iface,
-- 
2.34.1

