From 197bb91afcc7af54124fd5113f0acdb25fcf959d Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 29 May 2021 18:02:04 +1000
Subject: [PATCH 3/4] wbemprox: Support VT_BYREF in to_longlong

---
 dlls/wbemprox/query.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/wbemprox/query.c b/dlls/wbemprox/query.c
index ec2a43c3f5f..ccff6aa0b4f 100644
--- a/dlls/wbemprox/query.c
+++ b/dlls/wbemprox/query.c
@@ -1305,6 +1305,9 @@ HRESULT to_longlong( VARIANT *var, LONGLONG *val, CIMTYPE *type )
         *val = 0;
         return S_OK;
     }
+    if (V_VT( var ) & VT_BYREF)
+        var = V_VARIANTREF( var );
+
     if (V_VT( var ) & VT_ARRAY)
     {
         *val = (INT_PTR)to_array( var, type );
-- 
2.30.2

