From 8e4c76596335097ca35af9d289c2096dc274b4d3 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 6 Aug 2017 02:50:23 +0200
Subject: [PATCH] packager: Prefer native version.

---
 dlls/packager/Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/packager/Makefile.in b/dlls/packager/Makefile.in
index a90014ce25d..747eed203ae 100644
--- a/dlls/packager/Makefile.in
+++ b/dlls/packager/Makefile.in
@@ -1,7 +1,7 @@
 MODULE = packager.dll
 IMPORTS = uuid shell32 shlwapi user32
 
-EXTRADLLFLAGS = -mno-cygwin
+EXTRADLLFLAGS = -mno-cygwin -Wb,--prefer-native
 
 C_SRCS = \
 	packager_main.c
-- 
2.20.1

