From bc118983b67f87ed7858711d03cff34050b92770 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 6 Aug 2017 02:50:23 +0200
Subject: [PATCH] packager: Prefer native version.

---
 dlls/packager/Makefile.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/packager/Makefile.in b/dlls/packager/Makefile.in
index 5fc89d807a8..3589b37be9c 100644
--- a/dlls/packager/Makefile.in
+++ b/dlls/packager/Makefile.in
@@ -1,5 +1,6 @@
 MODULE = packager.dll
 IMPORTS = uuid shell32 shlwapi user32
+EXTRADLLFLAGS = -Wb,--prefer-native
 
 SOURCES = \
 	packager.rc \
-- 
2.42.0

