From 520b5366cb7ce1cba59a755a734f23dbd25fa093 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 22 Apr 2021 13:23:44 +1000
Subject: [PATCH] xactengine: Use nameless unions

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 include/xact3.h   | 2 +-
 include/xact3wb.h | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/xact3.h b/include/xact3.h
index 63eec707363..9b8f480829e 100644
--- a/include/xact3.h
+++ b/include/xact3.h
@@ -434,7 +434,7 @@ typedef struct XACT_NOTIFICATION
         XACT_NOTIFICATION_VARIABLE variable;
         XACT_NOTIFICATION_GUI gui;
         XACT_NOTIFICATION_WAVE wave;
-    } DUMMYUNIONNAME;
+    };
 } XACT_NOTIFICATION, *LPXACT_NOTIFICATION;
 typedef const XACT_NOTIFICATION *LPCXACT_NOTIFICATION;
 
diff --git a/include/xact3wb.h b/include/xact3wb.h
index 6d157a466f9..834b3e1ef54 100644
--- a/include/xact3wb.h
+++ b/include/xact3wb.h
@@ -121,7 +121,7 @@ typedef struct WAVEBANKENTRY
             DWORD Duration : 28;
         } DUMMYSTRUCTNAME;
         DWORD dwFlagsAndDuration;
-    } DUMMYUNIONNAME;
+    };
 
     WAVEBANKMINIWAVEFORMAT Format;
     WAVEBANKREGION PlayRegion;
-- 
2.30.2

