From 79afe407ea3e0a5496f6a4e488785cc73d1c3a5d Mon Sep 17 00:00:00 2001
From: llde <lorenzofersteam@live.it>
Date: Sat, 10 Nov 2018 01:22:09 +0100
Subject: Return correct position when the position is inferior to a  period

Signed-off-by: llde <lorenzofersteam@live.it>
---
 dlls/winepulse.drv/mmdevdrv.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/dlls/winepulse.drv/mmdevdrv.c b/dlls/winepulse.drv/mmdevdrv.c
index fb2e3678f8..d578031c63 100644
--- a/dlls/winepulse.drv/mmdevdrv.c
+++ b/dlls/winepulse.drv/mmdevdrv.c
@@ -2770,9 +2770,7 @@ static HRESULT WINAPI AudioClock_GetPosition(IAudioClock *iface, UINT64 *pos,
     *pos = This->clock_written - This->held_bytes;
 
     if(This->started){
-        if(*pos < This->period_bytes)
-            *pos = 0;
-        else if(This->held_bytes > This->period_bytes)
+        if(This->held_bytes > This->period_bytes)
             *pos -= This->period_bytes;
     }
 
-- 
2.19.1

