From 9c00130e319055e1b8fb336338e656f8ca5d64b9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 19 Jan 2021 13:47:45 +0100
Subject: [PATCH] include: Define DECLSPEC_SELECTANY on MinGW.

---
 include/winnt.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/winnt.h b/include/winnt.h
index 5b3efe61385..bb045fa2745 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -110,6 +110,8 @@ extern "C" {
 #ifndef DECLSPEC_SELECTANY
 #if defined(_MSC_VER) && (_MSC_VER >= 1100)
 #define DECLSPEC_SELECTANY __declspec(selectany)
+#elif defined(__MINGW32__)
+#define DECLSPEC_SELECTANY __attribute__((selectany))
 #else
 #define DECLSPEC_SELECTANY
 #endif
-- 
2.29.2

