From fcb515de7ef61b806c7da1fc57875fb60edb59db Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 4 Feb 2021 13:40:11 +0100
Subject: [PATCH] widl: Fold inherit cases by using typename rule in
 qualified_type.

---
 tools/widl/parser.y | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/tools/widl/parser.y b/tools/widl/parser.y
index d3a79e9fcb7..bbf71a2e3ef 100644
--- a/tools/widl/parser.y
+++ b/tools/widl/parser.y
@@ -892,7 +892,7 @@ namespace_pfx:
 	;
 
 qualified_type:
-	  aKNOWNTYPE				{ $$ = find_type_or_error(current_namespace, $1); }
+	  typename				{ $$ = find_type_or_error(current_namespace, $1); }
 	| namespace_pfx typename		{ $$ = find_type_or_error($1, $2); }
 	;
 
@@ -960,11 +960,6 @@ interfacedef: attributes interface inherit
 	  '{' int_statements '}' semicolon_opt	{ $$ = type_interface_define($2, $1, $3, $5);
 						  check_async_uuid($$);
 						}
-/* MIDL is able to import the definition of a base class from inside the
- * definition of a derived class, I'll try to support it with this rule */
-	| attributes interface ':' aIDENTIFIER
-	  '{' import int_statements '}'
-	   semicolon_opt			{ $$ = type_interface_define($2, $1, find_type_or_error(current_namespace, $4), $7); }
 	| dispinterfacedef semicolon_opt	{ $$ = $1; }
 	;
 
-- 
2.20.1

