From dc7e23bd6fc0a32c640957dfe8bab603ef800871 Mon Sep 17 00:00:00 2001
From: Steve Lhomme <robux4@ycbcr.xyz>
Date: Thu, 7 Jan 2021 14:14:22 +0100
Subject: [PATCH] widl: Define the C type name as an alias for the C++
 qualified name.

---
 tools/widl/header.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tools/widl/header.c b/tools/widl/header.c
index 60f0cd96fce..e635f6360f7 100644
--- a/tools/widl/header.c
+++ b/tools/widl/header.c
@@ -1483,6 +1483,8 @@ static void write_forward(FILE *header, type_t *iface)
   fprintf(header, "#define __%s_FWD_DEFINED__\n", iface->c_name);
   fprintf(header, "typedef interface %s %s;\n", iface->c_name, iface->c_name);
   fprintf(header, "#ifdef __cplusplus\n");
+  if (iface->namespace && !is_global_namespace(iface->namespace))
+    fprintf(header, "#define %s %s\n", iface->c_name, iface->qualified_name);
   write_namespace_start(header, iface->namespace);
   if (strchr(iface->name, '<')) write_line(header, 0, "template<> struct %s;", iface->name);
   else write_line(header, 0, "interface %s;", iface->name);
-- 
2.29.2

