From 7ab586e1da370ab17f5cc64dc97449a436f2fa5e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 25 Oct 2021 11:22:04 +0200
Subject: [PATCH] server: Stop enforcing relative rawinput mouse positions.

---
 server/queue.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index 87b25820cfe..a94bdf53939 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -1967,10 +1967,10 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
         raw_msg.message    = WM_INPUT;
 
         raw_msg.info                = input->mouse.info;
-        raw_msg.flags               = flags & ~(MOUSEEVENTF_ABSOLUTE | MOUSEEVENTF_VIRTUALDESK);
+        raw_msg.flags               = flags;
         raw_msg.rawinput.type       = RIM_TYPEMOUSE;
-        raw_msg.rawinput.mouse.x    = x - desktop->cursor.x;
-        raw_msg.rawinput.mouse.y    = y - desktop->cursor.y;
+        raw_msg.rawinput.mouse.x    = (flags & MOUSEEVENTF_MOVE) ? input->mouse.x : 0;
+        raw_msg.rawinput.mouse.y    = (flags & MOUSEEVENTF_MOVE) ? input->mouse.y : 0;
         raw_msg.rawinput.mouse.data = input->mouse.data;
 
         enum_processes( queue_rawinput_message, &raw_msg );
-- 
2.43.0

