From 3407a396a86366b81b243fde01e191aa1a2f7f5a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 26 Sep 2014 20:24:50 +0200
Subject: [PATCH] user32: Call UpdateWindow() during DIALOG_CreateIndirect.

---
 dlls/user32/dialog.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/dialog.c b/dlls/user32/dialog.c
index a226c764c49..91ce53f8297 100644
--- a/dlls/user32/dialog.c
+++ b/dlls/user32/dialog.c
@@ -703,6 +703,7 @@ static HWND DIALOG_CreateIndirect( HINSTANCE hInst, LPCVOID dlgTemplate,
         if (template.style & WS_VISIBLE && !(GetWindowLongW( hwnd, GWL_STYLE ) & WS_VISIBLE))
         {
            NtUserShowWindow( hwnd, SW_SHOWNORMAL );   /* SW_SHOW doesn't always work */
+            UpdateWindow( hwnd );
         }
         return hwnd;
     }
-- 
2.35.1

