From 29866a3cb6dc799b060565d94608c7b25bd26509 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Fri, 21 May 2021 23:44:39 -0500
Subject: [PATCH] d3d11/tests: Add a couple of extra tests for SRV/RTV
 conflict.

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 dlls/d3d11/tests/d3d11.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/d3d11/tests/d3d11.c b/dlls/d3d11/tests/d3d11.c
index 47c6613d904..5ab08632367 100644
--- a/dlls/d3d11/tests/d3d11.c
+++ b/dlls/d3d11/tests/d3d11.c
@@ -32364,6 +32364,15 @@ static void test_deferred_context_state(void)
     ID3D11DeviceContext_PSGetShaderResources(deferred, 0, 1, &ret_srv);
     ok(!ret_srv, "Got unexpected SRV %p.\n", ret_srv);
 
+    ID3D11DeviceContext_PSSetShaderResources(deferred, 0, 1, &srv);
+    ID3D11DeviceContext_PSGetShaderResources(deferred, 0, 1, &ret_srv);
+    ok(!ret_srv, "Got unexpected SRV %p.\n", ret_srv);
+
+    ID3D11DeviceContext_PSSetShaderResources(immediate, 0, 1, &srv);
+    ID3D11DeviceContext_PSGetShaderResources(immediate, 0, 1, &ret_srv);
+    ok(ret_srv == srv, "Got unexpected SRV %p.\n", ret_srv);
+    ID3D11ShaderResourceView_Release(ret_srv);
+
     ID3D11ShaderResourceView_Release(srv);
     ID3D11RenderTargetView_Release(rtv);
     ID3D11Texture2D_Release(texture);
-- 
2.30.2

