From 6fcb4ea61ecad68704e7793e56d69c5144e96716 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Wed, 23 Jun 2021 17:32:26 -0500
Subject: [PATCH] d3d11: Forbid map types other than DISCARD and NOOVERWRITE on
 a deferred context.

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 dlls/d3d11/device.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index 56f34c70ea1..d043af7051a 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -744,6 +744,10 @@ static HRESULT STDMETHODCALLTYPE d3d11_device_context_Map(ID3D11DeviceContext1 *
     if (map_flags)
         FIXME("Ignoring map_flags %#x.\n", map_flags);
 
+    if (context->type != D3D11_DEVICE_CONTEXT_IMMEDIATE
+            && map_type != D3D11_MAP_WRITE_DISCARD && map_type != D3D11_MAP_WRITE_NO_OVERWRITE)
+        return E_INVALIDARG;
+
     wined3d_resource = wined3d_resource_from_d3d11_resource(resource);
 
     wined3d_mutex_lock();
-- 
2.30.2

