diff --git a/dlls/wined3d/context.c b/dlls/wined3d/context.c
index 83e3563..abcc95f 100644
--- a/dlls/wined3d/context.c
+++ b/dlls/wined3d/context.c
@@ -1929,10 +1929,16 @@ struct wined3d_context *context_create(struct wined3d_swapchain *swapchain,
         goto out;
     }
 
-    if (!(context->hdc = GetDCEx(swapchain->win_handle, 0, DCX_USESTYLE | DCX_CACHE)))
-    {
+    context->hdc = NULL;
+
+   if (swapchain->win_handle == GetDesktopWindow())
+       TRACE("Attempted to retrieve device context of desktop window, trying swapchain backup.\n");
+   else if (!(context->hdc = GetDC(swapchain->win_handle)))
+
         WARN("Failed to retrieve device context, trying swapchain backup.\n");
 
+   if (!context->hdc)
+   {
         if ((context->hdc = swapchain_get_backup_dc(swapchain)))
             context->hdc_is_private = TRUE;
         else
