From 0d2d659fbfb73fe046626d256ca5ee2878d9288b Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 8 Feb 2021 13:16:52 +1100
Subject: [PATCH 2/2] ntdll: RtlFirstFreeAce only return FALSE on error

---
 dlls/ntdll/sec.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/sec.c b/dlls/ntdll/sec.c
index 5c1c9b1fbb0..9cebcb4f58e 100644
--- a/dlls/ntdll/sec.c
+++ b/dlls/ntdll/sec.c
@@ -1149,9 +1149,8 @@ BOOLEAN WINAPI RtlFirstFreeAce(
 			return FALSE;
 		ace = (PACE_HEADER)(((BYTE*)ace)+ace->AceSize);
 	}
-	if ((BYTE *)ace >= (BYTE *)acl + acl->AclSize)
-		return FALSE;
-	*x = ace;
+	if ((BYTE *)ace <= (BYTE *)acl + acl->AclSize)
+		*x = ace;
 	return TRUE;
 }
 
-- 
2.30.1

