From 1791ceee2e4f4c677ef65aef8dd79c85fb06d947 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 30 Jul 2022 18:08:00 +1000
Subject: [PATCH] winspool.drv: Add more tracing to ClosePrinter

---
 dlls/winspool.drv/info.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/winspool.drv/info.c b/dlls/winspool.drv/info.c
index 5a74cb29bd9..fe6782daf40 100644
--- a/dlls/winspool.drv/info.c
+++ b/dlls/winspool.drv/info.c
@@ -2925,15 +2925,18 @@ BOOL WINAPI ClosePrinter(HANDLE hPrinter)
             LIST_FOR_EACH_SAFE(cursor, cursor2, &printer->queue->jobs)
             {
                 job_t *job = LIST_ENTRY(cursor, job_t, entry);
+                TRACE("Scheduling Job: %p\n", job);
                 ScheduleJob(hPrinter, job->job_id);
             }
             free(printer->queue);
         }
 
         if (printer->backend_printer) {
+            TRACE("Closing Bankend printer\n");
             backend->fpClosePrinter(printer->backend_printer);
         }
 
+        TRACE("Freeing Printer entry\n");
         free_printer_entry( printer );
         printer_handles[i - 1] = NULL;
         LeaveCriticalSection(&printer_handles_cs);
-- 
2.35.1

