From 4aa49cd10f1cd190f088e009898cf70d19057b00 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 30 Jul 2022 18:08:00 +1000
Subject: [PATCH] winspool.drv: Add more tracing to ClosePrinter

---
 dlls/winspool.drv/info.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/winspool.drv/info.c b/dlls/winspool.drv/info.c
index 656477553ba..5a59916b0e2 100644
--- a/dlls/winspool.drv/info.c
+++ b/dlls/winspool.drv/info.c
@@ -2751,10 +2751,13 @@ BOOL WINAPI ClosePrinter(HANDLE hPrinter)
     if(printer)
     {
         TRACE("closing %s\n", debugstr_w(printer->name));
+
         if (printer->backend_printer) {
+            TRACE("Closing Bankend printer\n");
             backend->fpClosePrinter(printer->backend_printer);
         }
 
+        TRACE("Freeing Printer entry\n");
         free_printer_entry( printer );
         printer_handles[i - 1] = NULL;
         LeaveCriticalSection(&printer_handles_cs);
-- 
2.38.1

