From 5d62ab1fd576aa9e11a387a6b2a1b250aecc1803 Mon Sep 17 00:00:00 2001
From: Eric Pouech <eric.pouech@gmail.com>
Date: Fri, 5 Nov 2021 07:56:04 +1100
Subject: [PATCH] configure.ac: let dwarf4 be Wine's default debug format

---
 configure.ac | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/configure.ac b/configure.ac
index 57bd9a6f9ec..0196af56783 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1051,8 +1051,7 @@ then
         if test "x$ac_debug_format_seen" = x
         then
           case $CROSSDEBUG in
-            *dwarf)  WINE_TRY_CROSSCFLAGS([-gdwarf-2])
-                     WINE_TRY_CROSSCFLAGS([-gstrict-dwarf]) ;;
+            *dwarf)  WINE_TRY_CROSSCFLAGS([-gdwarf-4]) ;;
             pdb)     WINE_TRY_CROSSCFLAGS([-gcodeview]) ;;
           esac
         fi
@@ -1865,8 +1864,7 @@ char*f(const char *h,char n) {return strchr(h,n);}]])],[ac_cv_c_logicalop_noisy=
   done
   if test "x$ac_debug_format_seen" = xdefault
   then
-    WINE_TRY_CFLAGS([-gdwarf-2])
-    WINE_TRY_CFLAGS([-gstrict-dwarf])
+    WINE_TRY_CFLAGS([-gdwarf-4])
   fi
 
   dnl Disable gcc builtins except for Mingw
-- 
2.33.0

