From d1a6cc9b2bcc0a6acbf86ad96c76f79a84abff78 Mon Sep 17 00:00:00 2001
From: Gijs Vermeulen <gijsvrm@gmail.com>
Date: Sat, 7 Apr 2018 16:07:00 +0200
Subject: [PATCH] ntoskrnl.exe: Improve ObReferenceObjectByHandle stub.

Signed-off-by: Gijs Vermeulen <gijsvrm@gmail.com>
---
 dlls/ntoskrnl.exe/ntoskrnl.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index 7258c3773f..f80207daf6 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -2231,7 +2231,11 @@ NTSTATUS WINAPI ObReferenceObjectByHandle( HANDLE obj, ACCESS_MASK access,
                                            POBJECT_HANDLE_INFORMATION info)
 {
     FIXME( "stub: %p %x %p %d %p %p\n", obj, access, type, mode, ptr, info);
-    return STATUS_NOT_IMPLEMENTED;
+
+    if(ptr)
+        *ptr = UlongToHandle(0xdeadbeaf);
+
+    return STATUS_SUCCESS;
 }
 
  /***********************************************************************
-- 
2.13.6

