From ecc14aeb2e510520705e98ebb39e9b2fcb62b453 Mon Sep 17 00:00:00 2001
From: David Torok <dt@zeroitlab.com>
Date: Sun, 17 Nov 2019 19:08:12 +0100
Subject: [PATCH] Send WM_NCPOINTERUP on focus regain

---
 dlls/win32u/input.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index a4834b740d8..5d14779538c 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2045,6 +2045,9 @@ BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus, DWORD new
         send_message( hwnd, WM_ACTIVATE,
                       MAKEWPARAM( mouse ? WA_CLICKACTIVE : WA_ACTIVE, is_iconic(hwnd) ),
                       (LPARAM)previous );
+
+        send_message( hwnd, WM_NCPOINTERUP, 0, 0);
+
         if (NtUserGetAncestor( hwnd, GA_PARENT ) == get_desktop_window())
             NtUserPostMessage( get_desktop_window(), WM_PARENTNOTIFY, WM_NCACTIVATE, (LPARAM)hwnd );
     }
-- 
2.47.2

